## Bxy_strains
### Mucin Metabolism Genes Underlie the Key Renovations of Bacteroides xylanisolvens Genomes in Captive Great Apes

By Alex Nishida and Howard Ochman

2021

This manuscript has been submitted

#### How to use this repository

##### 1. Reproduce results 
[generate all figures and tables using processed results provided in this repository](https://github.com/ahnishida/Bxy_strains/blob/main/figures_tables.md)

##### 2. Explore generation of processed results
[genome assembly and QC of strains sequenced by this study](https://github.com/ahnishida/Bxy_strains/tree/main/scripts/isolate_genome_assembly)

[fetching published ncbi genomes and QC](https://github.com/ahnishida/Bxy_strains/tree/main/scripts/filter_ncbi_genomes)

[generating species pangenome, phylogeny, and annotation of gene groups](https://github.com/ahnishida/Bxy_strains/blob/main/scripts/pangenome_processing_annotation/processing_pangenome.md)

[run count and define phylogenetic independent constrasts](https://github.com/ahnishida/Bxy_strains/blob/main/scripts/pangenome_analyses/gene_gain_loss.md)

[determine potential source of genes gained by Bxy captive ape associated strains](https://github.com/ahnishida/Bxy_strains/blob/main/scripts/pangenome_analyses/identifying_source_genes.md)

[identify the PUL region gained by captive apes and output fastas](https://github.com/ahnishida/Bxy_strains/blob/main/scripts/pangenome_analyses/output_PUL_fasta.md)


##### 3. Apply gene-gain-loss methodology to your data
[Explore functions using test data provided](https://github.com/ahnishida/Bxy_strains/blob/main/scripts/pangenome_analyses/gene_gain_loss_functions.R)


