---
title: "Untitled"
author: "Alex Nishida"
date: "3/19/2021"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir ='/stor/work/Ochman/alex/captive_ape_strain_genomes')
library(ape)
library(treeio)
library(ggtree)
library(estimatr)
library(tidyverse)
library(seqinr)
library(stringr)
library(cowplot)
library(harrietr)
```


```{r}
getwd()
dir = 'results/pangenome/Bacteroides_xylanisolvens_outgroup'
system(paste0('mkdir -pv ',file.path(dir,'analyses')))
```

## Bxylan analyses

3 Bovatus genomes included in roary run

Raxml with 100 bootstraps run on tree

#### Reformat metadata
```{r}
metadata <- read.table('metadata/strain_ncbi_metadata_assembly_all.txt',header=T)
metadata <- metadata %>%
  mutate(taxonomy_Species = str_replace_all(taxonomy_Species,' ','_'),
         isolate = str_replace_all(isolate,'[_-]','.'),
         site = recode(site,'USA: Cambridge'='USA','USA:Boston'='USA',
                       'China: Shenzhen'='China','USA:Seattle'='USA',     
                       'USA:Baltimore'='USA', 'not applicable'='siteUnknown',
                       'missing'='siteUnknown')) %>%
  unite(host_site, host, site, sep = "_", remove = FALSE) 
```


#### Filter out Bxy assemblies low quality 
```{r}
#Bxy roary was run with exclude lower quality assemblies
#completeness and contamination associated with genome size and # of predicted genes
Bxy_metadata <- metadata %>% filter(taxonomy_Species == 'Bacteroides_xylanisolvens')
summary(lm_robust(Genome.size ~ Contamination + Completeness, data=Bxy_metadata))
summary(lm_robust(X..predicted.genes ~ Contamination+ Completeness, data=Bxy_metadata))

#apply qc filter
Bxy_metadata_filtered <- metadata %>% 
  filter(taxonomy_Species == 'Bacteroides_xylanisolvens') %>%
  filter(Completeness>94&Contamination<5)
summary(lm_robust(Genome.size ~ Contamination + Completeness, data=Bxy_metadata_filtered))
summary(lm_robust(X..predicted.genes ~ Contamination + Completeness, data=Bxy_metadata_filtered))
#completeness is negatively associated with # of predicted genes, s 
ggplot(Bxy_metadata_filtered,aes(x=Completeness,y=X..predicted.genes)) + 
  geom_point() + theme_bw()

```

#### Define samples to exclude
```{r}
samples_to_exclude <- metadata %>% 
  filter(Completeness<94|Contamination>5)
samples_to_exclude <- samples_to_exclude$isolate
metadata <- metadata %>% 
  filter(Completeness>94&Contamination<5)
```

#### Root tree and subset metadata 
```{r}
Bxy_tree_file <- "phylogeny/RAxML_bipartitions.Bacteroides_xylanisolvens_outgroup"
Bxy_tree <- ape::read.tree(file.path(dir,Bxy_tree_file))
Bxy_tree$tip.label <- str_replace_all(Bxy_tree$tip.label,'[_-]','.') #clean up names
Bxy_tree <- drop.tip(Bxy_tree,samples_to_exclude) #remove low QC assemblies
setdiff(Bxy_tree$tip.label,metadata$isolate)
Bxy_metadata <- metadata %>% filter(isolate %in% Bxy_tree$tip.label)
write.table(Bxy_metadata,file=file.path(dir,'analyses/Bxy_metadata.txt'))

ggtree(Bxy_tree) + geom_nodelab()
Bov <- metadata %>% filter(taxonomy_Species == 'Bacteroides_ovatus') 
Bov_outgroup <- intersect(Bxy_tree$tip.label,Bov$isolate)
Bxy_tree <- root(Bxy_tree,outgroup=Bov_outgroup) 

get_color_palette <- function(tips) {
  Bxy_metadata <- Bxy_metadata %>% filter(isolate %in% tips)
  vec <- sort(unique(Bxy_metadata$host_site))
  print(vec)
  return(recode(vec,
                          'human_USA'='blue3',
                          'human_China' = 'cyan',
                          'bonobo_Columbus_Zoo'='red2',
                          'chimpanzee_Houston_Zoo'='orange2',
                          'orangutan_Houston_Zoo'='purple4',
                          'gorilla_Columbus_Zoo'='green3',
                          'chicken_siteUnknown'='tan',
                          'missing_siteUnknown'='brown4'))}
host_site_color <- get_color_palette(Bxy_tree$tip.label)

Bxy_tree_plot <- ggtree(Bxy_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  geom_tiplab() + 
  xlim(NA,.5)
Bxy_tree_plot

write.tree(Bxy_tree,file=file.path(dir,'analyses/Bxy.tre'))
ggsave(Bxy_tree_plot,file=file.path(dir,'analyses/Bxy_phylogeny.pdf'),height=15,width=8)
```
#### Add clades to tree and metadata
```{r}
cladeA_MCRA <- ape::getMRCA(Bxy_tree,c('GCA.009102525.1.ASM910252v1','GCA.009093695.1.ASM909369v1'))
cladeA_tree <- extract.clade(Bxy_tree,cladeA_MCRA)
ggtree(cladeA_tree) %<+% metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeA_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)
cladeB_MCRA <- ape::getMRCA(Bxy_tree,c('P17.D4','GCA.003464445.1.ASM346444v1'))
cladeB_tree <- extract.clade(Bxy_tree,cladeB_MCRA)
ggtree(cladeB_tree) %<+% metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeB_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)
cladeC_MCRA <- ape::getMRCA(Bxy_tree,c('P21.4E','GCA.009102685.1.ASM910268v1'))
cladeC_tree <- extract.clade(Bxy_tree,cladeC_MCRA)
ggtree(cladeC_tree) %<+% metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeC_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

Bxy_metadata <- Bxy_metadata %>% mutate(clade = ifelse(
  isolate %in% cladeA_tree$tip.label,'cladeA',
    ifelse(isolate %in% cladeB_tree$tip.label,'cladeB',
      ifelse(isolate %in% cladeC_tree$tip.label,'cladeC','unassigned'))))
```

#### Captive clade assignment
```{r}
captive1Node  <- ape::getMRCA(Bxy_tree,c('P17.D4','P13.A9',"GCA.003468875.1.ASM346887v1"))
captive1Human <- c("GCA.003468875.1.ASM346887v1")
captive1tree <- extract.clade(Bxy_tree,captive1Node)
captive2Node  <- ape::getMRCA(Bxy_tree,c('P21.4E','P21.11C',"GCA.003458755.1.ASM345875v1"))
captive2tree <- extract.clade(Bxy_tree,captive2Node)
captive2Human <- c("GCA.003458755.1.ASM345875v1")
captive3Node  <- ape::getMRCA(Bxy_tree,c('P21.6E','P21.2G',"GCA.003436085.1.ASM343608v1"))
captive3tree <- extract.clade(Bxy_tree,captive3Node)
captive3Human <- c("GCA.003458755.1.ASM345875v1")

Bxy_tree_plot_cladelabels <- ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  xlim(NA,.5)+ 
  geom_cladelabel(node=captive1Node, color='black', offset=.01,
                  label="Mixed-host captive clade") + 
  geom_cladelabel(node=captive2Node, color='black', offset=.01,label="Captive gorilla clade2") +
  geom_cladelabel(node=captive3Node, color='black', offset=.01,label="Captive gorilla clade1") +
   geom_cladelabel(node=cladeA_MCRA, color='black', offset=.02,label="CladeA") + 
  geom_cladelabel(node=cladeB_MCRA, color='black', offset=.22,label="CladeB") +
  geom_cladelabel(node=cladeC_MCRA, color='black', offset=.2,label="CladeC") +
  xlim(NA,.5)
Bxy_tree_plot_cladelabels
ggsave(Bxy_tree_plot_cladelabels,
       file=file.path(dir,'analyses/Fig1_Bxy_phylogeny_cladeLabels.pdf'),height=9,width=8)


Bxy_metadata <- Bxy_metadata %>% mutate(captive_clade = ifelse(
  isolate %in% captive1tree$tip.label | isolate == captive1Human,'Mixed-host captive clade',
    ifelse(isolate %in% captive2tree$tip.label | isolate == captive2Human,'captive gorilla clade 2',
      ifelse(isolate %in% captive3tree$tip.label | isolate == captive3Human,'captive gorilla clade 1','unassigned'))))

```
#### Genome size and predicted genes by host species
```{r}
Predicted_genes <- Bxy_metadata %>% 
  filter(clade !='unassigned') %>% 
  unite(clade_host_site, clade, host_site, sep = "_", remove = FALSE)  %>%
  ggplot(aes(x=clade,y=X..predicted.genes)) + 
  geom_dotplot(binaxis = "y",
               stackdir = "center",stackratio=.75, dotsize=.5,
               aes(fill = host_site)) +
  theme_cowplot() + ylab('# of predicted genes') +
   scale_fill_manual(values=get_color_palette(Bxy_metadata$isolate)) +
  theme(axis.text.x = element_text(angle = 90)) 
ggsave(Predicted_genes,
       file=file.path(dir,'analyses/Fig_predicted_genes.pdf'),width=8)
```


#### Cluster isolate genomes
```{r}
#aln <- read.alignment(file.path(dir,'roary/core_gene_alignment.aln'),format='fasta')
#alnd <- as.matrix(dist.alignment(aln, matrix = "similarity",gap=0))
#colnames(alnd) <- str_replace_all(colnames(alnd),'[_-]','.')
#rownames(alnd) <- str_replace_all(rownames(alnd),'[_-]','.')
#alnd <- alnd[Bxy_metadata$isolate,Bxy_metadata$isolate]
#write.table(alnd,file = file.path(dir,'analyses/core_gene_alignment_dist.txt'),sep='\t')
```

##### Read in dist matrix of core alignment
```{r}
alnd <- read.table(file = file.path(dir,'analyses/core_gene_alignment_dist.txt'),sep='\t')
hc <- hclust(dist(alnd))

cluster99 <- cutree(hc,h=.01)
cluster99 <- as.data.frame(cluster99) %>% rownames_to_column(var='isolate')

cluster999 <- cutree(hc,h=.001)
cluster999 <- as.data.frame(cluster999) %>% rownames_to_column(var='isolate')

Bxy_metadata <- Bxy_metadata %>% left_join(cluster99,by='isolate') %>% left_join(cluster999,by='isolate')

```

#### Write metadata with clade and cluster assignments
```{r}
write.table(Bxy_metadata,file=file.path(dir,'analyses/Bxy_metadata.txt'),quote=F,sep='\t')
```


#### Read in presence absence table produced by roary
```{r}
pres_abs <- read.csv(file.path(dir,'roary/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(Bxy_metadata$isolate,colnames(pres_abs)) #any isolates missing?
isblank <- function(x) (ifelse(x=="",0,1))
pres_abs <- pres_abs %>% 
  select(Gene,Annotation,Bxy_metadata$isolate)  %>%
  mutate_at(vars(-Gene,-Annotation),isblank) %>%
  filter(rowSums(across(where(is.numeric)))>0)

roary_gene_table <- pres_abs %>% select(-Annotation)
write.table(roary_gene_table,
            file=file.path(dir,'analyses/roary_gene_table.txt'),
            sep='\t',quote=FALSE,row.names = FALSE)
roary_annotation <- pres_abs %>% select(Gene,Annotation) 
```

#### Read in annotations from eggnog mapper
```{r}
eggnog <- readr::read_tsv(
  file.path(dir,'eggnog_mapper/pan_genome_reference.emapper.annotations'),comment = '##')
eggnog <- eggnog %>% 
  rename('Gene.ID'='#query') %>% 
  select(Gene.ID,best_OG_cat,best_OG_name,best_OG_desc)

pan_genome_fasta <- read.fasta(file.path(dir,'roary/pan_genome_reference.fa'),whole.header = T)
full_sequence_names <- names(pan_genome_fasta) #link between Gene and Gene.ID
head(full_sequence_names) 
full_sequence_names <- as.data.frame(full_sequence_names) %>%
  separate(full_sequence_names,sep=' ',into=c('Gene.ID','Gene'))

pres_abs_eggnog <- pres_abs %>%
  left_join(full_sequence_names,by='Gene') %>%
  left_join(eggnog, by = 'Gene.ID') 
  
head(pres_abs_eggnog)
```


##### Analysis of COG categories between core and shell genomes
```{r}
BovNames <- Bxy_metadata$isolate[Bxy_metadata$taxonomy_Species!="Bacteroides_xylanisolvens"]
BxyNames <- Bxy_metadata$isolate[Bxy_metadata$taxonomy_Species=="Bacteroides_xylanisolvens"]

Bxy <- pres_abs_eggnog %>% 
  select(-BovNames) %>%  #removes B. ovatus outgroup
  mutate(gene_count = as.numeric(rowSums(across(where(is.numeric))))) %>% 
  filter(gene_count>0) #remove genes not present after excluding Bov

core = length(BxyNames)* .95 #set limits for core, shell, cloud genes 
shell = length(BxyNames)* .15
Bxy <- Bxy %>%  mutate(core_cat = if_else(gene_count>core,'core',
                    if_else(gene_count>shell,'shell','cloud')),
                    best_OG_cat = replace_na(best_OG_cat,'-'),
                    cog1 =str_sub(best_OG_cat, end=1)) %>% as.data.frame()
Bxy$core_cat <-  ordered(Bxy$core_cat,c('core','shell','cloud'))

#average breakdown of gene cat 
Bxy_core <- Bxy %>% select(core_cat,all_of(BxyNames)) %>%
  pivot_longer(all_of(BxyNames), names_to = 'isolate',values_to='n') %>% 
  group_by(core_cat,isolate) %>% 
  summarise(n = sum(n)) %>%
  as.data.frame() 

core_shell_cloud <- ggplot(Bxy_core, aes(x=core_cat,y=n)) + 
  geom_boxplot(aes(fill=core_cat)) + 
  theme_cowplot() + 
  theme(legend.position="none") +
  scale_fill_brewer() +
  ylab('Number of genes')
ggsave(core_shell_cloud,
       file=file.path(dir,'analyses/Fig_core_shell_cloud.pdf'))

Bxy_core_cog <- Bxy %>% 
  select(core_cat,cog1,all_of(BxyNames)) %>%
  pivot_longer(all_of(BxyNames), names_to = 'isolate',values_to='n') %>% 
  group_by(core_cat,cog1,isolate) %>% 
  summarise(n = sum(n)) %>%
  as.data.frame() 

ggplot(Bxy_core_cog, aes(x=cog1,y=n)) + 
  geom_boxplot(aes(fill=cog1)) + theme_bw()

ggplot(Bxy_core_cog, aes(x=cog1,y=n)) + 
  geom_boxplot(aes(fill=core_cat)) + theme_bw()

Bxy_core_cog_prop <- Bxy_core_cog %>% 
  left_join(Bxy_core,by=c('core_cat','isolate'),suffix=c('cog','core')) %>%
  mutate(n.prop = ncog/ncore *100)

core_vs_shell <- Bxy_core_cog_prop %>% 
  filter(cog1 != 'A') %>% 
  group_by(cog1) %>%
  filter(core_cat %in% c('shell','core')) %>%
  rstatix::kruskal_test(n.prop ~ core_cat) %>%
  mutate(comp='core_vs_shell') 
core_vs_shell <- core_vs_shell %>% mutate(padj = p * nrow(core_vs_shell))

ave_cog <- Bxy_core_cog_prop %>% 
  filter(cog1 != 'A', core_cat != 'cloud') %>% 
  group_by(cog1,core_cat) %>%
  summarise(mean_nprop = mean(n.prop)) %>%
  pivot_wider(names_from = core_cat, values_from = mean_nprop) %>%
  mutate(diff = core - shell,
         ratio = diff/core,
         rep = ifelse(diff<0,'shell','core')) 

#Bxy_core_cog_prop %>% group_by(core_cat,isolate) %>% summarise(sum(n.prop)) ##SANITY CHECK
shell_cog_order <- ave_cog %>% 
  arrange(-shell) %>% 
  pull(cog1) %>% 
  unique()

Bxy_core_cog_prop2 <- Bxy_core_cog_prop %>% 
  filter(core_cat != 'cloud' & cog1 != 'A') %>%
  left_join(ave_cog,by='cog1') %>%
  mutate(cog1 = ordered(cog1,shell_cog_order)) 
 

rep_shell <- Bxy_core_cog_prop2 %>% 
  filter(rep == 'shell') %>%
  ggplot(aes(x=cog1,y=n.prop)) + 
  geom_boxplot(aes(fill=core_cat)) + 
  theme_bw() + 
  xlab("COG category") + ylab("proportion")

rep_core <- Bxy_core_cog_prop2 %>% 
  filter(rep == 'core') %>%
  ggplot(aes(x=cog1,y=n.prop)) + 
  geom_boxplot(aes(fill=core_cat)) + 
  theme_bw()  + 
  xlab("COG category") + ylab("proportion")

library(cowplot)
rep <- plot_grid(
  rep_shell + theme(legend.position="none"),
  rep_core + theme(legend.position="none"),
  labels = c("A", "B"),
  rel_widths = c(1,2)
)

legend  <- get_legend(
 rep_shell + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom")
)

# add the legend underneath the row we made earlier. Give it 10%
# of the height of one plot (via rel_heights).
rep = plot_grid(rep, legend, ncol = 1, rel_heights = c(1, .1))
rep
ggsave(rep,
       file=file.path(dir,'analyses/Fig_cog_core_vs_shell.pdf'))

```



##### dbCAN
Use online dbCAN metada server (http://bcb.unl.edu/dbCAN2/index.php) to annotate pangenome reference produced by roary. Split files to meet size requirements. Run against dbCAN, CAZy, PPR.
```{r}
pan_genome_fasta <- read.fasta(file.path(dir,'roary/pan_genome_reference.fa'),whole.header = T)
nseqs = length(pan_genome_fasta)
half= nseqs/2
part1 = pan_genome_fasta[1:half] 
part2 = pan_genome_fasta[(half+1):nseqs] 
system(paste0('mkdir -pv ',file.path(dir,'analyses/annotation')))
seqinr::write.fasta(part1,names=names(part1),
                    file=file.path(dir,'analyses/annotation/pan_genome_reference_part1.fa'))
seqinr::write.fasta(part2,names=names(part2),
                    file=file.path(dir,'analyses/annotation/pan_genome_reference_part2.fa'))
```

Download results from dbCAN meta server
```{bash}
#wget http://bcb.unl.edu/dbCAN2/data/blast/20210319105041/overview.txt -O dbDCAN_part1.txt
#wget http://bcb.unl.edu/dbCAN2/data/blast/20210319111110/overview.txt -O dbDCAN_part2.txt
```

```{r}
#system(paste0('mv dbDCAN_part1.txt ',file.path(dir,'analyses/annotation/dbDCAN_part1.txt')))
#system(paste0('mv dbDCAN_part2.txt ',file.path(dir,'analyses/annotation/dbDCAN_part2.txt')))

#read in DBCan
dbcan_p1 <- read.table(file.path(dir,'analyses/annotation/dbDCAN_part1.txt'),sep='\t',header=T)
dbcan_p2 <- read.table(file.path(dir,'analyses/annotation/dbDCAN_part2.txt'),sep='\t',header=T)
head(dbcan_p1) #dbCAN split fasta sequence header so have to add back the Gene ortholog

dbcan <- dbcan_p1 %>% 
  add_row(dbcan_p2) %>% #add part2 to end of part2
  mutate(Gene.ID=str_sub(Gene.ID, end=-3)) %>% #remove '_1' from end of GeneID
  mutate_all(as.character) 

write.table(dbcan,file.path(dir,'analyses/annotation/dbDCAN_final.txt'),sep='\t',quote=F)


roary_eggnog_dbcan <- pres_abs_eggnog %>% 
  left_join(dbcan,by='Gene.ID') 
roary_eggnog_dbcan_annot <- roary_eggnog_dbcan %>% 
  select(-Bxy_metadata$isolate)
head(roary_eggnog_dbcan_annot)
write.table(roary_eggnog_dbcan_annot,file.path(dir,'analyses/roary_eggnog_dbcan.txt'),sep='\t',quote=F)

```

#### Format data for count
```{r}
system(paste0('rm -r ',dir,'/count'))
system(paste0('mkdir -pv ',dir,'/count'))
Bxy_tree_count <- Bxy_tree
Bxy_tree_count <- makeNodeLabel(Bxy_tree_count)
write.tree(Bxy_tree_count,file=file.path(dir,'count/Bxy_count.tree'))

pres_abs_count <- pres_abs %>% select(-Annotation)
write.table(pres_abs_count,
            file=file.path(dir,'count/roary_table.txt'),
            sep='\t',quote=FALSE,row.names = FALSE)
```

#### Function to run count
```{r}
runCount <- function(tree,data,output,gain) {
  system(paste('java -Xmx2048M -cp bin/Count/Count.jar ca.umontreal.iro.evolution.genecontent.AsymmetricWagner -gain ',
             gain,
             tree,
             data,'>',
             output,sep=' '))
  system(paste0("grep '# CHANGE' ",output," | sed 's/# //' > ",output,".CHANGE"))
  system(paste0("grep '# PRESENT' ",output," | sed 's/# //' > ",output,".PRESENT"))
  system(paste0("grep '# FAMILY' ",output," | sed 's/# //' > ",output,".FAMILY"))
  present <- read.table(paste0(output,".PRESENT"),sep='\t',header=TRUE)
  res <- present %>% 
    mutate(istip = if_else(node %in% metadata$isolate,'tip','node')) %>% 
    rstatix::t_test(genes ~ istip)
  res$gain_penalty <- gain
  return(res)
}

#compare gain penalty 
gain1 <-runCount(tree=file.path(dir,'count/Bxy_count.tree'),
                 data=file.path(dir,'count/roary_table.txt'),
                 output=file.path(dir,'count/roary_table'),
                 gain=1)
gain3 <-runCount(tree=file.path(dir,'count/Bxy_count.tree'),
                 data=file.path(dir,'count/roary_table.txt'),
                 output=file.path(dir,'count/roary_table'),
                 gain=3)
gain2 <-runCount(tree=file.path(dir,'count/Bxy_count.tree'),
                 data=file.path(dir,'count/roary_table.txt'),
                 output=file.path(dir,'count/roary_table'),
                 gain=2)
print(rbind(gain1,gain2,gain3))

```

#### Visualize gene gain/loss on roary output

```{r}
count_output <- read.table(file.path(dir,'count/roary_table.CHANGE'),sep='\t',header=TRUE)
count_tree <- read.tree(file=file.path(dir,'count/Bxy_count.tree'))

#read in count output
count_output <- as.data.frame(count_output) %>% 
    select(-CHANGE) %>%
    filter(node != 'total') %>%
    rename(label=node) %>%
    select(label,family_gain,family_loss,gene_gain,gene_loss) %>%
   mutate_at(vars(-label),as.numeric) %>%
   droplevels.data.frame()
  
#add count gene gain loss info to tree_object
tree_dat = count_tree %>% 
    treeio::as_tibble() %>% 
    left_join(count_output,by='label') %>%
    left_join(Bxy_metadata,by=c('label'='isolate')) %>%
    as.treedata()

extract_subtree_treeio <- function(tree_S4,taxa) {
  MCRA <- ape::getMRCA(as.phylo(tree_S4),taxa)
  subtree <- extract.clade(as.phylo(tree_S4),MCRA)
  to_drop <- setdiff(as.phylo(tree_S4)$tip.label,subtree$tip.label)
  tree_S4_subtree  <- treeio::drop.tip(tree_S4,to_drop)
  return(tree_S4_subtree)
}

get_nodes_to_include <- function(tree_dat){
 nodes_to_exclude <- c()
 tree_df <- as_tibble(tree_dat)
 for (cluster in unique(Bxy_metadata$cluster99)) {
      isolates <- tree_df %>% filter(cluster99==cluster) %>% pull(label) 
      if (length(isolates) > 1) {
        subtree <- extract_subtree_treeio(tree_dat,isolates)
        subtree_df <- as_tibble(subtree) %>% filter(parent!=node)
        nodes_to_exclude <- c(subtree_df$label,nodes_to_exclude)
        }
  }
  nodes_to_include <- setdiff(as_tibble(tree_dat)$label,nodes_to_exclude)
  return(nodes_to_include)
  }

###Whole tree (exclue B.ovatus) 
Bxy_whole_tree <- extract_subtree_treeio(tree_dat,c('GCA.009102525.1.ASM910252v1','P21.4E'))
nodes_to_include <- get_nodes_to_include(Bxy_whole_tree)
tree_df <- as_tibble(Bxy_whole_tree)
captive1Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P17.D4','P13.A9'))
captive1Node_parent <- tree_df$parent[tree_df$node == captive1Node]
captive2Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P21.4E','P21.11C'))
captive2Node_parent <- tree_df$parent[tree_df$node == captive2Node]
captive3Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P21.6E','P21.2G'))
captive3Node_parent <- tree_df$parent[tree_df$node == captive3Node]
parent_nodes <- c(captive1Node_parent,captive2Node_parent,captive3Node_parent)
(Bxy_whole_tree_plot <- ggtree(Bxy_whole_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree_dat)$tip.label)) + 
      geom_cladelabel(node=captive1Node, color='black', offset=.001,
                  label="Mixed-host captive clade") + 
      geom_cladelabel(node=captive2Node, 
                      color='black', offset=.001,label="Captive gorilla clade2") +
      geom_cladelabel(node=captive3Node, 
                      color='black', offset=.001,label="Captive gorilla clade1") +
    geom_treescale() +
    xlim(NA,.05))
ggsave(Bxy_whole_tree_plot,file=file.path(dir,'count/roary_whole_tree.pdf'),height=20,width=12)

####Clade A
count_cladeA_tree <- extract_subtree_treeio(tree_dat,
                                            c('GCA.009102525.1.ASM910252v1','GCA.009093695.1.ASM909369v1'))
nodes_to_include <- get_nodes_to_include(count_cladeA_tree)
(count_cladeA_tree_plot <- ggtree(count_cladeA_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(count_cladeA_tree)$tip.label)) + 
      geom_treescale()) + ggtitle('Clade A')
ggsave(count_cladeA_tree_plot,file=file.path(dir,'count/roary_cladeA.pdf'),height=10,width=8)


####Clade B
count_cladeB_tree <- extract_subtree_treeio(tree_dat,c('P17.D4','GCA.003464445.1.ASM346444v1'))
nodes_to_include <- get_nodes_to_include(count_cladeB_tree)
captive1Node  <- ape::getMRCA(as.phylo(count_cladeB_tree),c('P17.D4','P13.A9'))
captive1Node_parent <- as_tibble(count_cladeB_tree)$parent[as_tibble(count_cladeB_tree)$node == captive1Node]
captive3Node  <- ape::getMRCA(as.phylo(count_cladeB_tree),c('P21.6E','P21.2G'))
captive3Node_parent <- as_tibble(count_cladeB_tree)$parent[as_tibble(count_cladeB_tree)$node == captive3Node]
parent_nodes <- c(captive1Node_parent,captive3Node_parent)
(count_cladeB_tree_plot <- ggtree(count_cladeB_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree_dat)$tip.label)) + 
      geom_cladelabel(node=captive1Node, color='black', offset=.001,
                  label="Mixed-host captive clade") + 
      geom_cladelabel(node=captive3Node, color='black', offset=.001,label="Captive gorilla clade1") +
    geom_treescale() + ggtitle('Clade B') +
    xlim(NA,.05))
ggsave(count_cladeB_tree_plot,file=file.path(dir,'count/roary_cladeB.pdf'),height=10,width=8)

###Clade C
count_cladeC_tree <- extract_subtree_treeio(tree_dat,c('P21.4E','GCA.009102685.1.ASM910268v1'))
nodes_to_include <- get_nodes_to_include(count_cladeC_tree)
captive2Node  <- ape::getMRCA(as.phylo(count_cladeC_tree),c('P21.4E','P21.11C'))
captive2Node_parent <- as_tibble(count_cladeC_tree)$parent[as_tibble(count_cladeC_tree)$node == captive2Node]
parent_nodes <- c(captive2Node_parent)
(count_cladeC_tree_plot <- ggtree(count_cladeC_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(count_cladeC_tree)$tip.label)) + 
      geom_cladelabel(node=captive2Node, color='black', offset=.001,label="Captive gorilla clade2") +
    geom_treescale() + ggtitle('Clade C') +
   xlim(NA,.05))
ggsave(count_cladeC_tree_plot,file=file.path(dir,'count/roary_cladeC.pdf'),height=10)



plot_gene_gain <- function(tree,title,limit){
  ggtree(tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(gene_gain>limit|gene_loss>limit)),
             size =3, vjust=-.1, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(gene_gain>limit|gene_loss>limit)),
             size=3, vjust=1, color='blue') + 
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree)$tip.label)) + 
    geom_treescale() + ggtitle(title)
}


count_captive1Node  <- extract_subtree_treeio(tree_dat,c('P17.D4','P13.A9'))
(count_captive1Node_plot <- plot_gene_gain(count_captive1Node,'Mixed host captive clade',0))
ggsave(count_captive1Node_plot,file=file.path(dir,'count/roary_captive1.pdf'),height=10)
 
count_captive2Node  <- extract_subtree_treeio(tree_dat,c('P21.4E','P21.11C'))
(count_captive2Node_plot <- plot_gene_gain(count_captive2Node,'captive gorilla clade1',0))
ggsave(count_captive2Node_plot,file=file.path(dir,'count/roary_captive2.pdf'),height=10)
  
count_captive3Node  <- extract_subtree_treeio(tree_dat,c('P21.6E','P21.2G'))
(count_captive3Node_plot <- plot_gene_gain(count_captive3Node,'captive gorilla clade2',0))
ggsave(count_captive3Node_plot,file=file.path(dir,'count/roary_captive3.pdf'),height=10)

```


#### Run count on orthofinder gene table
```{r}
of_tab <- read_tsv(file=
          file.path(dir,'orthofinder/Results_Mar23/Orthogroups/Orthogroups.GeneCount.tsv'))
colnames(of_tab) <- str_replace_all(colnames(of_tab),'[-_]','.')
of_tab <- of_tab %>%
  mutate(sum = rowSums(select(of_tab, -Orthogroup) > 0)) %>% 
  filter(sum>0) %>%
  select(-sum)
write.table(of_tab,file=file.path(dir,'count/orthofinder_table.txt'),
            sep='\t',quote=FALSE,row.names = FALSE)

runCount(tree=file.path(dir,'count/Bxy_count.tree'),
         data=file.path(dir,'count/orthofinder_table.txt'),
         output=file.path(dir,'count/orthofinder_table'),
         gain=2)

count_output <- read.table(file.path(dir,'count/orthofinder_table.CHANGE'),sep='\t',header=TRUE)
count_tree <- read.tree(file=file.path(dir,'count/Bxy_count.tree'))

#read in count output
count_output <- as.data.frame(count_output) %>% 
    select(-CHANGE) %>%
    filter(node != 'total') %>%
    rename(label=node) %>%
    select(label,family_gain,family_loss,gene_gain,gene_loss) %>%
   mutate_at(vars(-label),as.numeric) %>%
   droplevels.data.frame()
  
#add count gene gain loss info to tree_object
tree_dat = count_tree %>% 
    treeio::as_tibble() %>% 
    left_join(count_output,by='label') %>%
    left_join(Bxy_metadata,by=c('label'='isolate')) %>%
    as.treedata()


###Whole tree (exclue B.ovatus) 
Bxy_whole_tree <- extract_subtree_treeio(tree_dat,c('GCA.009102525.1.ASM910252v1','P21.4E'))
nodes_to_include <- get_nodes_to_include(Bxy_whole_tree)
tree_df <- as_tibble(Bxy_whole_tree)
captive1Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P17.D4','P13.A9'))
captive1Node_parent <- tree_df$parent[tree_df$node == captive1Node]
captive2Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P21.4E','P21.11C'))
captive2Node_parent <- tree_df$parent[tree_df$node == captive2Node]
captive3Node  <- ape::getMRCA(as.phylo(Bxy_whole_tree),c('P21.6E','P21.2G'))
captive3Node_parent <- tree_df$parent[tree_df$node == captive3Node]
parent_nodes <- c(captive1Node_parent,captive2Node_parent,captive3Node_parent)
(Bxy_whole_tree_plot <- ggtree(Bxy_whole_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree_dat)$tip.label)) + 
      geom_cladelabel(node=captive1Node, color='black', offset=.001,
                  label="Mixed-host captive clade") + 
      geom_cladelabel(node=captive2Node, 
                      color='black', offset=.001,label="Captive gorilla clade2") +
      geom_cladelabel(node=captive3Node, 
                      color='black', offset=.001,label="Captive gorilla clade1") +
    geom_treescale() +
    xlim(NA,.05))
ggsave(Bxy_whole_tree_plot,file=file.path(dir,'count/orthofinder_whole_tree.pdf'),height=20,width=12)

####Clade A
count_cladeA_tree <- extract_subtree_treeio(tree_dat,
                                            c('GCA.009102525.1.ASM910252v1','GCA.009093695.1.ASM909369v1'))
nodes_to_include <- get_nodes_to_include(count_cladeA_tree)
(count_cladeA_tree_plot <- ggtree(count_cladeA_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(count_cladeA_tree)$tip.label)) + 
      geom_treescale()) + ggtitle('Clade A')
ggsave(count_cladeA_tree_plot,file=file.path(dir,'count/orthofinder_cladeA.pdf'),height=10,width=8)


####Clade B
count_cladeB_tree <- extract_subtree_treeio(tree_dat,c('P17.D4','GCA.003464445.1.ASM346444v1'))
nodes_to_include <- get_nodes_to_include(count_cladeB_tree)
captive1Node  <- ape::getMRCA(as.phylo(count_cladeB_tree),c('P17.D4','P13.A9'))
captive1Node_parent <- as_tibble(count_cladeB_tree)$parent[as_tibble(count_cladeB_tree)$node == captive1Node]
captive3Node  <- ape::getMRCA(as.phylo(count_cladeB_tree),c('P21.6E','P21.2G'))
captive3Node_parent <- as_tibble(count_cladeB_tree)$parent[as_tibble(count_cladeB_tree)$node == captive3Node]
parent_nodes <- c(captive1Node_parent,captive3Node_parent)
(count_cladeB_tree_plot <- ggtree(count_cladeB_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree_dat)$tip.label)) + 
      geom_cladelabel(node=captive1Node, color='black', offset=.001,
                  label="Mixed-host captive clade") + 
      geom_cladelabel(node=captive3Node, color='black', offset=.001,label="Captive gorilla clade1") +
    geom_treescale() + ggtitle('Clade B') +
    xlim(NA,.05))
ggsave(count_cladeB_tree_plot,file=file.path(dir,'count/orthofinder_cladeB.pdf'),height=10,width=8)

###Clade C
count_cladeC_tree <- extract_subtree_treeio(tree_dat,c('P21.4E','GCA.009102685.1.ASM910268v1'))
nodes_to_include <- get_nodes_to_include(count_cladeC_tree)
captive2Node  <- ape::getMRCA(as.phylo(count_cladeC_tree),c('P21.4E','P21.11C'))
captive2Node_parent <- as_tibble(count_cladeC_tree)$parent[as_tibble(count_cladeC_tree)$node == captive2Node]
parent_nodes <- c(captive2Node_parent)
(count_cladeC_tree_plot <- ggtree(count_cladeC_tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(label %in% nodes_to_include)),
             size =3, vjust=-.11, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(label %in% nodes_to_include)),
             size=3, vjust=1.01, color='blue') + 
      geom_point2(aes(subset=(node%in%parent_nodes)), color='black', size=4) + 
      geom_range(range='2', color="red", size = 1, alpha = 0.5)  +
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(count_cladeC_tree)$tip.label)) + 
      geom_cladelabel(node=captive2Node, color='black', offset=.001,label="Captive gorilla clade2") +
    geom_treescale() + ggtitle('Clade C') +
   xlim(NA,.05))
ggsave(count_cladeC_tree_plot,file=file.path(dir,'count/orthofinder_cladeC.pdf'),height=10)



plot_gene_gain <- function(tree,title,limit){
  ggtree(tree)  + 
      geom_text2(aes(x=branch, label=gene_gain,subset=(gene_gain>limit|gene_loss>limit)),
             size =3, vjust=-.1, color='firebrick') +
      geom_text2(aes(x=branch, label=gene_loss,subset=(gene_gain>limit|gene_loss>limit)),
             size=3, vjust=1, color='blue') + 
      geom_tippoint(aes(color=host_site), alpha=0.8) + 
      scale_colour_manual(values=get_color_palette(as.phylo(tree)$tip.label)) + 
    geom_treescale() + ggtitle(title)
}


count_captive1Node  <- extract_subtree_treeio(tree_dat,c('P17.D4','P13.A9'))
(count_captive1Node_plot <- plot_gene_gain(count_captive1Node,'Mixed host captive clade',0))
ggsave(count_captive1Node_plot,file=file.path(dir,'count/orthofinder_captive1.pdf'),height=10)
 
count_captive2Node  <- extract_subtree_treeio(tree_dat,c('P21.4E','P21.11C'))
(count_captive2Node_plot <- plot_gene_gain(count_captive2Node,'captive gorilla clade1',0))
ggsave(count_captive2Node_plot,file=file.path(dir,'count/orthofinder_captive2.pdf'),height=10)
  
count_captive3Node  <- extract_subtree_treeio(tree_dat,c('P21.6E','P21.2G'))
(count_captive3Node_plot <- plot_gene_gain(count_captive3Node,'captive gorilla clade2',0))
ggsave(count_captive3Node_plot,file=file.path(dir,'count/orthofinder_captive3.pdf'),height=10)


```


#### Convergent g

```{r}

nodes_of_interest <- c('Node7','Node12','Node29',
                       'GCA.003468875.1.ASM346887v1','GCA.003436085.1.ASM343608v1','GCA.003458755.1.ASM345875v1',
                       'Node8','Node11','Node28')
family_output_file = file.path(dir,'count/roary_table.FAMILY')
family_output = read_tsv(family_output_file)
family_output <- family_output %>% 
  select("name",nodes_of_interest) %>% 
  rename('captive_clade1' = 'Node7',
         'captive_clade2' = 'Node12',
         'captive_clade3' = 'Node29',
         'human_clade1' = 'GCA.003468875.1.ASM346887v1',
         'human_clade2' = 'GCA.003436085.1.ASM343608v1',
         'human_clade3' = 'GCA.003458755.1.ASM345875v1',
         'ancest_clade1'='Node8',
         'ancest_clade2'='Node11',
         'ancest_clade3'='Node28')

captive_gain <- family_output %>% 
  filter(captive_clade1 > ancest_clade1) %>%
  filter(captive_clade2 > ancest_clade2) %>%
  filter(captive_clade3 > ancest_clade3) %>%
  mutate(converge='captive_gain')

captive_loss <- family_output %>% 
  filter(captive_clade1 < ancest_clade1) %>%
  filter(captive_clade2 < ancest_clade2) %>%
  filter(captive_clade3 < ancest_clade3) %>%
  mutate(converge='captive_loss')

human_gain <- family_output %>% 
  filter(human_clade1 > ancest_clade1) %>%
  filter(human_clade2 > ancest_clade2) %>%
  filter(human_clade3 > ancest_clade3) %>%
  mutate(converge='human_gain')

human_loss <- family_output %>% 
  filter(human_clade1 < ancest_clade1) %>%
  filter(human_clade2 < ancest_clade2) %>%
  filter(human_clade3 < ancest_clade3) %>%
  mutate(converge='human_loss')

captive_gain_roary <- captive_gain %>%
  rename("Gene"=name) %>%
  select(Gene,converge) %>%
  left_join(roary_eggnog_dbcan, by ='Gene')
write_tsv(captive_gain_roary,file=file.path(dir,'count/roary_converge_table.txt'))
```

#### Convergence from orthofinder output
```{r}
family_output_file = file.path(dir,'count/orthofinder_table.FAMILY')
family_output = read_tsv(family_output_file)
family_output <- family_output %>% 
  select("name",nodes_of_interest) %>% 
  rename('captive_clade1' = 'Node7',
         'captive_clade2' = 'Node12',
         'captive_clade3' = 'Node29',
         'human_clade1' = 'GCA.003468875.1.ASM346887v1',
         'human_clade2' = 'GCA.003436085.1.ASM343608v1',
         'human_clade3' = 'GCA.003458755.1.ASM345875v1',
         'ancest_clade1'='Node8',
         'ancest_clade2'='Node11',
         'ancest_clade3'='Node28')

captive_gain <- family_output %>% 
  filter(captive_clade1 > ancest_clade1) %>%
  filter(captive_clade2 > ancest_clade2) %>%
  filter(captive_clade3 > ancest_clade3) %>%
  mutate(converge='captive_gain')

captive_loss <- family_output %>% 
  filter(captive_clade1 < ancest_clade1) %>%
  filter(captive_clade2 < ancest_clade2) %>%
  filter(captive_clade3 < ancest_clade3) %>%
  mutate(converge='captive_loss')

human_gain <- family_output %>% 
  filter(human_clade1 > ancest_clade1) %>%
  filter(human_clade2 > ancest_clade2) %>%
  filter(human_clade3 > ancest_clade3) %>%
  mutate(converge='human_gain')

human_loss <- family_output %>% 
  filter(human_clade1 < ancest_clade1) %>%
  filter(human_clade2 < ancest_clade2) %>%
  filter(human_clade3 < ancest_clade3) %>%
  mutate(converge='human_loss')


#format annotations for orthofinder orthogroups
orthofinder_eggnog <- read_tsv(file.path(dir,'orthofinder/emapper.emapper.annotations'),comment = '##')
orthofinder_eggnog <- orthofinder_eggnog %>% 
  rename('Gene'='#query') %>%
  select(Gene,evalue,best_OG_cat,best_OG_name,best_OG_desc,PFAMs)

orthogroups <- read_tsv(
  file.path(dir,'orthofinder/Results_Mar23/Orthogroups/Orthogroups.tsv'),
  comment = '##')
colnames(orthogroups) <- str_replace_all(colnames(orthogroups),'[_-]','.')
orthogroups <- orthogroups %>%  
  select(Orthogroup,Bxy_metadata$isolate)  %>%
  pivot_longer(cols=Bxy_metadata$isolate,names_to='isolate',values_to='Gene') %>%
  separate(col='Gene',sep='[,]',extra='drop',into = 'Gene') %>% 
  filter(!is.na(Gene)) 

orthonog <- orthogroups %>%
  left_join(orthofinder_eggnog,by='Gene') %>% 
  filter(!is.na(best_OG_cat))

orthonog_summary <- orthonog %>%
  group_by(Orthogroup,best_OG_cat,best_OG_name,best_OG_desc,PFAMs) %>% 
  tally() %>% 
  as.data.frame()

orthogroups_tophit <- orthonog_summary %>%
  arrange(-n) %>%
  group_by(Orthogroup) %>%
  top_n(n=1)

captive_gain_orthofinder <- captive_gain %>%
  rename("Orthogroup"=name) %>%
  select('Orthogroup','converge') %>%
  left_join(orthogroups_tophit,by=c("Orthogroup")) %>% 
  droplevels.data.frame() 
write_tsv(captive_gain_orthofinder,file=file.path(dir,'count/orthofinder_converge_table.txt'))
```


```{r}
captive_gain_both <- captive_gain_roary %>% 
  filter(best_OG_desc %in% captive_gain_orthofinder$best_OG_desc) %>%
  filter(!is.na(best_OG_name)) %>%
  droplevels()
table(captive_gain_both$Annotation)
```

#### Add Roary gene presence absence table to dbCAN data
```{r}

dbcan_isolates <- roary_eggnog_dbcan %>% 
  filter(!is.na(DIAMOND)) %>% 
  mutate(consensus = ifelse(DIAMOND != 'N',DIAMOND, 
                            ifelse(HMMER != 'N',HMMER,Hotpep))) %>%
  separate(consensus,sep='[(_]',into = c('consensus'),extra = 'drop')
  
dbcan_isolates <- dbcan_isolates %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  filter(present == 1) %>%
  group_by(isolate,consensus) %>%
  tally() 

summary_of_CAZY_domains <- dbcan_isolates %>% 
  group_by(consensus) %>% 
  summarise(total = sum(n)) %>% 
  arrange(-total)

CAZY_table <- dbcan_isolates %>% 
  as.data.frame() %>% 
  pivot_wider(names_from=consensus,values_from=n,values_fill=0) %>%
  column_to_rownames(var='isolate') 

HOST_table <- Bxy_metadata %>% 
  select(isolate,host_site) %>%
  column_to_rownames(var='isolate') 


```


```{r}

#'GH2',
cz<- c('GH16','GH82','GH110','GH127')
select_CAZY <- CAZY_table %>% select(all_of(cz))
p <- gheatmap(ggtree(Bxy_tree) + ylim(-10,NA),
              HOST_table,width=.3,colnames_angle=90,hjust=1)  + scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label)) 
library(ggnewscale)
p <- p + new_scale_fill()
p2 <- gheatmap(p,select_CAZY,offset=.1,colnames_angle=90,hjust=1) + scale_fill_viridis_c(option="C")

#add GH2 with new scale bc gene counts are so high
cz<- c('GH2')
select_CAZY <- CAZY_table %>% select(all_of(cz))
p2 <- p2 + new_scale_fill()
p3 <- gheatmap(p2,select_CAZY,offset=.33,colnames_angle=90,hjust=1,width = .25) + scale_fill_viridis_c(option="D")
p3
ggsave(p3,file=file.path(dir,'analyses/cazy_heatmap.pdf'),height=8)
```
```{r}
sulfatase <- roary_eggnog_dbcan %>% 
  filter(Annotation %in% c('N-acetylglucosamine-6-O-sulfatase',
      'N-acetylgalactosamine-6-O-sulfatase',
      'Arylsulfatase','Ulvan-active sulfatase')) %>% droplevels()
table(sulfatase$Annotation)
sulfatase  <- sulfatase  %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  filter(present == 1) %>%
  group_by(isolate,Annotation) %>%
  tally() 

Sulfatase_table <- sulfatase %>% 
  as.data.frame() %>% 
  pivot_wider(names_from=Annotation,values_from=n,values_fill=0) %>%
  column_to_rownames(var='isolate') 

p <- gheatmap(ggtree(Bxy_tree) + ylim(-50,NA),
              HOST_table,width=.3,colnames_angle=90,hjust=1)  + scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label)) 
library(ggnewscale)
p <- p + new_scale_fill()
p2 <- gheatmap(p,Sulfatase_table,offset=.1,colnames_angle=90,hjust=1) + scale_fill_viridis_c(option="B")
p2
ggsave(p2,file=file.path(dir,'analyses/sulfatase_heatmap.pdf'),height=10)

```



```{r}
sulfatase <- roary_eggnog_dbcan %>% filter(Annotation %in% c('Cobalt-zinc-cadmium resistance protein CzcA','Efflux pump membrane transporter BepE',
                                                'Multidrug resistance protein MdtA')) %>% droplevels()
table(sulfatase$Annotation)
sulfatase  <- sulfatase  %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  filter(present == 1) %>%
  group_by(isolate,Annotation) %>%
  tally() 

Sulfatase_table <- sulfatase %>% 
  as.data.frame() %>% 
  pivot_wider(names_from=Annotation,values_from=n,values_fill=0) %>%
  column_to_rownames(var='isolate') 

p <- gheatmap(ggtree(Bxy_tree) + ylim(-50,NA),
              HOST_table,width=.3,colnames_angle=90,hjust=1)  + scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label)) 
library(ggnewscale)
p <- p + new_scale_fill()
p2 <- gheatmap(p,Sulfatase_table,offset=.1,colnames_angle=90,hjust=1) + scale_fill_viridis_c(option="B")
p2
ggsave(p2,file=file.path(dir,'analyses/ARG_heatmap.pdf'),height=10)

```


```{r}
GH <- c('GH2','GH16','GH82','GH110','GH127')

CAZY_BC <- CAZY_table %>% 
  select(GH) %>% #
  rownames_to_column(var='isolate') %>%
  left_join(Bxy_metadata,by='isolate') %>%
  pivot_longer(cols=GH, names_to='GH', values_to='gene_count') %>%
  select(isolate,GH,gene_count,host_site,host,clade,captive_clade) %>% 
  filter(captive_clade != 'unassigned') %>%
  mutate(host_clade = paste0(captive_clade,host)) 
unique(CAZY_BC$host_clade) 
CAZY_BC$host_clade <-  ordered(CAZY_BC$host_clade,c("Mixed-host captive cladehuman",
                                                    "Mixed-host captive cladebonobo",
                                                    "Mixed-host captive cladechimpanzee",
                                                    "Mixed-host captive cladeorangutan",
                                                    "captive gorilla clade 1human",
                                                    "captive gorilla clade 1gorilla",
                                                    "captive gorilla clade 2human",
                                                    "captive gorilla clade 2gorilla"))
cazy_dotplot <- CAZY_BC %>% ggplot(aes(x=host_clade,y=gene_count)) +  
            geom_dotplot(binaxis = "y",
               stackdir = "center",stackratio=.2, dotsize=3,
               aes(fill = host_site,alpha=.3,xShowTitle=FALSE,)) + 
               theme_cowplot() + 
               facet_wrap(~GH,scale = 'free_y',ncol=2)+
               scale_fill_manual(values=get_color_palette(unique(CAZY_BC$isolate))) +
               scale_x_discrete(labels=c("Mixed-host captive cladehuman" = "", 
                                         "Mixed-host captive cladebonobo" = "",
                                         "Mixed-host captive cladechimpanzee" = "Mixed-host clade",
                                         "Mixed-host captive cladeorangutan"="",
                                         "captive gorilla clade 1human"="",
                                         "captive gorilla clade 1gorilla"= "captive gorilla clade 1",
                                         "captive gorilla clade 2human" = "",
                                         "captive gorilla clade 2gorilla"="captive gorilla clade 2")) +
                geom_vline(aes(xintercept=4.75)) +
                geom_vline(aes(xintercept=6.75)) + 
                theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
                scale_y_continuous(breaks=c(0,1,2,3,4,6,8,20,30,40)) +
                xlab('Host clade') + ylab('Gene count')
cazy_dotplot
ggsave(cazy_dotplot,file=file.path(dir,'analyses/cazy_dotplot.pdf'),height=8)


```


```{r}
dir
alnd <- read.table(file = file.path(dir,'analyses/core_gene_alignment_dist.txt'),sep='\t')
aln_distm <- melt_dist(alnd) %>% 
  rename(aln_dist=dist)

library(vegan)
library(harrietr)

roary_pan <- pres_abs %>% select(-Annotation) %>% column_to_rownames(var='Gene') %>% t()
roary_dist <- as.matrix(vegdist(roary_pan, method="bray"))

roary_distm <- melt_dist(roary_dist) %>% 
  rename(roary_dist=dist)

orthofinder_pan <- of_tab  %>% column_to_rownames(var='Orthogroup') %>% t()
orthofinder_dist <- as.matrix(vegdist(orthofinder_pan, method="bray"))
intersect(colnames(roary_dist),colnames(orthofinder_dist))
orthofinder_dist <- orthofinder_dist[colnames(roary_dist),colnames(roary_dist)]
orthofinder_distm <- melt_dist(orthofinder_dist) %>% 
  rename(orthofinder_dist=dist)

tree_dist <- cophenetic.phylo(Bxy_tree)[colnames(alnd),colnames(alnd)]
tree_distm <- melt_dist(tree_dist) %>% 
  rename(tree_dist=dist)

aln_pan_dist <- aln_distm %>% 
  left_join(tree_distm, by=c('iso1','iso2')) %>%
  left_join(roary_distm, by=c('iso1','iso2')) %>%
  left_join(orthofinder_distm, by=c('iso1','iso2')) 


Bxy_subset <- Bxy_metadata %>% select(isolate,host,clade,sample,taxonomy_Species,X..predicted.genes)
aln_pan_dist  <- aln_pan_dist  %>% 
  left_join(Bxy_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and individual2
  left_join(Bxy_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) %>%
  mutate(clade_comp = ifelse(clade.iso1 == clade.iso2,'same_clade','diff_clade'),
         sample_comp = ifelse(sample.iso1 == sample.iso2,'same_sample','diff_sample'),
         host_comp = ifelse(host.iso1 == host.iso2,'same_host','diff_host'))%>%
  filter(taxonomy_Species.iso1=='Bacteroides_xylanisolvens' & taxonomy_Species.iso2=='Bacteroides_xylanisolvens')

aln_pan_dist %>% 
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) + 
  geom_point() +
  theme_bw() 

aln_pan_dist %>% 
  ggplot() + 
  aes(x=tree_dist,y=orthofinder_dist) + 
  geom_point() + 
  theme_bw() 

aln_pan_dist %>% 
  filter(clade_comp == 'same_clade') %>% 
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) + geom_point()
 
aln_pan_dist %>% filter(host_comp == 'diff_host') %>% ggplot(aes(x=tree_dist)) + geom_density()
aln_pan_dist %>% filter(host_comp == 'same_host') %>% ggplot(aes(x=tree_dist)) + geom_density()
roary_same_diff <- aln_pan_dist %>% 
  filter(tree_dist>.01) %>%
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) +
  geom_point(aes(color=host_comp),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'diff_host'),
              method = "lm",color='pink', formula = y ~  x , size = 1, se=TRUE)+
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'same_host'),
              method = "lm",color='blue', formula = y ~  x , size = 1, se=TRUE)
ggsave(roary_same_diff,file=file.path(dir,'analyses/roary_same_diff.pdf'),height=8)
orthofinder_same_diff <- aln_pan_dist %>% 
  filter(tree_dist>.001) %>%
  droplevels() %>%
  ggplot() + 
  aes(x=tree_dist,y=orthofinder_dist) +
  geom_point(aes(color=host_comp),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'diff_host'&tree_dist>.001),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'same_host'&tree_dist>.001),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)
orthofinder_same_diff
ggsave(orthofinder_same_diff,file=file.path(dir,'analyses/orthofinder_same_diff.pdf'),height=8)

aln_pan_dist %>% 
  droplevels() %>%
  ggplot() + 
  aes(x=tree_dist,y=orthofinder_dist) +
  geom_point(aes(color=host_comp),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'diff_host'),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'same_host'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)


sisters <- aln_pan_dist %>% 
  filter(tree_dist<.01) 

sisters  %>%
  ggplot() + 
  aes(x=sample_comp,y=orthofinder_dist) +
  geom_boxplot() +
  theme_bw()

sisters  %>%
  filter(sample_comp != 'same_sample') %>%
  ggplot() + 
  aes(x=host_comp,y=orthofinder_dist,fill=host.iso1) +
  geom_boxplot()

sisters <- sisters %>% mutate(diff_genes = abs(X..predicted.genes.iso1-X..predicted.genes.iso2)) %>%
  mutate(both_host = paste(host.iso1,host.iso2))
sisters  %>%
  filter(sample_comp != 'same_sample') %>%
  ggplot() + 
  aes(x=both_host,y=orthofinder_dist,fill=host.iso1) +
  geom_boxplot()

```

```{r}
Bfr_metadata <- metadata %>% filter(taxonomy_Species == 'Bacteroides_fragilis')

dir = 'results/pangenome/Bacteroides_fragilis'
#aln <- read.alignment(file.path(dir,'roary/core_gene_alignment.aln'),format='fasta')
#alnd <- as.matrix(dist.alignment(aln, matrix = "similarity",gap=0))
#colnames(alnd) <- str_replace_all(colnames(alnd),'[_-]','.')
#rownames(alnd) <- str_replace_all(rownames(alnd),'[_-]','.')
#alnd <- alnd[Bfr_metadata$isolate,Bfr_metadata$isolate]

#write.table(alnd,file = file.path(dir,'roary/core_gene_alignment_dist.txt'),sep='\t')
alnd <- read.table(file = file.path(dir,'roary/core_gene_alignment_dist.txt'))
aln_distm <- melt_dist(alnd) %>% 
  rename(aln_dist=dist)

pres_abs <- read.csv(file.path(dir,'roary/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(Bfr_metadata$isolate,colnames(pres_abs)) #any isolates missing?
isblank <- function(x) (ifelse(x=="",0,1))
pres_abs <- pres_abs %>% 
  select(Gene,Annotation,Bfr_metadata$isolate)  %>%
  mutate_at(vars(-Gene,-Annotation),isblank) %>%
  filter(rowSums(across(where(is.numeric)))>0)
roary_pan <- pres_abs %>% select(-Annotation) %>% column_to_rownames(var='Gene') %>% t()
roary_dist <- as.matrix(vegdist(roary_pan, method="bray"))
roary_distm <- melt_dist(roary_dist) %>% 
  rename(roary_dist=dist)

Bfr_pan_dist <- aln_distm %>% 
  left_join(roary_distm, by=c('iso1','iso2')) 

Bfr_subset <- Bfr_metadata %>% select(isolate,host,sample,taxonomy_Species,X..predicted.genes)
Bfr_pan_dist  <- Bfr_pan_dist  %>% 
  left_join(Bfr_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and individual2
  left_join(Bfr_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2"))  %>%
  select(aln_dist,roary_dist,taxonomy_Species.iso1,taxonomy_Species.iso2)
  
Bfr_Bxy <- Bfr_pan_dist %>% add_row(select(aln_pan_dist,aln_dist,roary_dist,taxonomy_Species.iso1,taxonomy_Species.iso2)) 

Bfr_Bxy %>% 
 # filter(tree_dist>.01) %>%
  droplevels() %>%
  ggplot() + 
  aes(x=aln_dist,y=roary_dist) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(Bfr_Bxy,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bfr_Bxy,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)



pres_abs

```

```{r}
Bov_metadata <- metadata %>% filter(taxonomy_Species == 'Bacteroides_ovatus')

dir = 'results/pangenome/Bacteroides_ovatus'
#aln <- read.alignment(file.path(dir,'roary/core_gene_alignment.aln'),format='fasta')
#alnd <- as.matrix(dist.alignment(aln, matrix = "similarity",gap=0))
#colnames(alnd) <- str_replace_all(colnames(alnd),'[_-]','.')
#rownames(alnd) <- str_replace_all(rownames(alnd),'[_-]','.')
#alnd <- alnd[Bov_metadata $isolate,Bov_metadata$isolate]
#write.table(alnd,file = file.path(dir,'roary/core_gene_alignment_dist.txt'),sep='\t')
alnd <- read.table(file = file.path(dir,'roary/core_gene_alignment_dist.txt'))
aln_distm <- melt_dist(alnd) %>% 
  rename(aln_dist=dist)

pres_abs <- read.csv(file.path(dir,'roary/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(Bov_metadata$isolate,colnames(pres_abs)) #any isolates missing?
isblank <- function(x) (ifelse(x=="",0,1))
pres_abs <- pres_abs %>% 
  select(Gene,Annotation,Bov_metadata$isolate)  %>%
  mutate_at(vars(-Gene,-Annotation),isblank) %>%
  filter(rowSums(across(where(is.numeric)))>0)
roary_pan <- pres_abs %>% select(-Annotation) %>% column_to_rownames(var='Gene') %>% t()
roary_dist <- as.matrix(vegdist(roary_pan, method="bray"))
roary_distm <- melt_dist(roary_dist) %>% 
  rename(roary_dist=dist)

Bov_pan_dist <- aln_distm %>% 
  left_join(roary_distm, by=c('iso1','iso2')) 

Bov_subset <- Bov_metadata %>% select(isolate,host,sample,taxonomy_Species,X..predicted.genes)
Bov_pan_dist  <- Bov_pan_dist  %>% 
  left_join(Bov_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and individual2
  left_join(Bov_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2"))  %>%
  select(aln_dist,roary_dist,taxonomy_Species.iso1,taxonomy_Species.iso2)

dir = 'results/pangenome/Bacteroides_ovatus'
Bac <- Bfr_Bxy %>% add_row(Bov_pan_dist)

Bac %>% 
 # filter(tree_dist>.01) %>%
  droplevels() %>%
  ggplot() + 
  aes(x=aln_dist,y=roary_dist) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)

sisters <- Bac %>% 
  filter(aln_dist<.001) 

sisters  %>%
  ggplot() + 
  aes(x=taxonomy_Species.iso1,y=roary_dist,fill= taxonomy_Species.iso1) +
  geom_boxplot() +
  theme_bw()


```


```{r}
dbcan_annot_df <- roary_eggnog_dbcan %>% 
  mutate(dbcan_consensus = ifelse(is.na(DIAMOND),'nohit',
                            ifelse(DIAMOND != 'N',DIAMOND, 
                            ifelse(HMMER != 'N',HMMER,
                            ifelse(Hotpep != 'N',Hotpep,'nohit')))))  %>%
  separate(dbcan_consensus,sep='[(_]',into = c('dbcan_consensus'),extra = 'drop') 
 
dbcan_annot_df <- dbcan_annot_df %>% filter(Annotation != 'hypothetical protein') 

dbcan_annot_df <- dbcan_annot_df %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') 

dbcan_annot_df <- dbcan_annot_df %>%
  mutate(dbcan_annot = ifelse(dbcan_consensus == 'nohit',
                              as.character(dbcan_annot_df$Annotation),dbcan_consensus)) 


dbcan_annot_df <- dbcan_annot_df %>% select(dbcan_annot,isolate,present) %>%
  group_by(dbcan_annot,isolate) %>%
  summarise(count = sum(present))
Bxy_subset <- Bxy_metadata %>% select(isolate,host,host_site,clade,captive_clade)
dbcan_annot_df <- dbcan_annot_df %>% left_join(Bxy_subset,by='isolate') %>%
  mutate(host_ape_human = ifelse(host=='human','human',
                                 ifelse(host %in% c('chimpanzee','gorilla','bonobo','orangutan'),
                                        'captive_ape','other'
                                        )))


dbcan_annot_summary <- dbcan_annot_df %>% 
  filter(host_ape_human %in% c('human','captive_ape')) %>%
  filter(clade %in% c('cladeB','cladeC')) %>%
  group_by(dbcan_annot,clade,host_ape_human) %>%
  summarize(mean_count = mean(count))
dbcan_annot_summary_wide <- dbcan_annot_summary %>% 
  pivot_wider(names_from = c('host_ape_human','clade'),values_from='mean_count') %>%
  filter(sum(captive_ape_cladeB:human_cladeC)>0) %>%
  filter(captive_ape_cladeB != human_cladeB | captive_ape_cladeC != human_cladeC)

dbcan_annot_kw <- dbcan_annot_df %>% 
  filter(host_ape_human %in% c('human','captive_ape')) %>%
  filter(clade %in% c('cladeB','cladeC')) %>%
  filter(dbcan_annot %in% dbcan_annot_summary_wide$dbcan_annot) %>%
  group_by(dbcan_annot,clade) %>%
  rstatix::kruskal_test(count ~ host_ape_human) 

dbcan_annot_kw_sig <- dbcan_annot_kw %>% filter(p<.001)

dbcan_annot_sig2 <- dbcan_annot_kw_sig %>% group_by(dbcan_annot) %>% tally() %>% as.data.frame() %>% filter(n==2) %>% pull(dbcan_annot)

dbcan_annot_summary_sig <- dbcan_annot_summary_wide %>% 
  filter(dbcan_annot %in% dbcan_annot_sig2 )  

dbcan_annot_summary_sig %>% 
  filter(captive_ape_cladeB > human_cladeB & captive_ape_cladeC > human_cladeC)

converge_up_captive <- dbcan_annot_summary_sig %>% 
  filter(captive_ape_cladeB > human_cladeB & captive_ape_cladeC > human_cladeC)

converge_up_human <- dbcan_annot_summary_sig %>% filter(captive_ape_cladeB < human_cladeB & captive_ape_cladeC < human_cladeC)

```
```{r}
Bfr_metadata <- metadata %>% filter(taxonomy_Species == 'Bacteroides_ovatus')

dir = 'results/pangenome/Bacteroides_ovatus'

pres_abs <- read.csv(file.path(dir,'roary/gene_presence_absence.csv'))

colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(Bfr_metadata$isolate,colnames(pres_abs)) #any isolates missing?
isblank <- function(x) (ifelse(x=="",0,1))
pres_abs <- pres_abs %>% 
  select(Gene,Annotation,Bfr_metadata$isolate)  %>%
  mutate_at(vars(-Gene,-Annotation),isblank) %>%
  filter(rowSums(across(where(is.numeric)))>0)


dbcan_annot_df <- pres_abs %>% filter(Annotation != 'hypothetical protein') 

dbcan_annot_df <- dbcan_annot_df %>% 
  pivot_longer(cols=Bfr_metadata$isolate,
               names_to='isolate',values_to='present') 

dbcan_annot_df <- dbcan_annot_df %>% select(Annotation,isolate,present) %>%
  group_by(Annotation,isolate) %>%
  summarise(count = sum(present))
Bfr_subset <- Bfr_metadata %>% select(isolate,host)
dbcan_annot_df <- dbcan_annot_df %>% left_join(Bfr_subset,by='isolate') %>%
  mutate(host_ape_human = ifelse(host=='human','human',
                                 ifelse(host %in% c('chimpanzee','gorilla','bonobo','orangutan'),
                                        'captive_ape','other'
                                        )))
dbcan_annot_summary <- dbcan_annot_df %>% 
  filter(host_ape_human %in% c('human','captive_ape')) %>%
  group_by(Annotation,host_ape_human) %>%
  summarize(mean_count = mean(count)) %>%
   pivot_wider(names_from = c('host_ape_human'),values_from='mean_count')


table(dbcan_annot_df$Annotation)
dbcan_annot_kw <- dbcan_annot_df %>% 
  filter(host_ape_human %in% c('human','captive_ape')) %>%
  #filter(dbcan_annot %in% dbcan_annot_summary_wide$dbcan_annot) %>%
  group_by(Annotation) %>%
  rstatix::kruskal_test(count ~ host_ape_human) 
```

```{r}
dir = 'results/pangenome/Bacteroides_xylanisolvens_outgroup'
roary <- read.csv(file.path(dir,'roary/clustered_proteins'),sep=':',header=F)
colnames(roary) <- c('roary_og','prot')
roary <- roary %>% 
  mutate(prot = str_remove_all(as.character(prot),' ')) %>%
  separate_rows(prot, sep = '\t',convert = TRUE)
nrow(roary)
roary_nsp <- read.csv(file.path(dir,'roary_nosplitparalogs/clustered_proteins'),sep=':',header=F)
colnames(roary_nsp) <- c('roary_nsp_og','prot')
roary_nsp <- roary_nsp %>% 
  mutate(prot = str_remove_all(as.character(prot),' ')) %>%
  separate_rows(prot, sep = '\t',convert = TRUE) 
nrow(roary_nsp)
orthofinder <- read.csv(file.path(dir,'orthofinder/Results_Mar23/Orthogroups/Orthogroups.txt'),sep=':',header=F)
colnames(orthofinder) <- c('orthofinder','prot')
orthofinder <- orthofinder %>% 
  mutate(prot = as.character(prot)) %>%
  separate_rows(prot, sep = ' ',convert = TRUE) %>%
  filter(prot!="")
nrow(orthofinder)

merged <- roary %>% 
  left_join(roary_nsp, by = 'prot') %>%
  left_join(orthofinder, by = 'prot') 
merged <-merged %>% 
 separate(prot, into = c('prot_name','id'), sep = "_",remove=F) 

pres_abs <- read.csv(file.path(dir,'roary/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(Bxy_metadata$isolate,colnames(pres_abs)) #any isolates missing?

prot_name <- pres_abs %>% 
  select(Bxy_metadata$isolate)  
prot_name <- t(prot_name[1,])
iso_prot<- as.data.frame(prot_name)%>% 
 rownames_to_column(var='isolate') %>%
  rename('prot_name' = '1') %>%
  separate(col = 'prot_name', into='prot_name', sep='[__]')

isblank <- function(x) (ifelse(x=="",0,1))
pres_abs <- pres_abs %>% 
  select(Gene,Annotation,Bxy_metadata$isolate)  %>%
  mutate_at(vars(-Gene,-Annotation),isblank) %>%
  filter(rowSums(across(where(is.numeric)))>0)

merged_filt <- merged %>% 
  filter(roary_og %in% as.character(pres_abs$Gene)) %>%
  filter(prot_name %in% iso_prot$prot_name)

roary_para <- merged_filt %>% select(roary_og,roary_nsp_og) %>% unique()
length(unique(merged_filt$roary_og))

merged_test <- merged_filt %>% group_by(roary_og,orthofinder) %>% tally() %>% as.data.frame()

split_names <- names(table(merged_test$roary_og)[table(merged_test$roary_og)>2])
length(split)
split <- merged_test %>% 
  filter(roary_og %in% split_names)


merged_test <- merged_filt %>% group_by(roary_nsp_og,orthofinder) %>% tally() %>% as.data.frame()
split_names <- names(table(merged_test$roary_nsp_og)[table(merged_test$roary_nsp_og)>1])
length(split_names)
length(unique(merged_test$roary_nsp_og))

dbcan_annot_df <- roary_eggnog_dbcan %>% 
  mutate(dbcan_consensus = ifelse(is.na(DIAMOND),'nohit',
                            ifelse(DIAMOND != 'N',DIAMOND, 
                            ifelse(HMMER != 'N',HMMER,
                            ifelse(Hotpep != 'N',Hotpep,'nohit')))))  %>%
  separate(dbcan_consensus,sep='[(_]',into = c('dbcan_consensus'),extra = 'drop') %>%
  select(-Bxy_metadata$isolate)
head(dbcan_annot_df)
head(merged_filt)
merged_filt2 <- merged_filt %>% 
  left_join(dbcan_annot_df,by=c('roary_og'='Gene'))
GH2_og <- merged_filt2 %>% filter(dbcan_consensus=='GH2')


```



```{r}
nrow(pres_abs)
roary_nsp_gene_table <- pres_abs %>% 
  left_join(roary_para, by = c('Gene'='roary_og')) %>%
  select(roary_nsp_og,Bxy_metadata$isolate) %>%
  group_by(roary_nsp_og) %>% dplyr::summarise_all(sum) %>%
  column_to_rownames(var='roary_nsp_og') %>% t()


roary_dist <- as.matrix(vegdist(roary_nsp_gene_table, method="bray"))

roary_distm <- melt_dist(roary_dist) %>% 
  rename(roary_dist=dist)

orthofinder_pan <- of_tab  %>% column_to_rownames(var='Orthogroup') %>% t()
orthofinder_dist <- as.matrix(vegdist(orthofinder_pan, method="bray"))
intersect(colnames(roary_dist),colnames(orthofinder_dist))
orthofinder_dist <- orthofinder_dist[colnames(roary_dist),colnames(roary_dist)]
orthofinder_distm <- melt_dist(orthofinder_dist) %>% 
  rename(orthofinder_dist=dist)

tree_dist <- cophenetic.phylo(Bxy_tree)[colnames(alnd),colnames(alnd)]
tree_distm <- melt_dist(tree_dist) %>% 
  rename(tree_dist=dist)

aln_pan_dist <- aln_distm %>% 
  left_join(tree_distm, by=c('iso1','iso2')) %>%
  left_join(roary_distm, by=c('iso1','iso2')) %>%
  left_join(orthofinder_distm, by=c('iso1','iso2')) 


Bxy_subset <- Bxy_metadata %>% select(isolate,host,clade,sample,taxonomy_Species,X..predicted.genes)
aln_pan_dist  <- aln_pan_dist  %>% 
  left_join(Bxy_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and individual2
  left_join(Bxy_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) %>%
  mutate(clade_comp = ifelse(clade.iso1 == clade.iso2,'same_clade','diff_clade'),
         sample_comp = ifelse(sample.iso1 == sample.iso2,'same_sample','diff_sample'),
         host_comp = ifelse(host.iso1 == host.iso2,'same_host','diff_host'))%>%
  filter(taxonomy_Species.iso1=='Bacteroides_xylanisolvens' & taxonomy_Species.iso2=='Bacteroides_xylanisolvens')

aln_pan_dist %>% 
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) + 
  geom_point() +
  theme_bw() 

aln_pan_dist %>% 
  ggplot() + 
  aes(x=tree_dist,y=orthofinder_dist) + 
  geom_point() + 
  theme_bw() 

aln_pan_dist %>% 
  filter(clade_comp == 'same_clade') %>% 
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) + geom_point()

c <- .01
roary_same_diff_host <- aln_pan_dist %>% 
  filter(tree_dist>c) %>%
  filter(host.iso1 == 'human'|host.iso2 == 'human') %>%
  ggplot() + 
  aes(x=tree_dist,y=roary_dist) +
  geom_point(aes(color=host_comp),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'diff_host'&tree_dist>c),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'same_host'&tree_dist>c),
              method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE) 
ggsave(roary_same_diff_host,file=file.path(dir,'analyses/roary_same_diff_host.pdf'),height=8)
orthofinder_same_diff_host <- aln_pan_dist %>% 
  filter(tree_dist>c) %>%
  filter(host.iso1 == 'human'|host.iso2 == 'human') %>%
  droplevels() %>%
  ggplot() + 
  aes(x=tree_dist,y=orthofinder_dist) +
  geom_point(aes(color=host_comp),alpha=.2,size=2) + 
  theme_bw() +
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'diff_host'&tree_dist>c),
              method = "lm",color='pink', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(aln_pan_dist,host_comp == 'same_host'&tree_dist>c),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)
ggsave(orthofinder_same_diff_host,file=file.path(dir,'analyses/orthofinder_same_diff_host.pdf'),height=8)

test <- aln_pan_dist %>% filter(tree_dist > .025)

lm_robust(roary_dist ~ host_comp + tree_dist +  host_comp *tree_dist + X..predicted.genes.iso1 + X..predicted.genes.iso2, data = test)

lm_robust(orthofinder_dist ~ host_comp + tree_dist + X..predicted.genes.iso1 + X..predicted.genes.iso2, data = test)
```

#16S results
```{r}
library(phyloseq)
library(dada2)
physeq <- readRDS('16S/phyloseq_rare10000_local.rds')
Bacteroides <- subset_taxa(physeq, Genus == 'Bacteroides')

ASVfasta <- readDNAStringSet('16S/ASVs.fasta')
Bacteroides_fasta <- ASVfasta[names(ASVfasta) %in% taxa_names(Bacteroides)]
writeXStringSet(Bacteroides_fasta,file='16S/Bacteroides_ASVs.fasta')
```
```{bash}
#blastn -db nr -query 16S/Bacteroides_ASVs.fasta -out 16S/Bacteroides_ASV_blastout.txt -outfmt "6 qseqid sseqid salltitles pident evalue" -remote 
```



```{r warning=FALSE}
blast_tab <- read_tsv(file = '16S/Bacteroides_ASV_blastout.txt',comment='#',col_names = FALSE)
colnames(blast_tab) <- c('qseqid', 'sseqid','salltitles', 'pident', 'evalue')

blast_tab2 <- blast_tab %>% 
  filter(pident == 100) %>% 
  separate(col = 'salltitles',into=c('Genus','Species'),extra='drop',sep=' ') %>%
  filter(Genus == 'Bacteroides') %>%
  group_by(qseqid,Species) %>% tally() %>% as.data.frame() 

Bxy_ASVs <- blast_tab2$qseqid[blast_tab2$Species == 'xylanisolvens']
Bxy_Sp <- blast_tab2 %>% filter(qseqid %in% Bxy_ASVs) %>% filter(Species != 'sp.') %>% pull(Species)
Bxy_Sp_ASVs <-blast_tab2 %>% filter(Species %in% Bxy_Sp) %>% pull(qseqid)
Bxy_Sp_ASVs 
tt <- as.data.frame(tax_table(physeq))
tt_edit <- tt %>% rownames_to_column(var='ASV') %>%
                    mutate(
                    Genus = ifelse(is.na(as.character(Genus)),'unassigned',as.character(Genus)),
                    bxy = if_else(rownames(tt) %in% Bxy_Sp_ASVs,1,0),
                    bac = if_else(Genus=='Bacteroides',1,0),
                    bac_bxy = paste(bac,bxy,sep='.')) %>% 
                    column_to_rownames(var='ASV') 
tax_table(physeq) <- as.matrix(tt_edit)
physeq_melt <- psmelt(physeq) 

unique(physeq_melt$Old_Sample)
physeq_melt_subset <- physeq_melt  %>% 
  mutate(Old_SampleID = str_replace(Old_SampleID,pattern = '[.]',replace='_'))  %>%
  filter(Old_SampleID %in% Bxy_metadata$sample)

physeq_melt_subset %>%
  ggplot(aes(x=Sample,y=Abundance,fill=bac_bxy)) + 
  geom_bar(position="stack", stat="identity") +
  scale_fill_brewer(palette="Set3") +
  theme_bw() +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x = element_text(angle = 90,vjust =.4,hjust=1)) +
  ylab("Relative abundance") 

physeq_melt %>%
  filter(captivity_status=='captive') %>%
  group_by(Sample,bac_bxy) %>%
  summarize(Abundance = sum(Abundance)) %>%
  ggplot(aes(x=Sample,y=Abundance,fill=bac_bxy)) + 
  geom_bar(position="stack", stat="identity") +
  scale_fill_brewer(palette="Set3") +
  theme_bw() +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x = element_text(angle = 90,vjust =.4,hjust=1)) +
  ylab("Relative abundance") 

summary <- physeq_melt %>%
  filter(bac_bxy!='0.0') %>%
  mutate(bac_bxy = recode(bac_bxy, '1.1'='B.xylan','1.0'='Other_Bacteroides'))%>%
  group_by(Sample,captivity_status,bac_bxy) %>%
  summarize(Abundance = sum(Abundance)/10000) %>%
  filter(Abundance > 0) %>%
  group_by(captivity_status,bac_bxy) %>% tally()

physeq_melt %>%
  filter(bac_bxy!='0.0') %>%
  mutate(bac_bxy = recode(bac_bxy, '1.1'='B.xylan','1.0'='Other_Bacteroides'))%>%
  group_by(Sample,captivity_status,bac_bxy) %>%
  summarize(log_Relative_Abundance = log(sum(Abundance)/10000)) %>%
  ggplot(aes(x=captivity_status,y=log_Relative_Abundance,fill=bac_bxy)) + 
  geom_boxplot()

physeq_melt %>%
  filter(bac_bxy!='0.0') %>%
  mutate(bac_bxy = recode(bac_bxy, '1.1'='B.xylan','1.0'='Other_Bacteroides'))%>%
  group_by(Sample,captivity_status,bac_bxy) %>%
  summarize(Relative_Abundance = sum(Abundance)/10000) %>%
  ggplot(aes(x=captivity_status,y=Relative_Abundance,fill=bac_bxy)) + 
  geom_boxplot()
```

