---
title: "Bxy2"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir ='/stor/work/Ochman/alex/captive_ape_strain_genomes')
library(ape)
library(treeio)
library(ggtree)
library(estimatr)
library(tidyverse)
library(seqinr)
library(stringr)
library(cowplot)
library(harrietr)
library(vegan)
```


### Label major clades and strain lineages isolated from captive apes in metadata 
```{r}
dir='results/pangenome/Bacteroides_xylanisolvens'
system('mkdir results/pangenome/Bacteroides_xylanisolvens/output')
species = 'Bacteroides_xylanisolvens'

#read-in metadata
Bxy_metadata_file = file.path(dir,"metadata.txt")
Bxy_metadata = read_tsv(Bxy_metadata_file,col_types = cols())
Bxy_metadata <- Bxy_metadata %>%
  mutate(taxonomy_Species = str_replace_all(taxonomy_Species,' ','_'),
         isolate = str_replace_all(isolate,'[_-]','.'),
         site = recode(site,'USA: Cambridge'='USA','USA:Boston'='USA',
                       'China: Shenzhen'='China','USA:Seattle'='USA',     
                       'USA:Baltimore'='USA', 'not applicable'='siteUnknown',
                       'not determined'='siteUnknown','Norway:Oslo'='Norway',
                       'missing'='siteUnknown')) %>%
  unite(host_site, host, site, sep = "_", remove = FALSE) 
#recode host site info
table(Bxy_metadata$host_site)
#look up and recode metadata missing from select samples
#Bxy_metadata$isolate[Bxy_metadata$host_site=='missing_siteUnknown']
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002959635.1.ASM295963v1"] <- 'human_USA'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.900114865.1.IMG.taxon.2654588180.annotated.assembly"] <- 'rumen_USA'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.900107825.1.IMG.taxon.2623620516.annotated.assembly"] <- 'rumen_USA'
#Bxy_metadata$isolate[Bxy_metadata$host=='chicken']
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002161135.1.ASM216113v1"] <- 'chicken_Europe'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002161115.1.ASM216111v1"] <- 'chicken_Europe'
Bxy_metadata <- Bxy_metadata %>%
  mutate(host_site = recode(host_site,'human_Norway'='human',
                            'human_USA'='human',
                            'human_siteUnknown'='human'))
table(Bxy_metadata$host_site)

#root tree
outgroup_names <- Bxy_metadata %>% 
  filter(taxonomy_Species!=species) %>% 
  pull(isolate)
Bxy_tree_file = file.path(dir,paste0("phylogeny/RAxML_bipartitions.",species))
Bxy_tree <- ape::read.tree(Bxy_tree_file)
Bxy_tree$tip.label = str_replace_all(Bxy_tree$tip.label,'[_-]','.') #clean up names
Bxy_tree <- root(Bxy_tree,outgroup=outgroup_names) 
write.tree(Bxy_tree,file = file.path(dir,'output',paste0(species,'.tre')))
tree_plot <- ggtree(Bxy_tree) %<+% Bxy_metadata +
   geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
   geom_tippoint(aes(color=host_site), alpha=0.8)  + 
   geom_tiplab() + 
   xlim(NA,.5)

Bxy_tree <- drop.tip(Bxy_tree,outgroup_names)
#setdiff(Bxy_tree$tip.label,Bxy_metadata$isolate)
#setdiff(Bxy_metadata$isolate,Bxy_tree$tip.label)

get_color_palette <- function(tips) {
  Bxy_metadata <- Bxy_metadata %>% filter(isolate %in% tips)
  vec <- sort(unique(Bxy_metadata$host_site))
  return(recode(vec,
                          'human'='cadetblue4',
                          'human_China' = 'deepskyblue',
                          'rumen_USA' = 'brown4',
                          'bonobo_Columbus_Zoo'='red2',
                          'chimpanzee_Houston_Zoo'='orange2',
                          'orangutan_Houston_Zoo'='purple4',
                          'gorilla_Columbus_Zoo'='green3',
                          'chicken_Europe'='tan',
                          'missing_siteUnknown'='brown4'))}
host_site_color <- get_color_palette(Bxy_tree$tip.label)
Bxy_tree_plot <- ggtree(Bxy_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 90),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  geom_tiplab() + 
  xlim(NA,.5) +
  geom_treescale(x=.15,y=.25)
ggsave(Bxy_tree_plot,file=file.path(dir,'output/Bxy.pdf'),height=20)

cladeA_MCRA <- ape::getMRCA(Bxy_tree,
                            c('P17.D4','GCA.003474245.1.ASM347424v1'))
cladeA_tree <- extract.clade(Bxy_tree,cladeA_MCRA)
cladeA_tree_plot <- ggtree(cladeA_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeA_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

cladeB_MCRA <- ape::getMRCA(Bxy_tree,
                            c('P21.4E','GCA.009102685.1.ASM910268v1'))
cladeB_tree <- extract.clade(Bxy_tree,cladeB_MCRA)
cladeB_tree_plot <- ggtree(cladeB_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeB_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

cladeC_MCRA <- ape::getMRCA(Bxy_tree,
                            c('GCA.009102525.1.ASM910252v1','GCA.902374095.1.MGYG.HGUT.01345'))
cladeC_tree <- extract.clade(Bxy_tree,cladeC_MCRA)
cladeC_tree_plot <- ggtree(cladeC_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeC_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

Bxy_metadata <- Bxy_metadata %>% mutate(clade = ifelse(
  isolate %in% cladeA_tree$tip.label,'cladeA',
    ifelse(isolate %in% cladeB_tree$tip.label,'cladeB',
      ifelse(isolate %in% cladeC_tree$tip.label,'cladeC','unassigned'))))

mixedhostNode  <- ape::getMRCA(Bxy_tree,c('P14.E4','P15.E10'))
mixedhostHuman <- c("GCA.003468875.1.ASM346887v1")
mixedhosttree <- extract.clade(Bxy_tree,mixedhostNode)
gorilla1Node  <- ape::getMRCA(Bxy_tree,c('P21.6D','P21.2G'))
gorilla1tree <- extract.clade(Bxy_tree,gorilla1Node)
gorilla1Human <- c("GCA.003458755.1.ASM345875v1")
gorilla2Node  <- ape::getMRCA(Bxy_tree,c('P21.4G','P21.11A'))
gorilla2tree <- extract.clade(Bxy_tree,gorilla2Node)
gorilla2Human <- c("GCA.003458755.1.ASM345875v1")
Bxy_metadata <- Bxy_metadata %>% mutate(captive_clade = ifelse(
  isolate %in% mixedhosttree$tip.label,'mixedhost',
    ifelse(isolate %in% gorilla1tree$tip.label,'gorilla1',
      ifelse(isolate %in% gorilla2tree$tip.label,'gorilla2','unassigned'))))
Bxy_metadata <- Bxy_metadata %>% 
  mutate(human_ape = 
  if_else(host %in% c('chimpanzee','bonobo','gorilla','orangutan'),
                                                  'ape',host))
write_tsv(Bxy_metadata,file=file.path(dir,'metadata_edit.txt'))

(Bxy_tree_plot_cladelabels <- ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  xlim(NA,.5)+ 
  geom_cladelabel(node=mixedhostNode, color='black', offset=.01,
                  label="Mixed-host clade") + 
  geom_cladelabel(node=gorilla1Node, color='black', offset=.01,label="Gorilla clade1") +
  geom_cladelabel(node=gorilla2Node, color='black', offset=.01,label="Gorilla clade2") +
  geom_cladelabel(node=cladeA_MCRA, color='black', offset=.14,label="CladeA") + 
  geom_cladelabel(node=cladeB_MCRA, color='black', offset=.13,label="CladeB") +
  geom_cladelabel(node=cladeC_MCRA, color='black', offset=.02,label="CladeC") +
  xlim(NA,.5) +
  geom_treescale(x=.05,y=.05))
ggsave(Bxy_tree_plot_cladelabels,file=file.path(dir,'output/Bxy_cladelabels.pdf'),height=20)
```

#### Genome size heatmap
```{r}
(Bxy_tree_plot_cladelabels <- ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site))  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  geom_cladelabel(node=mixedhostNode, color='black', offset=.005,
                  label="Mixed-host clade") + 
  geom_cladelabel(node=gorilla1Node, color='black', offset=.005,label="Gorilla clade1") +
  geom_cladelabel(node=gorilla2Node, color='black', offset=.005,label="Gorilla clade2") +
  geom_cladelabel(node=cladeA_MCRA, color='black', offset=.08,label="CladeA") + 
  geom_cladelabel(node=cladeB_MCRA, color='black', offset=.095,label="CladeB") +
  geom_cladelabel(node=cladeC_MCRA, color='black', offset=.081,label="CladeC") +
  geom_treescale(x=.05,y=.05))

predicted.genes = Bxy_metadata %>% select(isolate,X..predicted.genes) %>%
  dplyr::rename('genes'='X..predicted.genes') %>%
  mutate(genes=as.numeric(genes)) %>%
  column_to_rownames(var='isolate')
(Bxy_tree_genes <- gheatmap(Bxy_tree_plot_cladelabels + ylim(-10,NA) + xlim(NA,.17),
              predicted.genes,width=.4,colnames_angle=90,offset=.04,hjust=1) + 
  scale_fill_viridis_c(option="D"))
ggsave(Bxy_tree_genes,file=file.path(dir,'output/Bxy_genesNum.pdf'),height=11,width=8)
```

### Kruskal 
```{r}
dir = 'results/pangenome/Bacteroides_xylanisolvens'
annotation = read_tsv(file= file.path(dir,'output/roary_nosplitparalogs_annotation.txt'),
                      col_types = cols())
gene_table = read_tsv(file= file.path(dir,'output/roary_nosplitparalogs_gene_table.txt'),
                      col_types = cols())

func <- annotation %>% 
  left_join(gene_table,by='Gene') %>%
  filter(func_consensus!='hypothetical protein',!is.na(func_consensus)) %>%
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  select(func_consensus,isolate,present) %>%
  group_by(func_consensus,isolate) %>%
  summarise(count = sum(present))

func <- Bxy_metadata %>% 
  select(isolate,host,human_ape,host_site,clade,captive_clade) %>% 
  left_join(func,by='isolate',fill=0)

func_mean <- func %>% 
  group_by(func_consensus,clade,human_ape) %>%
  summarize(mean_count = mean(count)) %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) 

func_mean_wide <- func_mean %>%
  filter(human_ape=='human'|human_ape=='ape') %>%
  filter(clade!='unassigned') %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=T) %>%  
  pivot_wider(names_from = host_clade,values_from=mean_count) %>%
  filter(cladeA_ape != cladeA_human,cladeB_ape != cladeB_human) #some difference to observe

func_humanape <- func %>%
  filter(human_ape %in% c('human','ape'),clade!='unassigned')  %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) %>%
  filter(func_consensus %in% func_mean_wide$func_consensus)

kw = func_humanape %>%
  filter(clade != 'cladeC') %>%
  group_by(func_consensus,clade) %>%
  rstatix::kruskal_test(count ~ host_clade) 
kw$p_adj = p.adjust(kw$p, method = 'fdr')

kw_mean <- kw %>% 
  left_join(func_mean_wide,by=c('func_consensus')) %>%
  mutate(ape_mean = ifelse(clade == 'cladeA', cladeA_ape, cladeB_ape))%>%
  mutate(human_mean = ifelse(clade == 'cladeA', cladeA_human, cladeB_human)) %>%
  select(-c(cladeA_ape, cladeB_ape,cladeA_human, cladeB_human, cladeC_human)) %>%
  select(clade, ape_mean, human_mean, everything())
write_tsv(kw_mean,file=file.path(dir,'output/table_kw.txt'))

kw_convergent_captive = kw_mean %>% filter(p_adj < .001)  %>% 
  filter(ape_mean>human_mean) %>%
  add_count(func_consensus,name='count') %>%
  filter(count == 2) 
kw_convergent_captive  = kw_convergent_captive  %>% mutate(diff = ape_mean-human_mean)
kw_convergent_human = kw_mean %>% filter(p_adj < .001)  %>% 
  filter(ape_mean<human_mean) %>%
  add_count(func_consensus,name='count') %>%
  filter(count == 2) 


#GH2,GH20,GH16 have multiple functions to be sure mucin chain metabolism is up
#sanity check to look at function gain by EC numbers show to degrade mucin
EC <- annotation %>% 
  left_join(gene_table,by='Gene') %>%
  filter(EC!='-',!is.na(func_consensus)) %>%
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  select(EC,isolate,present) %>%
  group_by(EC,isolate) %>%
  summarise(count = sum(present)) %>%
  filter(str_detect(EC,'3.2.1.23')|str_detect(EC,'3.2.1.52'))

EC <- Bxy_metadata %>% 
  select(isolate,host,human_ape,host_site,clade,captive_clade) %>% 
  left_join(EC,by='isolate',fill=0)

EC_mean <- EC %>% 
  group_by(EC,clade,human_ape) %>%
  summarize(mean_count = mean(count)) %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) 

EC_mean_wide <- EC_mean %>%
  filter(human_ape=='human'|human_ape=='ape') %>%
  filter(clade!='unassigned') %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=T) %>%  
  pivot_wider(names_from = host_clade,values_from=mean_count) %>%
  filter(cladeA_ape != cladeA_human,cladeB_ape != cladeB_human) #some difference to observe

EC_humanape <- EC %>%
  filter(human_ape %in% c('human','ape'),clade!='unassigned')  %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) %>%
  filter(EC %in% EC_mean_wide$EC)

EC_kw = EC_humanape %>%
  filter(clade != 'cladeC') %>%
  group_by(EC,clade) %>%
  rstatix::kruskal_test(count ~ host_clade) 
EC_kw$p_adj = p.adjust(EC_kw$p, method = 'fdr')

EC_kw_mean <- EC_kw %>% 
  left_join(EC_mean_wide,by=c('EC')) %>%
  mutate(ape_mean = ifelse(clade == 'cladeA', cladeA_ape, cladeB_ape))%>%
  mutate(human_mean = ifelse(clade == 'cladeA', cladeA_human, cladeB_human)) %>%
  select(-c(cladeA_ape, cladeB_ape,cladeA_human, cladeB_human, cladeC_human)) %>%
  select(clade, ape_mean, human_mean, everything())

EC_kw_convergent_captive = EC_kw_mean %>% filter(p_adj < .001)  %>% 
  filter(ape_mean>human_mean) %>%
  add_count(EC,name='count') %>%
  filter(count == 2) 
```



```{r}
sulfatase_func <- kw_convergent_captive %>% 
  filter(str_detect(func_consensus,'sulfatase')) %>%
  pull(func_consensus) %>% unique()

get_func_copynumber_table <- function(list_of_func){
  func_copynumber_table  <- annotation %>% 
    left_join(gene_table,by='Gene') %>%
    filter(func_consensus %in% list_of_func) %>%
    pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
    select(func_consensus,isolate,present) %>%
    group_by(func_consensus,isolate) %>%
    summarise(count = sum(present)) %>%
    pivot_wider(names_from = func_consensus, values_from = count) %>% 
    column_to_rownames(var='isolate')
    func_copynumber_table[func_copynumber_table==0] <- NA
  
  return(func_copynumber_table)
}  


sulfa_mucin <- get_func_copynumber_table(c(sulfatase_func,
                                          'GH2','GH16','GH20',
                                          'GH110','Sialate O-acetylesterase'))
sulfa_mucin <- sulfa_mucin[,c(sulfatase_func,
                              'Sialate O-acetylesterase','GH110',
                             'GH16','GH20', 'GH2')]
carrageenen <- get_func_copynumber_table(c('GH150','GH82','GH167'))
carrageenen <- carrageenen[,c('GH150','GH82','GH167')]

HOST_table <- Bxy_metadata %>% 
  select(isolate,host_site) %>%
  column_to_rownames(var='isolate') 

Bxy_tree = read.tree(file = file.path(dir,'output/Bacteroides_xylanisolvens.tre'))
outgroup_names <- Bxy_metadata %>% 
  filter(taxonomy_Species!=species) %>% 
  pull(isolate)
Bxy_tree <- drop.tip(Bxy_tree,outgroup_names)

library(ggnewscale)
library(RColorBrewer)
tree_v1 = ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 90),size=.75) +
  geom_treescale()

p <- gheatmap(tree_v1 + ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1)  + 
  scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label))
p <- p + new_scale_fill()
(p2 <- gheatmap(p,sulfa_mucin,colnames_angle=90,hjust=1,offset=.01,width=.9) + scale_fill_viridis_c(direction = -1, option="B"))
p2 <- p2 + new_scale_fill()
(p3 <- gheatmap(p2,carrageenen,colnames_angle=90,hjust=1,offset=.07,width=.3)+ scale_fill_viridis_c(option="D",direction = -1))
ggsave(p3,file = file.path(dir,'output/Figure1.pdf'),width = 10, height=10)

(tree_v2 = tree_v1 +
   geom_cladelabel(node=cladeA_MCRA,label="cladeA",offset = .005)+
   geom_cladelabel(node=cladeB_MCRA,label="cladeB",offset = .005)+
   geom_cladelabel(node=cladeC_MCRA,label="cladeC",offset = .005)+
   geom_cladelabel(node=mixedhostNode,label="mixedhost",offset = .01) +
   geom_cladelabel(node=gorilla1Node,label="gorilla1",offset = .01)+ 
   geom_cladelabel(node=gorilla2Node,label="gorilla2",offset = .01) +
   geom_tippoint(aes(subset=(label=='GCA.000210075.1.ASM21007v1')),color='red') +
   xlim(NA,.25)) 
p <- gheatmap(tree_v2 + ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1,offset=.04)  + 
  scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label))
(p <- p + new_scale_fill())
(p2 <- gheatmap(p,sulfa_mucin,colnames_angle=90,hjust=1,offset=.05,width=.9) + scale_fill_viridis_c(direction = -1, option="B"))
p2 <- p2 + new_scale_fill()
(p3 <- gheatmap(p2,carrageenen,colnames_angle=90,hjust=1,offset=.11,width=.3)+ scale_fill_viridis_c(option="D",direction = -1))
ggsave(p3,file = file.path(dir,'output/Figure1_v2.pdf'),width = 10, height=10)


p <- gheatmap(tree_v2 + 
                ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1,offset=.05)  + 
  scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label))
p <- p + new_scale_fill()
(p1b <- gheatmap(p,predicted.genes,colnames_angle=90,hjust=1,offset=.06,width=.1) + scale_fill_viridis_c(direction = -1, option="D"))
p1b <- p1b + new_scale_fill()
(p2 <- gheatmap(p1b,sulfa_mucin,colnames_angle=90,hjust=1,offset=.07,width=.9) + scale_fill_viridis_c(direction = -1, option="B"))
p2 <- p2 + new_scale_fill()
(p3 <- gheatmap(p2,carrageenen,colnames_angle=90,hjust=1,offset=.13,width=.3)+ scale_fill_viridis_c(option="D",direction = -1))
ggsave(p3,file = file.path(dir,'output/Figure1_v3_long.pdf'),width = 10, height=12)
ggsave(p3,file = file.path(dir,'output/Figure1_v3_short.pdf'),width = 10, height=6)
```


### GH2,GH16,GH20 have multiple functions
```{r}
sulfatase_genes <- annotation %>% filter(sulfatase_family=='S1') %>% pull(Gene)
get_gene_table <- function(list_of_genes) {
  gene_table <- annotation %>% 
    left_join(gene_table,by='Gene') %>%
    filter(Gene %in% list_of_genes) %>%
    select(Gene,Bxy_metadata$isolate) %>%
    column_to_rownames(var='Gene') %>%
    t() %>% as.data.frame()
  return(gene_table)
}
sulfatase_gene_table <- get_gene_table(sulfatase_genes)

get_function <- function(gene,column) {
  m <- annotation %>% column_to_rownames('Gene') %>% as.matrix()
  res <- m[gene,column]
  return(res)} 
get_function('group_1490','Annotation')

code_gene_table_by_func <- function(gene_table,column) {
  isolates <- rownames(gene_table)
  gene_table <- gene_table %>% 
      transmute_all(funs(ifelse(. > 0, deparse(substitute(.)), NA))) 
  for (gene in colnames(gene_table)) {
      gene_table[gene_table==gene] <- get_function(gene,column) 
      }
  rownames(gene_table) <- isolates
  return(gene_table)
  }
sulfatase_gene_table_funct <- code_gene_table_by_func(sulfatase_gene_table,'func_consensus')

library(ggnewscale)
library(RColorBrewer)
p <- gheatmap(ggtree(Bxy_tree) + ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1)  + 
  scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label))
p <- p + new_scale_fill()

(sulfatase_heatmap_full <- gheatmap(p,sulfatase_gene_table_funct,offset=.01,colnames_angle=90,hjust=1,
               legend_title="sulfatase function", colnames=FALSE) +  
    scale_fill_brewer(palette="Paired")  )
ggsave(sulfatase_heatmap_full,file=file.path(dir,'output/sulfatase_heatmap_full.pdf'),height=6,width=10)

#GH16
GH16_genes <- annotation %>% filter(func_consensus=='GH16') %>% pull(Gene)
GH16_gene_table <- get_gene_table(GH16_genes)
GH16_gene_table_funct <- code_gene_table_by_func(GH16_gene_table,'Annotation')
(GH16_gene_table_funct_plot <- gheatmap(p,GH16_gene_table_funct ,offset=.01,colnames_angle=90,hjust=1,
               legend_title="sulfatase function", colnames=FALSE) +  
    scale_fill_brewer(palette="Paired"))

#GH2
GH2_genes <- annotation %>% filter(func_consensus=='GH2') %>% pull(Gene)
GH2_gene_table <- get_gene_table(GH2_genes)
GH2_gene_table_funct <- code_gene_table_by_func(GH2_gene_table,'Annotation')
(GH2_gene_table_funct_plot <- gheatmap(p,GH2_gene_table_funct ,offset=.01,colnames_angle=90,hjust=1,
               legend_title="sulfatase function", colnames=FALSE) +  
    scale_fill_brewer(palette="Paired"))

#GH20
GH20_genes <- annotation %>% filter(func_consensus=='GH20') %>% pull(Gene)
GH20_gene_table <- get_gene_table(GH20_genes)
GH20_gene_table_funct <- code_gene_table_by_func(GH20_gene_table,'Annotation')
(GH20_gene_table_funct_plot <- gheatmap(p,GH20_gene_table_funct ,offset=.01,colnames_angle=90,hjust=1,
               legend_title="sulfatase function", colnames=FALSE) +  
    scale_fill_brewer(palette="Paired"))
```




### Figure 2: XB1A model to captive ape strains

### AAI
```{bash}
#ls results/pangenome/Bacteroides_xylanisolvens/faa/
#cat results/pangenome/Bacteroides_xylanisolvens/faa/* > #results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa
```


## Not in use
```{r}
library(Biostrings)
dir = 'results/pangenome/Bacteroides_xylanisolvens'
Bxy_metadata <- read_tsv(file.path(dir,'metadata_edit.txt'))
pres_abs <- read.csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),
                     col_types = cols())
colnames(pres_abs)[str_detect(pres_abs[1,],'FEMEHMHH')]
annotation <- read_tsv(file= file.path(dir,'output/roary_nosplitparalogs_annotation.txt'),
                       col_types = cols())
sulfatase_genes = annotation %>% filter(sulfatase_family=='S1')
sulfatase_genes$Gene

faa_file = 'results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa'
db = readAAStringSet(faa_file)
names(db) <- sapply(strsplit(names(db)," "), `[`, 1)
system('mkdir results/pangenome/Bacteroides_xylanisolvens/output/og_seqs')

get_ave_AAI <- function(gene){
  prot <- pres_abs %>% filter(Gene==gene) %>% select(Bxy_metadata$isolate) %>% as.character()
  prot <- unlist(strsplit(prot, "[\t]"))
  prot = prot[!is.na(prot)]
  prot_seqs = db[names(db) %in% prot]
  print(paste(gene,length(prot_seqs)))
  if (length(prot_seqs)>1){
    file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/og_seqs',paste0(gene,'.faa'))
    writeXStringSet(prot_seqs,file)
    system(paste0('mafft --auto ',file,' > ',file,'.align'))
    prot_align = read.alignment(file=paste0(file,'.align'),format='fasta')
    print(file)
    prot_align
    dist = dist.alignment(prot_align,"identity")
    dist = as.matrix(dist)
    dist_file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/og_seqs',
                          paste0(gene,'_dist.txt'))
    write_tsv(as.data.frame(dist),file=dist_file)
    dist[lower.tri(dist,diag = TRUE)] <- NA
    Ave_AAI = (1-mean(dist,na.rm=TRUE))*100
    print(Ave_AAI)
    return(Ave_AAI)} else{
       value=NA
       dist_file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/og_seqs',
                          paste0(gene,'_dist.txt'))
       write_tsv(value,file=dist_file)
      return(NA)}}
get_ave_AAI(gene = "group_3050") 

completed = list.files('results/pangenome/Bacteroides_xylanisolvens/output/og_seqs/',
                       pattern = '_dist.txt')
(completed = unique(sapply(str_split(completed,'_dist'), `[`, 1)))
(sulfatase_to_do = setdiff(sulfatase_genes$Gene,completed))
for (gene in sulfatase_to_do) {
  get_ave_AAI(gene)}

get_AAI_from_dist <- function(gene){
   dist_file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/og_seqs',
                          paste0(gene,'_dist.txt'))
   dist = read_tsv(file=dist_file)
   dist=as.matrix(dist)
   dist[lower.tri(dist,diag = TRUE)] <- NA
   Ave_AAI = (1-mean(dist,na.rm=TRUE))*100
   return(Ave_AAI) 
}
get_AAI_from_dist(gene = "group_3050") 

#OG_AAI = lapply(sulfatase_genes,get_AAI_from_dist)
#names(OG_AAI) <- sulfatase_genes
#OG_AAI_df = OG_AAI %>% 
#  as.data.frame() %>%
#  t() %>% as.data.frame() %>% 
#  rownames_to_column(var='Gene')
#colnames(OG_AAI_df)<-c('Gene','Ave_AAI')
#write_tsv(OG_AAI_df,file='results/pangenome/Bacteroides_xylanisolvens/output/og_seqs/all_Ave_AAI.txt')
OG_AAI_df = read_tsv(file='results/pangenome/Bacteroides_xylanisolvens/output/og_seqs/all_Ave_AAI.txt')
```

```{r}
sulfatase_genes <- annotation %>% filter(sulfatase_family=='S1') %>% pull(Gene)

present_in_isolates <- function(list_of_genes,list_of_isolates,name) {
  isolate_gene_table <- annotation %>% 
  left_join(gene_table,by='Gene') %>%
  filter(Gene %in% list_of_genes) %>%
  select(Gene,list_of_isolates) %>%
  mutate(total = rowSums(select_(., "-Gene"))) %>%
  select(Gene,total)  
  print(min(isolate_gene_table$total[isolate_gene_table$total!=0])) 
  isolate_gene_table = isolate_gene_table %>%
         mutate(binary = ifelse(total>0,name,NA),
         prop = total/max(total)) %>%
         column_to_rownames('Gene') 
  isolate_gene_table = isolate_gene_table %>% select(binary)
  colnames(isolate_gene_table) <-c(name)
  return(isolate_gene_table)
}

XBA1 = c("GCA.000210075.1.ASM21007v1")
XBA1 = present_in_isolates(sulfatase_genes,XBA1,'XBA1')
captive_isolates = Bxy_metadata %>% filter(human_ape=='ape') %>% pull(isolate)
ape = present_in_isolates(sulfatase_genes,captive_isolates,'ape')
mixedhost_isolates =  Bxy_metadata %>% filter(captive_clade=='mixedhost') %>% pull(isolate)
mixedhost = present_in_isolates(sulfatase_genes,mixedhost_isolates,'mixedhost')
gor1_isolates =  Bxy_metadata %>% filter(captive_clade=='gorilla1') %>% pull(isolate)
gor1 = present_in_isolates(sulfatase_genes,gor1_isolates,'gorilla1')
gor2_isolates =  Bxy_metadata %>% filter(captive_clade=='gorilla2') %>% pull(isolate)
gor2 = present_in_isolates(sulfatase_genes,gor2_isolates,'gorilla2')
sulfadata <- cbind(XBA1,ape,mixedhost,gor1,gor2)

sulfadata <- sulfadata %>%
  rownames_to_column(var='Gene') %>%
  select(Gene,XBA1,mixedhost,gorilla1,gorilla2) %>%
  left_join(annotation,by='Gene')

##make tree of sulfa genes
#faa_file = 'results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa'
#db = readAAStringSet(faa_file)
#names(db) <- sapply(strsplit(names(db)," "), `[`, 1)
#prot_seqs = db[names(db) %in% sulfadata$Gene.ID]
file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/sulfatase.faa')
#writeXStringSet(prot_seqs,file)
#system(paste0('mafft --auto ',file,' > ',file,'.align'))
#system(paste0('fasttree ',file,'.align > ',file,'.tree'))
sulfa_tree <- read.tree(paste0(file,'.tree'))
get_group <- function(prot){sulfadata$Gene[sulfadata$Gene.ID == prot]}
sulfa_tree$tip.label <- as.character(lapply(sulfa_tree$tip.label,get_group))

pres <- sulfadata %>% 
  dplyr::select(Gene,XBA1,mixedhost,gorilla1,gorilla2) %>% 
  column_to_rownames(var='Gene')
func <- sulfadata %>% 
  dplyr::select(Gene,func_consensus) %>% 
  column_to_rownames(var='Gene') %>% as.matrix() 
subfamily <- sulfadata %>% 
  dplyr::select(Gene,sulfatase_subfamily) %>% 
  filter(Gene %in% sulfa_tree$tip.label) %>% 
  mutate(sulfatase_subfamily = as.character(sulfatase_subfamily)) %>%
  column_to_rownames(var='Gene') %>% as.matrix() 
colourCount = length(unique(subfamily[,'sulfatase_subfamily']))
subfamily_pal = colorRampPalette(brewer.pal(colourCount, "Paired"))(colourCount)

not_in_XBA1 = rownames(XBA1)[is.na(XBA1$XBA1)]
sulfa_tree_gg = ggtree(sulfa_tree,layout='circular') +
  geom_point2(aes(subset=(label%in%not_in_XBA1)), 
              shape=21, size=1, fill='black') + 
  geom_treescale() 
p <- gheatmap(sulfa_tree_gg, pres, colnames_angle=90,hjust=1,width=.4)  + scale_fill_manual(values=brewer.pal(4, "Dark2"))
p <- p + new_scale_fill()
(p2 <- gheatmap(p,func,colnames_angle=90,hjust=1,offset=.99,width=.05) +
    scale_fill_manual(values = brewer.pal(9, "Paired")))
ggsave(p2,file=file.path(dir,'output/Figure2_sulfatase_circle_v1.pdf'),height=12,width=10)

sulfa_tree$node.label
sulfa_tree_gg + geom_hilight(node=60, fill="darkgreen", alpha=.1)
sub = '8'
got_subfam_node <- function(sub){
  subfam = sulfadata$Gene[sulfadata$sulfatase_subfamily==sub]
  subfam = intersect(subfam,sulfa_tree$tip.label)
  subfam_node = getMRCA(sulfa_tree,subfam)
  return(subfam_node)
}
got_subfam_node('8')
table(sulfadata$sulfatase_subfamily)
nodes = lapply(unique(sulfadata$sulfatase_subfamily),got_subfam_node)
names(nodes) = unique(sulfadata$sulfatase_subfamily)
unlist(nodes)
sulfa_tree_gg_v2 = sulfa_tree_gg +
  geom_hilight(node=79, fill="darkgreen", alpha=.1) + 
  geom_cladelabel(node=79,label="S1_15") +  
  geom_hilight(node=94, fill="blue", alpha=.1) +
  geom_cladelabel(node=94,label="S1_9") +  
  geom_hilight(node=71, fill="red", alpha=.1) +
  geom_cladelabel(node=71,label="S1_20") +  
  geom_hilight(node=93, fill="orange", alpha=.1) +
  geom_cladelabel(node=93,label="S1_11") +
  geom_hilight(node=99, fill="yellow", alpha=.1) + 
  geom_cladelabel(node=99,label="S1_67") +  
  geom_hilight(node=84, fill="purple", alpha=.1) +
  geom_cladelabel(node=84,label="S1_4") +  
  geom_hilight(node=75, fill="royalblue", alpha=.1) +
  geom_cladelabel(node=75,label="S1_14") +  
  geom_hilight(node=58, fill="pink", alpha=.1) +
  geom_cladelabel(node=58,label="S1_8") +
  geom_hilight(node=78, fill="darkorange", alpha=.1) +
  geom_cladelabel(node=78,label="S1_22") +  
  geom_hilight(node=89, fill="green", alpha=.1) +
  geom_cladelabel(node=89,label="S1_16") +  
  geom_hilight(node=55, fill="red2", alpha=.1) +
  geom_cladelabel(node=55,label="S1_30") +  
  geom_hilight(node=100, fill="yellow4", alpha=.1) +
  geom_cladelabel(node=100,label="S1_28") + 
  geom_treescale() 

p <- gheatmap(sulfa_tree_gg_v2, pres, colnames_angle=90,hjust=1,width=.4)  + scale_fill_manual(values=brewer.pal(4, "Dark2"))
p <- p + new_scale_fill()
(p2 <- gheatmap(p,func,colnames_angle=90,hjust=1,offset=.99,width=.05) +
    scale_fill_manual(values = brewer.pal(9, "Paired")))
ggsave(p2,file=file.path(dir,'output/Figure2_sulfatase_circle_v2.pdf'),height=12,width=10)


```


#### Figure 3
```{r}
dir = 'results/pangenome/Bacteroides_xylanisolvens'
species = 'Bacteroides_xylanisolvens'
Bxy_metadata = read_tsv(file=file.path(dir,'metadata_edit.txt'),col_types = cols())
Bxy_metadata_subset <- Bxy_metadata %>% dplyr::select(isolate,host,human_ape,captive_clade)
annotation = read_tsv(file=file.path(dir,'output/roary_nosplitparalogs_annotation.txt'),
                      col_types = cols())
gene_table = read_tsv(file=file.path(dir,'output/roary_nosplitparalogs_gene_table.txt'),
                      col_types = cols())
Bxy_tree <- read.tree(file = file.path(dir,paste0(species,'.tre')))

captive_clade_table <- gene_table %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>% 
  filter(present>0) %>%
  left_join(Bxy_metadata_subset,by='isolate') %>%
  group_by(Gene,captive_clade) %>%
  tally() %>%
  pivot_wider(names_from=captive_clade,values_from=n,values_fill=0) 
  
captive_clade_table <- captive_clade_table %>%  
  mutate(mixedhost = as.numeric(mixedhost),
         gorilla1 = as.numeric(gorilla1),
         gorilla2 = as.numeric(gorilla2),
         cladecount = sum(if_else(mixedhost>0 & (mixedhost>.5*max(captive_clade_table$mixedhost)),1,0),
                          if_else(gorilla1>0 & (gorilla1>.5*max(captive_clade_table$gorilla1)),1,0),
                          if_else(gorilla2>0 & (gorilla2>.5*max(captive_clade_table$gorilla2)),1,0))) 
captive_clade_table <- captive_clade_table %>%  
  filter(unassigned<3,cladecount>1) 
captive_convergent_genes <- captive_clade_table$Gene
captive_gene_table <- gene_table %>% 
  filter(Gene %in% captive_convergent_genes) %>%
  column_to_rownames(var='Gene') %>%
  t()
HOST_table <- Bxy_metadata %>% 
  dplyr::select(isolate,host_site) %>%
  column_to_rownames(var='isolate') 
library(RColorBrewer)
library(ggnewscale)

p <- gheatmap(ggtree(Bxy_tree) + ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1)  +
              scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label)) 
p <- p + new_scale_fill()
(p2 <- gheatmap(p,captive_gene_table,offset=.05,colnames_angle=90,hjust=1) +
                scale_fill_viridis_c(option="C"))
ggsave(p2,file =file.path(dir,'output/convergenetly_gained_captive.pdf'))
```


```{r warning = FALSE, message=FALSE}
source('scripts/genomic_island_function.R')
#system(paste0('rm -r ',file.path(dir,'captive_convergent_genes')))
list_of_genes = captive_convergent_genes
pres_abs <- read_csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),
                     col_types = cols())
Bxy_metadata = read_tsv(file.path(dir,'metadata_edit.txt'),col_types = cols())
window_size=10
get_captive_convergent_genes = function(isolate){

    isolate_old = metadata$isolate_old[metadata$isolate==isolate]
    gff_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.gff')
    genomic_islands_outfile = file.path(dir,'captive_convergent_genes/gff',
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
    gff_islands = get_genomic_island(pres_abs=pres_abs,
                               list_of_genes=captive_convergent_genes,
                               isolate=isolate,
                               gff_file=gff_file,
                               window_size = window_size,
                               outfile=genomic_islands_outfile)
}

get_captive_convergent_genes('P17.C2')
captive_isolates = Bxy_metadata$isolate[Bxy_metadata$human_ape=='ape']
for (isolate in captive_isolates) {
  print(isolate)
  get_captive_convergent_genes(isolate)
}

#generate heatmap only with clusters longer than 10 genes
captive_PUL_table = data.frame()
captive_PUL_table = data.frame(Gene=captive_convergent_genes)
for  (isolate in captive_isolates) {
  genomic_islands_outfile = file.path(dir,'captive_convergent_genes/gff',
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
  iso = read_tsv(file=genomic_islands_outfile)
  iso = iso %>% filter(size>5) %>%
  select(Gene,gene_span) %>% 
  mutate(Gene=make.names(Gene, unique=TRUE)) 
  colnames(iso) = c('Gene',isolate)
  captive_PUL_table = captive_PUL_table %>% full_join(iso,by='Gene')}

#cluster_over_10 = captive_PUL_table$Gene[rowSums(!is.na(captive_PUL_table))>1]
#captive_PUL_table = captive_PUL_table %>% 
#  filter(Gene %in% cluster_over_10) 

captive_PUL_tableM = captive_PUL_table %>% 
  column_to_rownames(var='Gene') %>%
  arrange_all() %>% t()
not_captive = Bxy_tree$tip.label[!Bxy_tree$tip.label %in% captive_isolates]
Bxy_tree_captive = drop.tip(Bxy_tree,not_captive)
(cluster_heatmap_plot <- gheatmap(ggtree(Bxy_tree_captive) +geom_tiplab() + ylim(-10,NA),captive_PUL_tableM,colnames_angle=90,hjust=1,offset = .01) +
    scale_fill_manual(values=colorRampPalette(brewer.pal(13, "Paired"))(13)))

captive_PUL_table_df = captive_PUL_table %>% 
  pivot_longer(cols=all_of(captive_isolates),names_to='isolate',values_to='gene_span') %>%
  select(gene_span,isolate) %>% 
  distinct() %>%
  filter(!is.na(gene_span))

updown_size=20
gene_span='group_16996_group_20520'
isolate = 'P17.C2'

get_island_fna = function(isolate,gene_span){

    isolate_old = metadata$isolate_old[metadata$isolate==isolate]
    gff_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.gff')
    fna_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.fna')
    genomic_islands_outfile = file.path(dir,'captive_convergent_genes/gff',
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
    island_fasta_outfile = file.path(dir,'captive_convergent_genes/fna/',gene_span,
                                 paste0(isolate,'_',gene_span,'_updown_',updown_size,'.fna'))
    output_island_fasta(
          isolate=isolate,
          gene_span = gene_span,
          genomic_islands_outfile = genomic_islands_outfile,
          island_fasta_outfile  = island_fasta_outfile,
          updown_size = updown_size ,
          fna_file = fna_file,
          gff_file = gff_file
          )
}


mapply(get_island_fna,captive_PUL_table_df$isolate,captive_PUL_table_df$gene_span)



```

```{r}
species = 'Bacteroides_ovatus'
dir = paste0('results/pangenome/',species)
Bxy_sum = read_tsv(file.path(dir,'pairwise_gene_gain_50strains_summary.txt'),col_types = cols())
Bxy_gi = read_tsv(file.path(dir,'pairwise_gene_gain_50strains_gff_islands.txt'),col_types = cols())

df = Bxy_sum
dist_cutoff = .01
df_filt = filter(df,tree_dist>dist_cutoff)

#Pangenome dist
summary(lm_robust(dist_bray_curtis ~ tree_dist,data = df))
summary(lm_robust(dist_bray_curtis ~ I(tree_dist^2) + tree_dist,data = df))
cor.test(df$dist_bray_curtis,df$tree_dist,method = 'spearman')

bray = ggplot(Bxy_sum,aes(x=tree_dist,y=dist_bray_curtis)) + 
  geom_point(color='blue4',alpha = .05,size=3) + 
  theme_bw() +
  stat_smooth(method = "lm", color = 'blue4',formula = y ~ x + I(x^2), size = 1) +
  ylab('Pangenome distance (Bray-Curtis)')

#Events
summary(lm_robust(n_unique_genes ~ tree_dist,data = df))
summary(lm_robust(n_unique_genes ~ I(tree_dist^2) + tree_dist,data = df))
cor.test(df$n_unique_genes,df$tree_dist,method = 'spearman')
summary(lm_robust(bray_CN100 ~ tree_dist ,data = df))
summary(lm_robust(bray_CN100 ~ I(tree_dist^2) + tree_dist, data = df))
cor.test(df$bray_CN100,df$tree_dist,method = 'spearman')
(events = ggplot(df,aes(x=tree_dist)) + 
  geom_point(aes(y=n_unique_genes),color='darkorange',alpha = .05,size=3) + 
  stat_smooth(aes(y=n_unique_genes), 
              method = "lm", color = 'darkorange',formula = y ~ I(x^2) + x, size = 1) +
  geom_point(aes(y=bray_CN100),color='purple',alpha = .1,size=3)   +
  stat_smooth(aes(y=bray_CN100), 
              method = "lm", color = 'purple',formula = y ~ I(x^2) + x, size = 1) +
    theme_bw() +
    scale_y_continuous(
    name = "Number of unique genes",
    sec.axis = sec_axis( trans=~., name="Number of events")
    ))

##Fractions of events/genes by size 
Bxy_gi_filt = Bxy_gi %>% 
  filter(tree_dist>0) %>% 
  mutate(cat='gene') %>%
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,210)))
mean(Bxy_gi_filt$size)
median(Bxy_gi_filt$size)

Bxy_gi_clust = Bxy_gi_filt %>% 
  distinct(comp,tree_dist,name,size) %>%
  mutate(cat='event') %>%
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,210)))
mean(Bxy_gi_clust$size)
median(Bxy_gi_clust$size)

Bxy_gi_combined = Bxy_gi_filt %>% add_rows(Bxy_gi_clust) %>% 
  group_by(cat,size_category) %>%
  summarise(count = n()) %>%
  mutate(freq = count / sum(count)) %>% 
  as.data.frame()

(frac_events_genes = Bxy_gi_combined %>% 
  ggplot(aes(x=size_category,y=freq,fill=cat)) +
  geom_bar(stat="identity", position=position_dodge()) + 
  theme_bw() +
  xlab('Genomic island size')+
  scale_y_continuous(
    name = "Fraction of events",
    sec.axis = sec_axis( trans=~., name="Fraction of genes")
    )+ 
  theme(legend.position = c(.95, .95), 
       legend.justification = c(.95, .95)
       ))

#island size
island_size = Bxy_gi_clust %>% 
  group_by(comp,tree_dist) %>% 
  summarize(mean_size = mean(size))
summary(lm(mean_size~tree_dist,island_size))
summary(lm(mean_size~I(tree_dist^2) + tree_dist,island_size))
cor.test(island_size$mean_size,island_size$tree_dist,method = 'spearman')
island_size_plot = island_size %>% 
  ggplot(aes(x=tree_dist,y=mean_size)) + 
  geom_point(color='red2',alpha = .05,size=3)   +
  stat_smooth(aes(y=mean_size), 
              method = "lm", color = 'red2',formula = y ~ I(x^2) + x, size = 1) +
  theme_bw()  +
  ylab('Average genomic island size')

(figure4 = plot_grid(bray,events,island_size_plot,frac_events_genes,nrow=2,labels=c('A','B','C','D')))
ggsave(figure4,file =file.path(dir,'output/figure4.pdf'),width=8,height=8)

```


```{r}
species = 'Bacteroides_xylanisolvens'
dir = paste0('results/pangenome/',species)
Bxy = read_tsv('results/pangenome/Bacteroides_xylanisolvens/pairwise_gene_gain_50strains.txt',col_types = cols())
Bov = read_tsv('results/pangenome/Bacteroides_ovatus/pairwise_gene_gain_50strains.txt',col_types = cols())
Bfr = read_tsv('results/pangenome/Bacteroides_fragilis/pairwise_gene_gain_50strains.txt',col_types = cols())
Bth = read_tsv('results/pangenome/Bacteroides_thetaiotaomicron/pairwise_gene_gain_50strains.txt',col_types = cols())
Bac = Bxy %>% add_row(Bov) %>% add_row(Bfr) %>% add_row(Bth)


Bac <- Bac %>% group_by(taxonomy_Species.iso1)%>% mutate(
  max_tree_dist = max(tree_dist),
  tree_dist_norm = tree_dist/mean(tree_dist)) %>% as.data.frame()

Bac_tree_dist_sum = Bac %>% 
  group_by(taxonomy_Species.iso1) %>% 
  summarize(min_tdn = min(tree_dist_norm),max_tdn = max(tree_dist_norm))
Bac <- Bac %>%  filter(tree_dist_norm>max(Bac_tree_dist_sum$min_tdn),
                       tree_dist_norm<min(Bac_tree_dist_sum$max_tdn))

summary(lm_robust(dist_bray_curtis ~ I(tree_dist_norm^2) + tree_dist_norm + taxonomy_Species.iso1, data = Bac))
summary(lm_robust(dist_bray_curtis ~  tree_dist_norm + taxonomy_Species.iso1, data = Bac))

figure5 = Bac %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=dist_bray_curtis) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=1,size=2) +
  theme_bw()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~ I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~  I(x^2) + x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  ylab('Pangenome distance (Bray-Curtis)')
  
ggsave(figure5,file = file.path(dir,'output/figure5_bcdist.pdf'),width=8,height=6)

summary(lm_robust(bray_CN100 ~ I(tree_dist_norm^2) + tree_dist_norm + taxonomy_Species.iso1, data = Bac))
summary(lm_robust(bray_CN100 ~  tree_dist_norm + taxonomy_Species.iso1, data = Bac))
lapply(df$model,summary)
(figure5B = Bac %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=bray_CN100) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=1,size=2) +
  theme_bw()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~ x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~ x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  x, size = 1, se=TRUE)+
    ylab('Number of events'))

ggsave(figure5B,file = file.path(dir,'output/figure5_CN100.pdf'),width=8,height=6)



```
```{r}
Bxy = read_tsv('results/pangenome/Bacteroides_xylanisolvens/pairwise_gene_gain_50strains_gff_islands.txt',
               col_types = cols()) 
Bov = read_tsv('results/pangenome/Bacteroides_ovatus/pairwise_gene_gain_50strains_gff_islands.txt',
               col_types = cols())
Bfr = read_tsv('results/pangenome/Bacteroides_fragilis/pairwise_gene_gain_50strains_gff_islands.txt',
               col_types = cols())
Bth = read_tsv('results/pangenome/Bacteroides_thetaiotaomicron/pairwise_gene_gain_50strains_gff_islands.txt',
               col_types = cols()) 
Bac = Bxy %>% add_row(Bov) %>% add_row(Bfr) %>% add_row(Bth)
Bac <- Bac %>% group_by(taxonomy_Species.iso1)%>% mutate(
  max_tree_dist = max(tree_dist),
  tree_dist_norm = tree_dist/mean(tree_dist)) %>% as.data.frame()
Bac = Bac %>% filter(tree_dist_norm>max(Bac_tree_dist_sum$min_tdn),
                       tree_dist_norm<min(Bac_tree_dist_sum$max_tdn)) 
Bac = Bac %>% distinct(comp,tree_dist,name,size,taxonomy_Species.iso1) %>% 
  group_by(taxonomy_Species.iso1,comp) %>% summarise(mean_size = mean(size)) 

Bac %>% ggplot() + 
  aes(x=taxonomy_Species.iso1,y=mean_size,fill=taxonomy_Species.iso1) + geom_boxplot()
```


```{r}
mean(test_clust$size)

summary(lm(bray_CL50 ~ tree_dist,data = df))
summary(lm(bray_CL50 ~ I(tree_dist^2) + tree_dist,data = df))
ggplot(df,aes(x=tree_dist)) + 
  geom_point(aes(y=bray_CL50),color='green4',alpha = .1,size=3) + 
  stat_smooth(aes(y=bray_CL50), 
               method = "lm", color = 'green4',formula = y ~ x, size = 1) 
summary(lm(bray_CL50 ~ tree_dist,data = filter(df,tree_dist>.001)))
summary(lm(bray_CL50 ~ I(tree_dist^2) + tree_dist,data = filter(df,tree_dist>.001)))

ggplot(filter(df,tree_dist>.001),aes(x=tree_dist)) + 
  geom_point(aes(y=bray_CL50),color='green4',alpha = .1,size=3) + 
  stat_smooth(aes(y=bray_CL50), 
               method = "lm", color = 'green4',formula = y ~ x + I(x^2), size = 1) 

ngenes = df %>% 
  dplyr::select(tree_dist,n_unique_genes,bray_CN100) %>%
  pivot_longer(cols=c(n_unique_genes,bray_CN100),names_to='events',values_to='n') 
ngenes %>%
  ggplot(aes(x=tree_dist,y=n)) + 
  geom_hex(data=filter(ngenes,events=='bray_CN100'),alpha=.5)+
  scale_fill_gradient(low="lightblue", high="blue4")  +  new_scale_fill() +
 geom_hex(data=filter(ngenes,events=='n_unique_genes'),alpha=.5) +
  scale_fill_gradient(low="lightgreen", high="green4") +
  theme_bw() 
ggplot(df,aes(x=tree_dist,y=bray_CL50)) + geom_point() + theme_bw() 


#get mean
species = 'Bacteroides_xylanisolvens'
dir = paste0('results/pangenome/',species)
metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
pres_abs <- read_csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),col_types = cols())
pw_outdir = file.path(dir,'pairwise_gene_gain')
print(pw_outdir)
window_size = 5
tree_file = file.path(dir,paste0("phylogeny/RAxML_bipartitions.",species))
tree = read.tree(tree_file)
tree = drop.tip(tree,outgroup)
tree_dist <- cophenetic.phylo(tree)
tree_dist

#dereplication closely related strains
set.seed(123)
hc = hclust(as.dist(tree_dist))
cl = cutree(hc,k=50)
df = data.frame(cl) %>% 
  rownames_to_column(var='isolate') %>%
  group_by(cl) %>% sample_n(1)
tree_distm <- melt_dist(tree_dist) %>% 
  dplyr::rename(tree_dist=dist) %>%
  filter(iso1 %in% df$isolate, iso2 %in% df$isolate)
tree_distm = tree_distm %>% 
  mutate(comp = paste(iso1,iso2,sep='_')) 
#exclude outgroup from comparisons
files = list.files(file.path(pw_outdir),'.island_gff.txt')
files = paste0(pw_outdir,'/',files)
print(length(files))
completed = sapply(strsplit(files, "/"), "[", 5)
completed = sapply(strsplit(completed,"_window"),"[", 1)
files = files[completed %in% tree_distm$comp]
library(data.table)
read_file = function(x){
  df = read.csv(x,sep='\t')
  comp =  sapply(strsplit(x, "/"), "[", 5)
  comp = sapply(strsplit(comp,"_window"),"[", 1)
  df = df %>% mutate(comp = comp)
}
test = rbindlist(lapply(files,read_file))
test = test %>% right_join(tree_distm, by='comp')
mean(test$size)

test_mean=test %>% group_by(comp,tree_dist) %>% summarise(mean_size=mean(size),
                                                          median_size=median(size)) 
mean(test_filt$median_size)
mean(test_filt$mean_size)
ggplot(filter(test_mean),aes(x=tree_dist)) + 
  geom_point(aes(y=mean_size),color='green4',alpha = .1,size=3) + 
  stat_smooth(data=filter(test_mean,tree_dist>.001),aes(y=mean_size), 
              method = "lm", color = 'green4',formula = y ~ x, size = 1) +
  geom_point(aes(y=median_size),color='royalblue',alpha = .1,size=3) + 
  stat_smooth(data=filter(test_mean,tree_dist>.001),aes(y=median_size), 
              method = "lm", color = 'royalblue',formula = y ~ x, size = 1)  
summary(lm(mean_size~tree_dist,filter(test_mean,tree_dist>.01)))
summary(lm(median_size~tree_dist,filter(test_mean,tree_dist>.01)))
test_filt = filter(test_mean,tree_dist>.01)
mean(test_filt$median_size)
mean(test_filt$mean_size)

test_clust = test %>% 
  distinct(comp,tree_dist,name,size) %>%
  mutate(cat='event') %>%
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,1000)))
test_clust %>% ggplot(aes(size_category)) +
  geom_bar(aes(y=..count../sum(..count..)))
mean(test_clust$size)

test = test %>% 
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,1000)))
test %>% ggplot(aes(size_category)) +
  geom_bar(aes(y=..count../sum(..count..)))



mean(test_filt$median_size)
mean(test_filt$mean_size)


test %>% ggplot(aes(x=size)) + geom_histogram()
test %>% filter(tree_dist>.001) %>% ggplot() + geom_density(aes(x=size),adjust = 5) + theme_bw() + geom_vline(xintercept = median(test$size))
test %>% filter(tree_dist>.001) %>%  
  ggplot(aes(x=size)) + 
  geom_histogram(aes(y=..count../sum(..count..)),binwidth=5,fill='royalblue') + 
  theme_bw() + 
  geom_vline(xintercept = mean(test$size),linetype=2)
ggplot(test, aes(x=size)) + geom_bar(aes(y = ..prop.., group = 1))

ggplot(filter(test_mean,tree_dist>.001),aes(x=tree_dist)) + 
  geom_point(aes(y=mean_size),color='green4',alpha = .1,size=3) + 
  stat_smooth(aes(y=mean_size), 
              method = "lm", color = 'green4',formula = y ~ x, size = 1) 
summary(lm(mean_size~tree_dist,filter(test_mean,tree_dist>.01)))

test_nuin=test %>% group_by(comp,tree_dist) %>% tally()
ggplot(test_nuin,aes(x=tree_dist,y=n)) + 
  
  geom_point(color='green4',alpha = .1,size=3) +
  stat_smooth(aes(y=n), 
              method = "lm", color = 'green4',formula = y ~ x, size = 1) 


summary(lm(n~tree_dist,filter(test_nuin,tree_dist>.01)))
summary(lm(n~I(tree_dist^2)+tree_dist,filter(test_nuin,tree_dist>.01)))


test_CN100=test %>% 
  distinct(comp,tree_dist,name) %>% 
  group_by(comp,tree_dist) %>% tally() %>% 
  filter(tree_dist>.01)
ggplot(test_CN100,aes(x=tree_dist,y=n)) + 
  geom_point(color='green4',alpha = .1,size=3) 
summary(lm(n~tree_dist,filter(test_CN100,tree_dist>.01)))

predict(fit1,tree_dist=.01)
ggplot(test_CN100,aes(x=tree_dist,y=n)) + 
  geom_point(color='green4',alpha = .1,size=3) 
summary(lm(n~tree_dist,test_CN100))
summary(lm(n~I(tree_dist^2)+tree_dist,test_CN100))



```


## Figure 5 pangenome distance

```{r}
pandist <- function(species){
dir=paste0('results/pangenome/',species)
system(paste0('mkdir ',dir,'/output'))

#read-in metadata
metadata_file = file.path(dir,"metadata.txt")
metadata = read_tsv(metadata_file)
table(metadata$site)
table(metadata$host)
outgroup_names <- metadata %>% 
  filter(taxonomy_Species!=species) %>% 
  pull(isolate)

(tree_file = file.path(dir,paste0("phylogeny/RAxML_bipartitions.",species)))
tree <- ape::read.tree(tree_file)
tree$tip.label = str_replace_all(tree$tip.label,'[_-]','.') #clean up names
tree <- root(tree,outgroup=outgroup_names) 
write.tree(tree,file = file.path(dir,paste0(species,'.tre')))
(tree_plot <- ggtree(tree) %<+% metadata +
   geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
   geom_tippoint(aes(color=host_site), alpha=0.8)  + 
   geom_tiplab() + 
   xlim(NA,.5))

pres_abs <- read.csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(metadata$isolate,colnames(pres_abs))
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- pres_abs %>% 
   select(Gene,metadata$isolate)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)
write_tsv(gene_table, file= file.path(dir,'output/roary_nosplitparalogs_gene_table.txt'))

tree_dist <- cophenetic.phylo(tree)[metadata$isolate,metadata$isolate]
tree_distm <- melt_dist(tree_dist) %>% 
  rename(tree_dist=dist)

pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_bray <- as.matrix(vegdist(pan_dist, method="bray"))
pan_dist_bray <- pan_dist_bray[metadata$isolate,metadata$isolate]
pan_distm_bray <- melt_dist(pan_dist_bray) %>%
  rename(pan_dist_bray=dist)
pan_dist_jaccard <- as.matrix(vegdist(pan_dist, method="jaccard"))
pan_dist_jaccard <- pan_dist_jaccard[metadata$isolate,metadata$isolate]
pan_distm_jaccard <- melt_dist(pan_dist_jaccard) %>%
  rename(pan_dist_jaccard=dist)
tree_pan_dist <-tree_distm %>% 
   left_join(pan_distm_bray, by=c('iso1','iso2'))  %>% 
   left_join(pan_distm_jaccard, by=c('iso1','iso2')) 
metadata_subset <- metadata %>% select(isolate,host,sample,taxonomy_Species,X..predicted.genes)
tree_pan_dist_meta  <- tree_pan_dist  %>%    
  left_join(metadata_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and 
  left_join(metadata_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) %>%
  filter(taxonomy_Species.iso1==species,taxonomy_Species.iso2==species)
write_tsv(tree_pan_dist_meta, file= file.path(dir,'output/pangenome_dist.txt'))
}

write_annotation <- function(species){
dir=paste0('results/pangenome/',species)
system(paste0('mkdir ',dir,'/output'))
pres_abs <- read.csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
annotation <- pres_abs %>% select(Gene,Annotation) 
eggnog <- readr::read_tsv(
  file.path(dir,'eggnog_mapper/pan_genome_reference.emapper.annotations'),
  comment = '##')
eggnog <- eggnog %>% 
   rename('Gene.ID'='#query') %>% 
   select(Gene.ID,evalue,best_OG_cat,best_OG_name,best_OG_desc)
pan_genome_fasta <- read.fasta(
  file.path(dir,'roary_nosplitparalogs/pan_genome_reference.fa'), 
  whole.header = T)
full_sequence_names <- names(pan_genome_fasta) #link between Gene and Gene.ID
head(full_sequence_names)
full_sequence_names <- as.data.frame(full_sequence_names) %>%
  separate(full_sequence_names,sep=' ',into=c('Gene.ID','Gene'))
annotation <- annotation %>%
   left_join(full_sequence_names,by='Gene') %>%
   left_join(eggnog, by = 'Gene.ID')
write_tsv(annotation, file= file.path(dir,'output/roary_nosplitparalogs_annotation.txt'))

}
pandist('Bacteroides_xylanisolvens')
pandist('Bacteroides_ovatus')
write_annotation('Bacteroides_ovatus')
pandist('Bacteroides_fragilis')
write_annotation('Bacteroides_fragilis')
pandist('Bacteroides_thetaiotaomicron')
write_annotation('Bacteroides_thetaiotaomicron')
```


```{r}
dir=paste0('results/pangenome/Bacteroides_xylanisolvens')
Bxy_pandist <- read_tsv(file= file.path(dir,'output/pangenome_dist.txt'))
Bxy_pandist <- Bxy_pandist %>% mutate(tree_dist_norm = tree_dist/max(tree_dist))
dir=paste0('results/pangenome/Bacteroides_ovatus')
Bov_pandist <- read_tsv(file= file.path(dir,'output/pangenome_dist.txt'))
Bov_pandist <- Bov_pandist %>% mutate(tree_dist_norm = tree_dist/max(tree_dist))
dir=paste0('results/pangenome/Bacteroides_fragilis')
Bfr_pandist <- read_tsv(file= file.path(dir,'output/pangenome_dist.txt'))
Bfr_pandist <- Bfr_pandist %>% mutate(tree_dist_norm = tree_dist/max(tree_dist))
dir=paste0('results/pangenome/Bacteroides_thetaiotaomicron')
Bth_pandist <- read_tsv(file= file.path(dir,'output/pangenome_dist.txt'))
Bth_pandist <- Bth_pandist %>% mutate(tree_dist_norm = tree_dist/max(tree_dist))

Bt_pandist_all <- Bxy_pandist %>% 
  add_row(Bov_pandist) %>% 
  add_row(Bfr_pandist)  %>% 
  add_row(Bth_pandist)

Bt_pandist <- Bt_pandist_all %>% 
  droplevels() %>%
  filter(tree_dist>.0001)

Bt_pandist_bray_plot <- Bt_pandist_all %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=pan_dist_bray) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=.1,size=2) +
  theme_cowplot()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~  I(x^2) + x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  I(x^2) + x, size = 1, se=TRUE)
Bt_pandist_bray_plot
Bt_pandist_jaccard_plot <- Bt_pandist_all %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=pan_dist_jaccard) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=.1,size=2) +
  theme_cowplot()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~  I(x^2) + x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bt_pandist,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  I(x^2) + x, size = 1, se=TRUE)
Bt_pandist_jaccard_plot

dir=paste0('results/pangenome/Bacteroides_xylanisolvens')
ggsave(Bt_pandist_jaccard_plot,file=file.path(dir,'output/Bt_pandist_jaccard_plot.pdf'),height=6,width=8)
ggsave(Bt_pandist_bray_plot,file=file.path(dir,'output/Bt_pandist_bray_plot.pdf'),height=6,width=8)

sisters <- Bt_pandist_all %>%
  filter(tree_dist<.0001)

pan_dist_sisters <- sisters  %>%
  ggplot() + 
  aes(x=taxonomy_Species.iso1,y=pan_dist_bray,fill= taxonomy_Species.iso1) +
  geom_boxplot() +
  theme_bw() +
  scale_fill_manual(values=c('red2','blue','green2','orange')) 
pan_dist_sisters
ggsave(pan_dist_sisters,file=file.path(dir,'output/Bt_pandist_sisters.pdf'),height=6,width=8)
```
### PCOA

```{r}
species = 'Bacteroides_ovatus'
folder = 'Bacteroides_test'
output_rep_genomes <- function(species,folder){
dir=paste0('results/pangenome/',species)
metadata <- read_tsv("results/pangenome/Bacteroides_metadata.txt")
metadata <- metadata %>% filter(taxonomy_Species==species)
pres_abs <- read.csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
setdiff(metadata$isolate,colnames(pres_abs))
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- pres_abs %>% 
   select(Gene,metadata$isolate)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)
pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_bray <- vegdist(pan_dist, method="bray")
fit <- hclust(pan_dist_bray)
plot(fit) # display dendogram
groups <- cutree(fit, k=5) # cut tree into 5 clusters
df <- as.data.frame(groups) %>% 
  rownames_to_column(var='isolate') %>%  
  group_by(groups) %>%
  top_n(n = 2, wt = isolate)

system(paste0('mkdir -pv results/pangenome/',folder,'/gff'))

for (isolate in df$isolate) {
    isolate_old = metadata$isolate_old[metadata$isolate==isolate]
    print(isolate_old)
    infile = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.gff')
    outfile = paste0('results/pangenome/',folder,'/gff/',isolate,'.gff')
    system(paste0('cp ',infile,' ',outfile))
    }  
}

output_rep_genomes(species = 'Bacteroides_fragilis',
                   folder = 'Bacteroides_test')
output_rep_genomes(species = 'Bacteroides_ovatus',
                   folder = 'Bacteroides_test')
output_rep_genomes(species = 'Bacteroides_thetaiotaomicron',
                   folder = 'Bacteroides_test')
output_rep_genomes(species = 'Bacteroides_xylanisolvens',
                   folder = 'Bacteroides_test')
```
```{r}
species = 'Bacteroides_test'
dir=paste0('results/pangenome/',species)


#read-in metadata
metadata_file = "results/pangenome/Bacteroides_metadata.txt"
metadata = read_tsv(metadata_file)

pres_abs <- read.csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'))
colnames(pres_abs) <- str_replace_all(colnames(pres_abs),'[_-]','.')
isolates = intersect(colnames(pres_abs),metadata$isolate)
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- pres_abs %>% 
   select(Gene,isolates)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)
pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_bray <- vegdist(pan_dist, method="bray")
ord_nMDS<- metaMDS(comm= pan_dist_bray,
                   k = 2,
                   maxit = 1000, #have to increase tries for unifrac distances, otherwise they don't converge
                   trymax = 500,
                   wascores = FALSE,
                   autotransform = FALSE,
                   trace = 2,
                   noshare = FALSE,
                   parallel = 4)
ndim_stress = paste0('k = ',ord_nMDS$ndim,', stress = ',round(ord_nMDS$stress,2)) #extract stress and dimensions

ord_nMDS_df <- as.data.frame(scores(ord_nMDS))%>% 
  rownames_to_column(var='isolate') %>% 
  left_join(metadata,by='isolate') 
(pcoa_4Bt <- ggplot(ord_nMDS_df, aes(x = NMDS1, y = NMDS2)) +
  geom_point(size=3,alpha=.5, aes(colour = taxonomy_Species)) +
  theme_bw())
ggsave(pcoa_4Bt,file=file.path(dir,'pcoa_4Bt.pdf'))
as.matrix(pan_dist)

only_Bovox <- metadata %>% filter(taxonomy_Species=='Bacteroides_ovatus' | taxonomy_Species=='Bacteroides_xylanisolvens')
isolates = intersect(colnames(pres_abs),only_Bovox$isolate)
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- pres_abs %>% 
   select(Gene,isolates)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)
pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_bray <- vegdist(pan_dist, method="bray")
ord_nMDS<- metaMDS(comm= pan_dist_bray,
                   k = 2,
                   maxit = 1000, #have to increase tries for unifrac distances, otherwise they don't converge
                   trymax = 500,
                   wascores = FALSE,
                   autotransform = FALSE,
                   trace = 2,
                   noshare = FALSE,
                   parallel = 4)
ndim_stress = paste0('k = ',ord_nMDS$ndim,', stress = ',round(ord_nMDS$stress,2)) #extract stress and dimensions

ord_nMDS_df <- as.data.frame(scores(ord_nMDS))%>% 
  rownames_to_column(var='isolate') %>% 
  left_join(metadata,by='isolate') 

(pcoa_Bxy_Bov <- ggplot(ord_nMDS_df, aes(x = NMDS1, y = NMDS2)) +
  geom_point(size=3,alpha=.5, aes(colour = taxonomy_Species)) +
  theme_bw())
ggsave(pcoa_Bxy_Bov,file=file.path(dir,'pcoa_Bxy_Bov.pdf'))

only_Bovox <- metadata %>% filter(taxonomy_Species=='Bacteroides_ovatus' | taxonomy_Species=='Bacteroides_xylanisolvens')
isolates = intersect(colnames(pres_abs),only_Bovox$isolate)
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- pres_abs %>% 
   select(Gene,isolates)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)

pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_acc <- pan_dist[,colSums(pan_dist)<10]
dim(pan_dist_acc)
pan_dist_bray <- vegdist(pan_dist_acc, method="bray")
ord_nMDS<- metaMDS(comm= pan_dist_bray,
                   k = 2,
                   maxit = 1000, #have to increase tries for unifrac distances, otherwise they don't converge
                   trymax = 500,
                   wascores = FALSE,
                   autotransform = FALSE,
                   trace = 2,
                   noshare = FALSE,
                   parallel = 4)
ndim_stress = paste0('k = ',ord_nMDS$ndim,', stress = ',round(ord_nMDS$stress,2)) #extract stress and dimensions

ord_nMDS_df <- as.data.frame(scores(ord_nMDS))%>% 
  rownames_to_column(var='isolate') %>% 
  left_join(metadata,by='isolate') 
ggplot(ord_nMDS_df, aes(x = NMDS1, y = NMDS2)) +
  geom_point(size=3,alpha=.5, aes(colour = taxonomy_Species)) +
  theme_bw()


only_Bovox <- metadata %>% filter(taxonomy_Species=='Bacteroides_xylanisolvens'|
                                  taxonomy_Species=='Bacteroides_ovatus' )

sulfa <- dplyr::filter(pres_abs, grepl("sulfatase",Annotation))

isolates = intersect(colnames(pres_abs),only_Bovox$isolate)
isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
gene_table <- sulfa %>% 
   select(Gene,isolates)  %>%
   mutate_at(vars(-Gene),isblank) %>%
   filter(rowSums(across(where(is.numeric)))>0)

pan_dist <- gene_table %>% column_to_rownames(var='Gene') %>% t()
pan_dist_bray <- vegdist(pan_dist, method="bray")
ord_nMDS<- metaMDS(comm= pan_dist_bray,
                   k = 2,
                   maxit = 1000, #have to increase tries for unifrac distances, otherwise they don't converge
                   trymax = 500,
                   wascores = FALSE,
                   autotransform = FALSE,
                   trace = 2,
                   noshare = FALSE,
                   parallel = 4)
ndim_stress = paste0('k = ',ord_nMDS$ndim,', stress = ',round(ord_nMDS$stress,2)) #extract stress and dimensions

ord_nMDS_df <- as.data.frame(scores(ord_nMDS))%>% 
  rownames_to_column(var='isolate') %>% 
  left_join(metadata,by='isolate') 
ggplot(ord_nMDS_df, aes(x = NMDS1, y = NMDS2)) +
  geom_point(size=3,alpha=.5, aes(colour = taxonomy_Species)) +
  theme_bw()

```



```{r}
source('scripts/genomic_island_function.R')
dir = 'results/pangenome/Bacteroides_xylanisolvens'

Bxy_metadata = read_tsv(file.path(dir,'metadata_edit.txt'),col_types = cols())
pres_abs <- read_csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),col_types = cols())
pw_outdir = file.path(dir,'pairwise_gene_gain')
window_size = 5

pairwise_gene_gain_runner = function(iso1,iso2) {
  iso1_old = Bxy_metadata$isolate_old[Bxy_metadata$isolate==iso1]
  iso2_old = Bxy_metadata$isolate_old[Bxy_metadata$isolate==iso2]
  iso1_gff_file = paste0('results/pangenome/prokka/',iso1_old,'/',iso1_old,'.gff')
  iso2_gff_file = paste0('results/pangenome/prokka/',iso2_old,'/',iso2_old,'.gff')
  print(iso1)
  print(iso2)
  pairwise_gene_gain(
    pres_abs = pres_abs,
    iso1=iso1,
    iso2=iso2,
    iso1_gff_file = iso1_gff_file,
    iso2_gff_file = iso2_gff_file,
    window_size = window_size,
    pw_outdir = pw_outdir)
}


pairwise_gene_gain_runner(
 iso1="P17.C2",
 iso2="GCA.009102525.1.ASM910252v1")

Bxy_tree_file = file.path(dir,paste0("phylogeny/RAxML_bipartitions.Bacteroides_xylanisolvens"))
Bxy_tree = read.tree(Bxy_tree_file)
tree_dist <- cophenetic.phylo(Bxy_tree)
tree_distm <- melt_dist(tree_dist) %>% 
  rename(tree_dist=dist)
tree_distm = tree_distm[1:10,]

library(parallel)
library(doParallel)
cl <- makePSOCKcluster(30, setup_timeout = 0.5)
registerDoParallel(cl)
mapply(pairwise_gene_gain_runner,tree_distm$iso1,tree_distm$iso2)
stopCluster(cl)

files = list.files(file.path(pw_outdir),'summary.txt')
files = paste0(pw_outdir,'/',files)
res <- do.call(rbind, lapply(files, function(x) read_tsv(x,col_types = cols())))
res = as.data.frame(as.matrix(res))

```

