---
title: "Figures and tables"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir ='/stor/work/Ochman/alex/captive_ape_strain_genomes')
library(ape)
library(treeio)
library(ggtree)
library(estimatr)
library(tidyverse)
library(seqinr)
library(stringr)
library(cowplot)
library(harrietr)
library(vegan)
library(ggnewscale)
library(RColorBrewer)
library(Biostrings)
```


### Edit Bxy meteVisualize Bxy strain phylogeny
Label major clades and strain lineages isolated from captive apes in metadata 
Produces edited metadata and rooted phylogeny
```{r}
dir='results/pangenome/Bacteroides_xylanisolvens'
dir.create(file.path(dir,'output'))
species = 'Bacteroides_xylanisolvens'

#read-in metadata
Bxy_metadata_file = file.path(dir,"metadata.txt")
Bxy_metadata = read_tsv(Bxy_metadata_file,col_types = cols())
Bxy_metadata <- Bxy_metadata %>%
  mutate(taxonomy_Species = str_replace_all(taxonomy_Species,' ','_'),
         isolate = str_replace_all(isolate,'[_-]','.'),
         site = recode(site,'USA: Cambridge'='USA','USA:Boston'='USA',
                       'China: Shenzhen'='China','USA:Seattle'='USA',     
                       'USA:Baltimore'='USA', 'not applicable'='siteUnknown',
                       'not determined'='siteUnknown','Norway:Oslo'='Norway',
                       'missing'='siteUnknown')) %>%
  unite(host_site, host, site, sep = "_", remove = FALSE) 


##### Move this section
table(Bxy_metadata$host_site)
#look up and recode metadata missing from select samples
Bxy_metadata$isolate[Bxy_metadata$host_site=='missing_siteUnknown']
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002959635.1.ASM295963v1"] <- 'human_USA'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.900114865.1.IMG.taxon.2654588180.annotated.assembly"] <- 'rumen_USA'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.900107825.1.IMG.taxon.2623620516.annotated.assembly"] <- 'rumen_USA'
#Bxy_metadata$isolate[Bxy_metadata$host=='chicken']
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002161135.1.ASM216113v1"] <- 'chicken_Europe'
Bxy_metadata$host_site[Bxy_metadata$isolate==
                         "GCA.002161115.1.ASM216111v1"] <- 'chicken_Europe'
######


Bxy_metadata <- Bxy_metadata %>%
  mutate(host_site = recode(host_site,'human_Norway'='human',
                            'human_USA'='human',
                            'human_siteUnknown'='human'))
table(Bxy_metadata$host_site)

#root tree
outgroup_names <- Bxy_metadata %>% 
  filter(taxonomy_Species!=species) %>% 
  pull(isolate)
Bxy_tree_file = file.path(dir,paste0("phylogeny/RAxML_bipartitions.",species))
Bxy_tree <- ape::read.tree(Bxy_tree_file)
Bxy_tree$tip.label = str_replace_all(Bxy_tree$tip.label,'[_-]','.') #clean up names
Bxy_tree <- root(Bxy_tree,outgroup=outgroup_names) 
write.tree(Bxy_tree,file = file.path(dir,'output',paste0(species,'.tre')))
tree_plot <- ggtree(Bxy_tree) %<+% Bxy_metadata +
   geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
   geom_tippoint(aes(color=host_site), alpha=0.8)  + 
   geom_tiplab() + 
   xlim(NA,.5)

Bxy_tree <- drop.tip(Bxy_tree,outgroup_names)
#setdiff(Bxy_tree$tip.label,Bxy_metadata$isolate)
#setdiff(Bxy_metadata$isolate,Bxy_tree$tip.label)

get_color_palette <- function(tips) {
  Bxy_metadata <- Bxy_metadata %>% filter(isolate %in% tips)
  vec <- sort(unique(Bxy_metadata$host_site))
  return(recode(vec,
                          'human'='cadetblue4',
                          'human_China' = 'deepskyblue',
                          'rumen_USA' = 'brown4',
                          'bonobo_Columbus_Zoo'='red2',
                          'chimpanzee_Houston_Zoo'='orange2',
                          'orangutan_Houston_Zoo'='purple4',
                          'gorilla_Columbus_Zoo'='green3',
                          'chicken_Europe'='tan',
                          'missing_siteUnknown'='brown4'))}
host_site_color <- get_color_palette(Bxy_tree$tip.label)
Bxy_tree_plot <- ggtree(Bxy_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 90),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  geom_tiplab() + 
  xlim(NA,.5) +
  geom_treescale(x=.15,y=.25)
#ggsave(Bxy_tree_plot,file=file.path(dir,'output/Bxy_tree.pdf'),height=20)

cladeA_MCRA <- ape::getMRCA(Bxy_tree,
                            c('P17.D4','GCA.003474245.1.ASM347424v1'))
cladeA_tree <- extract.clade(Bxy_tree,cladeA_MCRA)
cladeA_tree_plot <- ggtree(cladeA_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeA_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

cladeB_MCRA <- ape::getMRCA(Bxy_tree,
                            c('P21.4E','GCA.009102685.1.ASM910268v1'))
cladeB_tree <- extract.clade(Bxy_tree,cladeB_MCRA)
cladeB_tree_plot <- ggtree(cladeB_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeB_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

cladeC_MCRA <- ape::getMRCA(Bxy_tree,
                            c('GCA.009102525.1.ASM910252v1','GCA.902374095.1.MGYG.HGUT.01345'))
cladeC_tree <- extract.clade(Bxy_tree,cladeC_MCRA)
cladeC_tree_plot <- ggtree(cladeC_tree) %<+% Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  +
  scale_colour_manual(values=get_color_palette(cladeC_tree$tip.label)) + 
  geom_tiplab() +
  xlim(NA,.5)

Bxy_metadata <- Bxy_metadata %>% mutate(clade = ifelse(
  isolate %in% cladeA_tree$tip.label,'cladeA',
    ifelse(isolate %in% cladeB_tree$tip.label,'cladeB',
      ifelse(isolate %in% cladeC_tree$tip.label,'cladeC','unassigned'))))

mixedhostNode  <- ape::getMRCA(Bxy_tree,c('P14.E4','P15.E10'))
mixedhostHuman <- c("GCA.003468875.1.ASM346887v1")
mixedhosttree <- extract.clade(Bxy_tree,mixedhostNode)
gorilla1Node  <- ape::getMRCA(Bxy_tree,c('P21.6D','P21.2G'))
gorilla1tree <- extract.clade(Bxy_tree,gorilla1Node)
gorilla1Human <- c("GCA.003458755.1.ASM345875v1")
gorilla2Node  <- ape::getMRCA(Bxy_tree,c('P21.4G','P21.11A'))
gorilla2tree <- extract.clade(Bxy_tree,gorilla2Node)
gorilla2Human <- c("GCA.003458755.1.ASM345875v1")
Bxy_metadata <- Bxy_metadata %>% mutate(captive_clade = ifelse(
  isolate %in% mixedhosttree$tip.label,'mixedhost',
    ifelse(isolate %in% gorilla1tree$tip.label,'gorilla1',
      ifelse(isolate %in% gorilla2tree$tip.label,'gorilla2','unassigned'))))
Bxy_metadata <- Bxy_metadata %>% 
  mutate(human_ape = 
  if_else(host %in% c('chimpanzee','bonobo','gorilla','orangutan'),
                                                  'ape',host))
write_tsv(Bxy_metadata,file=file.path(dir,'metadata_edit.txt'))

(Bxy_tree_plot_cladelabels <- ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 50),size=.75) +
  geom_tippoint(aes(color=host_site), alpha=0.8)  + 
  scale_colour_manual(values=get_color_palette(Bxy_tree$tip.label)) + 
  xlim(NA,.5)+ 
  geom_cladelabel(node=mixedhostNode, color='black', offset=.01,
                  label="Mixed-host clade") + 
  geom_cladelabel(node=gorilla1Node, color='black', offset=.01,label="Gorilla clade1") +
  geom_cladelabel(node=gorilla2Node, color='black', offset=.01,label="Gorilla clade2") +
  geom_cladelabel(node=cladeA_MCRA, color='black', offset=.14,label="CladeA") + 
  geom_cladelabel(node=cladeB_MCRA, color='black', offset=.13,label="CladeB") +
  geom_cladelabel(node=cladeC_MCRA, color='black', offset=.02,label="CladeC") +
  xlim(NA,.5) +
  geom_treescale(x=.05,y=.05))
#ggsave(Bxy_tree_plot_cladelabels,file=file.path(dir,'output/Bxy_tree_cladelabels.pdf'),height=20)
```

### Kruskal-Wallis test 
#####Identify gene function convergently enriched in strains isolated from captive apes relative to those isolated from humans in both major clades A and B 

```{r}
#inputs annotation, gene table, Bxy metdata
dir = 'results/pangenome/Bacteroides_xylanisolvens'
annotation = read_tsv(file= file.path(dir,'output/roary_nosplitparalogs_annotation.txt'),
                      col_types = cols())
gene_table = read_tsv(file= file.path(dir,'output/roary_nosplitparalogs_gene_table.txt'),
                      col_types = cols())
Bxy_metadata = read_tsv(file=file.path(dir,'metadata_edit.txt'),
                      col_types = cols())
#output: Kruskal-Wallis table copy number difference in function groups 
print(length(unique(annotation$func_group)))

#get gene counts for each func group for all isolates 
func <- annotation %>% 
  left_join(gene_table,by='Gene') %>%
  filter(!is.na(func_group)) %>%
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  select(func_group,func_annot,isolate,present) %>%
  group_by(func_group,func_annot,isolate) %>%
  summarise(count = sum(present))
head(func)

#add metadata
func <- Bxy_metadata %>% 
  select(isolate,host,human_ape,host_site,clade,captive_clade) %>% 
  left_join(func,by='isolate',fill=0)

#create new df that show mean gene count across all func groups in apes and humans
func_mean <- func %>% 
  group_by(func_group,func_annot,clade,human_ape) %>%
  summarize(mean_count = mean(count)) %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) 

#subset to just captive apes and humans from clades A and B
func_mean_wide <- func_mean %>%
  filter(human_ape=='human'|human_ape=='ape') %>%
  filter(clade!='unassigned') %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=T) %>%  
  pivot_wider(names_from = host_clade,values_from=mean_count) %>%
  filter(cladeA_ape != cladeA_human,cladeB_ape != cladeB_human) #some difference to observe
head(func_mean_wide)

#subset to just captive apes and humans from clades A and B
func_humanape <- func %>%
  filter(human_ape %in% c('human','ape'),clade!='unassigned')  %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) %>%
  filter(func_group %in% func_mean_wide$func_group)

#run kruskal wallis on functional groups
kw = func_humanape %>%
  filter(clade != 'cladeC') %>%
  group_by(func_group,func_annot,clade) %>%
  rstatix::kruskal_test(count ~ host_clade) 
kw$p_adj = p.adjust(kw$p, method = 'fdr')

#add back mean gene count info to kw results
kw_mean <- kw %>% 
  left_join(func_mean_wide,by=c('func_group','func_annot')) %>%
  mutate(ape_mean = ifelse(clade == 'cladeA', cladeA_ape, cladeB_ape))%>%
  mutate(human_mean = ifelse(clade == 'cladeA', cladeA_human, cladeB_human)) %>%
  mutate(diff = ape_mean-human_mean) %>%
  select(-c(cladeA_ape, cladeB_ape,cladeA_human, cladeB_human, cladeC_human)) %>%
  select(clade, ape_mean, human_mean, everything()) 

#identify func groups that are convergently enriched in captive apes
kw_convergent = kw_mean %>% 
  filter(p_adj < .001, ape_mean > human_mean)  %>% 
  add_count(func_group,name='count') %>%
  filter(count == 2) %>%
  filter(func_group != 'hypothetical protein')

#label func group in table
kw_mean = kw_mean %>% 
  mutate(converge_captive = if_else(
    func_group %in% kw_convergent$func_group & ape_mean > human_mean, 1, 0)) %>%
  arrange(desc(converge_captive),desc(diff)) %>%
  select(func_group,func_annot,clade,ape_mean,human_mean,method,n,statistic,df,p,p_adj,converge_captive)
head(kw_mean)
write_tsv(kw_mean,file=file.path(dir,'output/table_kw_func_groups.txt'))

```

### Determine whether sulfatase functions within COG3119 are enriched in Bxy strains isolated from captive apes
```{r}
#inputs: annotation, gene table, Bxy metdata
#output: Kruskal-Wallis table copy number difference in sulfatase functions 

#get gene counts for each func group for all isolates 
COG3119  <- annotation %>% 
  left_join(gene_table,by='Gene') %>%
  filter(eggnog_best_OG_name=='COG3119@2|Bacteria') %>%
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
  select(Annotation,isolate,present) %>%
  group_by(Annotation,isolate) %>%
  summarise(count = sum(present))
head(COG3119)

#add metadata
COG3119 <- Bxy_metadata %>% 
  select(isolate,host,human_ape,host_site,clade,captive_clade) %>% 
  left_join(COG3119,by='isolate',fill=0)

#add mean gene count across all apes and humans
COG3119_mean <- COG3119 %>% 
  group_by(Annotation,clade,human_ape) %>%
  summarize(mean_count = mean(count)) %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) 

#subset to just captive apes and humans from clades A and B
COG3119_mean_wide <- COG3119_mean %>%
  filter(human_ape=='human'|human_ape=='ape') %>%
  filter(clade!='unassigned') %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=T) %>%  
  pivot_wider(names_from = host_clade,values_from=mean_count) %>%
  filter(cladeA_ape != cladeA_human,cladeB_ape != cladeB_human) #some difference to observe
head(COG3119_mean_wide)

#subset to just captive apes and humans from clades A and B
COG3119_humanape <- COG3119 %>%
  filter(human_ape %in% c('human','ape'),clade!='unassigned')  %>%
  unite(col=host_clade,c(clade,human_ape),sep='_',remove=F) %>%
  filter(Annotation %in% COG3119_mean_wide$Annotation)

#run kruskal wallis on functional groups
kw_COG3119 = COG3119_humanape  %>%
  filter(clade != 'cladeC') %>%
  group_by(Annotation,clade) %>%
  rstatix::kruskal_test(count ~ host_clade) 
kw_COG3119$p_adj = p.adjust(kw_COG3119$p, method = 'fdr')

#add back mean gene count info to kw results
kw_COG3119_mean <- kw_COG3119 %>% 
  left_join(COG3119_mean_wide,by=c('Annotation')) %>%
  mutate(ape_mean = ifelse(clade == 'cladeA', cladeA_ape, cladeB_ape))%>%
  mutate(human_mean = ifelse(clade == 'cladeA', cladeA_human, cladeB_human)) %>%
  mutate(diff = ape_mean-human_mean) %>%
  select(-c(cladeA_ape, cladeB_ape,cladeA_human, cladeB_human, cladeC_human)) %>%
  select(clade, ape_mean, human_mean, everything()) 

#identify func groups that are convergently enriched in captive apes
kw_COG3119_convergent = kw_COG3119_mean %>% 
  filter(p_adj < .001, ape_mean > human_mean)  %>% 
  add_count(Annotation,name='count') %>%
  filter(count == 2) 

#label func group in table
kw_COG3119_mean = kw_COG3119_mean %>% 
  mutate(converge_captive = if_else(
    Annotation %in% kw_COG3119_convergent$Annotation & ape_mean > human_mean, 1, 0)) %>%
  arrange(desc(converge_captive),desc(diff)) %>%
  select(Annotation,clade,ape_mean,human_mean,method,n,statistic,df,p,p_adj,converge_captive)
print(kw_COG3119_mean)
write_tsv(kw_COG3119_mean,file=file.path(dir,'output/table_kw_sulfatase_annotation.txt'))
```

### Figure 1
Visualize strain host site info, # predicted genes, and copy number for gene functions related to mucin and carrageenan biosynthesis
```{r}
#inputs: annotation,gene table
get_func_copynumber_table <- function(list_of_func){
  #given list of function groups returns table with gene copy numbers for each isolate
  func_copynumber_table  <- annotation %>% 
    left_join(gene_table,by='Gene') %>%
    filter(func_group %in% list_of_func) %>%
    pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
    select(func_group,isolate,present) %>%
    group_by(func_group,isolate) %>%
    summarise(count = sum(present)) %>%
    pivot_wider(names_from = func_group, values_from = count) %>% 
    column_to_rownames(var='isolate')
    func_copynumber_table[func_copynumber_table==0] <- NA
  return(func_copynumber_table)
}  

mucin <- get_func_copynumber_table(c('COG3119@2|Bacteria_K01136_NA', 
                                     'COG3119@2|Bacteria_K01565_NA', 
                                     'COG1649@2|Bacteria_K05970_NA', 
                                     'COG5434@2|Bacteria_NA_GH110' ,
                                     'COG3525@2|Bacteria_K12373_GH20', 
                                     'COG3250@2|Bacteria_K01195_GH2', 
                                     'COG3250@2|Bacteria_K01190_GH2' 
                                   )) 

mucin <- mucin %>% as.data.frame() %>%
                                    dplyr::rename('iduronate 2-sulfatase(K01136)'='COG3119@2|Bacteria_K01136_NA',
                                           'N-sulfoglucosamine sulfohydrolase(K01565)'='COG3119@2|Bacteria_K01565_NA',
                                           'sialate O-acetylesterase(K05970)'='COG1649@2|Bacteria_K05970_NA',
                                           'alpha-1,3-galactosidase(GH110)'='COG5434@2|Bacteria_NA_GH110',
                                           'hexosaminidase(K12373)'='COG3525@2|Bacteria_K12373_GH20',
                                           'beta-glucuronidase(K01195)'='COG3250@2|Bacteria_K01195_GH2',
                                           'beta-galactosidase(K01190)'='COG3250@2|Bacteria_K01190_GH2')

carrageenen <- get_func_copynumber_table(c('COG1874@2|Bacteria_NA_GH167','NA_NA_GH167',
                                           '33PQM@2|Bacteria_NA_GH150','COG5434@2|Bacteria_NA_GH82')) 
carrageenen <- carrageenen %>% as.data.frame() %>%
                                        dplyr::rename('Lambda-carrageenase(GH150)'='33PQM@2|Bacteria_NA_GH150',
                                               'Iota-carrageenase(GH82)'='COG5434@2|Bacteria_NA_GH82',
                                               'GH167a'='COG1874@2|Bacteria_NA_GH167',
                                               'GH167b'='NA_NA_GH167') %>%
                                        mutate(beta_carrageenase=GH167a+GH167b) %>%
                                        dplyr::rename('beta_carrageenase(GH167)'='beta_carrageenase') %>%
                                        select(-GH167a,-GH167b)

get_annot_copynumber_table <- function(list_of_annot){
  #given list of function groups returns table with gene copy numbers for each isolate
  annot_copynumber_table  <- annotation %>% 
    left_join(gene_table,by='Gene') %>%
    filter(Annotation %in% list_of_annot) %>%
    pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>%
    select(Annotation,isolate,present) %>%
    group_by(Annotation,isolate) %>%
    summarise(count = sum(present)) %>%
    pivot_wider(names_from = Annotation, values_from = count) %>% 
    column_to_rownames(var='isolate')
    annot_copynumber_table[annot_copynumber_table==0] <- NA
  return(annot_copynumber_table)
}  
sulfatase = get_annot_copynumber_table(c('N-acetylgalactosamine-6-O-sulfatase',
                                         'N-acetylglucosamine-6-O-sulfatase',
                                         'Ulvan-active sulfatase','Arylsulfatase')) 
sulfatase = sulfatase %>% as.data.frame() %>%
                  dplyr::rename('N-acetylgalactosamine-6-O-sulfatase(K01132)'='N-acetylgalactosamine-6-O-sulfatase',
                         'N-acetylglucosamine-6-O-sulfatase(K01137)'='N-acetylglucosamine-6-O-sulfatase'
                         )
sulfa_mucin = cbind(sulfatase,mucin)   
sulfa_mucin = sulfa_mucin[c("N-acetylgalactosamine-6-O-sulfatase(K01132)",
              "N-acetylglucosamine-6-O-sulfatase(K01137)",
              'Arylsulfatase',
              'Ulvan-active sulfatase',
              "sialate O-acetylesterase(K05970)",
              "alpha-1,3-galactosidase(GH110)",
              "hexosaminidase(K12373)",
              "beta-glucuronidase(K01195)",
              "beta-galactosidase(K01190)")]
HOST_table <- Bxy_metadata %>% 
  select(isolate,host_site) %>%
  column_to_rownames(var='isolate') 

predicted.genes = Bxy_metadata %>% select(isolate,predicted.genes) %>%
  mutate(predicted.genes=as.numeric(predicted.genes)) %>%
  column_to_rownames(var='isolate')

#format tree
Bxy_tree = read.tree(file = file.path(dir,'output/Bacteroides_xylanisolvens.tre'))
outgroup_names <- Bxy_metadata %>% 
  filter(taxonomy_Species!='Bacteroides_xylanisolvens') %>% 
  pull(isolate)
Bxy_tree <- drop.tip(Bxy_tree,outgroup_names)
tree_v1 = ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 90),size=.75) +
  geom_treescale()
tree_v2 = tree_v1 + #add clade labels
   geom_cladelabel(node=cladeA_MCRA,label="cladeA",offset = .005)+
   geom_cladelabel(node=cladeB_MCRA,label="cladeB",offset = .005)+
   geom_cladelabel(node=cladeC_MCRA,label="cladeC",offset = .005)+
   geom_cladelabel(node=mixedhostNode,label="mixedhost",offset = .01) +
   geom_cladelabel(node=gorilla1Node,label="gorilla1",offset = .01)+ 
   geom_cladelabel(node=gorilla2Node,label="gorilla2",offset = .01) +
   geom_tippoint(aes(subset=(label=='GCA.000210075.1.ASM21007v1')),color='red') +
   xlim(NA,.25)
p <- gheatmap(tree_v2 + ylim(-10,NA),
    HOST_table,width=.1,colnames_angle=90,hjust=1,offset=.05)  + 
    scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label))
p <- p + new_scale_fill()
p1b <- gheatmap(p,predicted.genes,colnames_angle=90,hjust=1,offset=.06,width=.1) + 
    scale_fill_viridis_c(direction = -1, option="D")
p1b <- p1b + new_scale_fill()
p2 <- gheatmap(p1b,sulfa_mucin,colnames_angle=90,hjust=1,offset=.07,width=.9) + 
    scale_fill_viridis_c(direction = -1, option="B")
p2 <- p2 + new_scale_fill()
(p3 <- gheatmap(p2,carrageenen,colnames_angle=90,hjust=1,offset=.13,width=.3)+ 
    scale_fill_viridis_c(option="D",direction = -1))
ggsave(p3,file = file.path(dir,'output/Figure1_long.pdf'),width = 10, height=12)
ggsave(p3,file = file.path(dir,'output/Figure1_short.pdf'),width = 10, height=6)
```


### Figure 2: Compare distribution of sulfatase genes in XB1A model to strains isolated from captive apes

#### Aggregate faas files into a single faa file
```{bash}
#ls results/pangenome/Bacteroides_xylanisolvens/faa/
#cat results/pangenome/Bacteroides_xylanisolvens/faa/* > results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa
```

```{r}
#filter to orthologous gene groups that are within the sulfatlas
sulfatase_genes <- annotation %>% filter(sulfatlas_family=='S1') %>% pull(Gene)

present_in_isolates <- function(list_of_genes,list_of_isolates,name) {
  #given list of genes and isolates, determine the total number of ge
  isolate_gene_table <- annotation %>% 
    left_join(gene_table,by='Gene') %>%
    filter(Gene %in% list_of_genes) %>%
    select(Gene,list_of_isolates) %>%
    mutate(total = rowSums(select_(., "-Gene"))) %>%
    select(Gene,total)  
  isolate_gene_table = isolate_gene_table %>%
         mutate(binary = ifelse(total>0,name,NA),
         prop = total/max(total)) %>%
         column_to_rownames('Gene') 
  print(head(isolate_gene_table))
  isolate_gene_table = isolate_gene_table %>% select(binary)
  colnames(isolate_gene_table) <-c(name)
  return(isolate_gene_table)
}

#determine which of the sulfatase_genes are present in XBA1 the captive ape lineages
XBA1 = c("GCA.000210075.1.ASM21007v1")
XBA1 = present_in_isolates(sulfatase_genes,XBA1,'XBA1')
captive_isolates = Bxy_metadata %>% filter(human_ape=='ape') %>% pull(isolate)
ape = present_in_isolates(sulfatase_genes,captive_isolates,'ape')
mixedhost_isolates =  Bxy_metadata %>% filter(captive_clade=='mixedhost') %>% pull(isolate)
mixedhost = present_in_isolates(sulfatase_genes,mixedhost_isolates,'mixedhost')
gor1_isolates =  Bxy_metadata %>% filter(captive_clade=='gorilla1') %>% pull(isolate)
gor1 = present_in_isolates(sulfatase_genes,gor1_isolates,'gorilla1')
gor2_isolates =  Bxy_metadata %>% filter(captive_clade=='gorilla2') %>% pull(isolate)
gor2 = present_in_isolates(sulfatase_genes,gor2_isolates,'gorilla2')
sulfadata <- cbind(XBA1,ape,mixedhost,gor1,gor2)

sulfadata <- sulfadata %>%
  rownames_to_column(var='Gene') %>%
  select(Gene,XBA1,mixedhost,gorilla1,gorilla2) %>%
  left_join(annotation,by='Gene')
head(sulfadata)

##make tree of sulfa genes
#faa_file = 'results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa'
#db = readAAStringSet(faa_file)
#names(db) <- sapply(strsplit(names(db)," "), `[`, 1)
#prot_seqs = db[names(db) %in% sulfadata$Gene.ID]
file = file.path('results/pangenome/Bacteroides_xylanisolvens/output/sulfatase.faa')
#writeXStringSet(prot_seqs,file)
#system(paste0('mafft --auto ',file,' > ',file,'.align'))
#system(paste0('fasttree ',file,'.align > ',file,'.tree'))
sulfa_tree <- read.tree(paste0(file,'.tree'))
get_group <- function(prot){sulfadata$Gene[sulfadata$Gene.ID == prot]}
sulfa_tree$tip.label <- as.character(lapply(sulfa_tree$tip.label,get_group))

pres <- sulfadata %>% 
  dplyr::select(Gene,XBA1,mixedhost,gorilla1,gorilla2) %>% 
  column_to_rownames(var='Gene')
func <- sulfadata %>% 
  dplyr::select(Gene,Annotation) %>% 
  column_to_rownames(var='Gene') %>% as.matrix() 
subfamily <- sulfadata %>% 
  dplyr::select(Gene,sulfatlas_subfamily) %>% 
  filter(Gene %in% sulfa_tree$tip.label) %>% 
  mutate(sulfatase_subfamily = as.character(sulfatlas_subfamily)) %>%
  column_to_rownames(var='Gene') %>% as.matrix() 
colourCount = length(unique(subfamily[,'sulfatlas_subfamily']))
subfamily_pal = colorRampPalette(brewer.pal(colourCount, "Paired"))(colourCount)

not_in_XBA1 = rownames(XBA1)[is.na(XBA1$XBA1)]
sulfa_tree_gg = ggtree(sulfa_tree,layout='circular') +
  geom_point2(aes(subset=(label%in%not_in_XBA1)), 
              shape=21, size=1, fill='black') + 
  geom_treescale() 

got_subfam_node <- function(sub){
  subfam = sulfadata$Gene[sulfadata$sulfatlas_subfamily==sub]
  subfam = intersect(subfam,sulfa_tree$tip.label)
  subfam_node = getMRCA(sulfa_tree,subfam)
  return(subfam_node)
}
got_subfam_node('4')
table(sulfadata$sulfatlas_subfamily)
nodes = lapply(unique(sulfadata$sulfatlas_subfamily),got_subfam_node)
names(nodes) = unique(sulfadata$sulfatlas_subfamily)
unlist(nodes)

sulfa_tree_gg_v2 = sulfa_tree_gg +
  geom_hilight(node=79, fill="darkgreen", alpha=.1) + 
  geom_cladelabel(node=79,label="S1_15") +  
  geom_hilight(node=94, fill="blue", alpha=.1) +
  geom_cladelabel(node=94,label="S1_9") +  
  geom_hilight(node=71, fill="red", alpha=.1) +
  geom_cladelabel(node=71,label="S1_20") +  
  geom_hilight(node=93, fill="orange", alpha=.1) +
  geom_cladelabel(node=93,label="S1_11") +
  geom_hilight(node=99, fill="yellow", alpha=.1) + 
  geom_cladelabel(node=99,label="S1_67") +  
  geom_hilight(node=84, fill="purple", alpha=.1) +
  geom_cladelabel(node=84,label="S1_4") +  
  geom_hilight(node=75, fill="royalblue", alpha=.1) +
  geom_cladelabel(node=75,label="S1_14") +  
  geom_hilight(node=58, fill="pink", alpha=.1) +
  geom_cladelabel(node=58,label="S1_8") +
  geom_hilight(node=78, fill="darkorange", alpha=.1) +
  geom_cladelabel(node=78,label="S1_22") +  
  geom_hilight(node=89, fill="green", alpha=.1) +
  geom_cladelabel(node=89,label="S1_16") +  
  geom_hilight(node=55, fill="red2", alpha=.1) +
  geom_cladelabel(node=55,label="S1_30") +  
  geom_hilight(node=100, fill="yellow4", alpha=.1) +
  geom_cladelabel(node=100,label="S1_28") + 
  geom_treescale() 

p <- gheatmap(sulfa_tree_gg_v2, pres, colnames_angle=90,hjust=1,width=.4)  + scale_fill_manual(values=brewer.pal(4, "Dark2"))
p <- p + new_scale_fill()
(p2 <- gheatmap(p,func,colnames_angle=90,hjust=1,offset=.99,width=.05) +
    scale_fill_manual(values = brewer.pal(9, "Paired")))
ggsave(p2,file=file.path(dir,'output/Figure2_sulfatase_circle.pdf'),height=12,width=10)
```

#### D
```{r}
faa_file = 'results/pangenome/Bacteroides_xylanisolvens/output/all_prot.faa'
db = readAAStringSet(faa_file)
names(db) <- sapply(strsplit(names(db)," "), `[`, 1)
system('mkdir results/pangenome/Bacteroides_xylanisolvens/output/og_seqs')

```


#### Figure 3
```{r}
dir = 'results/pangenome/Bacteroides_xylanisolvens'
species = 'Bacteroides_xylanisolvens'
Bxy_metadata = read_tsv(file=file.path(dir,'metadata_edit.txt'),col_types = cols())
Bxy_metadata_subset <- Bxy_metadata %>% dplyr::select(isolate,host,human_ape,captive_clade)
annotation = read_tsv(file=file.path(dir,'output/roary_nosplitparalogs_annotation.txt'),
                      col_types = cols())
gene_table = read_tsv(file=file.path(dir,'output/roary_nosplitparalogs_gene_table.txt'),
                      col_types = cols())
Bxy_tree <- read.tree(file = file.path(dir,paste0(species,'.tre')))

captive_clade_table <- gene_table %>% 
  pivot_longer(cols=Bxy_metadata$isolate,
               names_to='isolate',values_to='present') %>% 
  filter(present>0) %>%
  left_join(Bxy_metadata_subset,by='isolate') %>%
  group_by(Gene,captive_clade) %>%
  tally() %>%
  pivot_wider(names_from=captive_clade,values_from=n,values_fill=0) 
  
captive_clade_table <- captive_clade_table %>%  
  mutate(mixedhost = as.numeric(mixedhost),
         gorilla1 = as.numeric(gorilla1),
         gorilla2 = as.numeric(gorilla2),
         cladecount = sum(if_else(mixedhost>0 & (mixedhost>.5*max(captive_clade_table$mixedhost)),1,0),
                          if_else(gorilla1>0 & (gorilla1>.5*max(captive_clade_table$gorilla1)),1,0),
                          if_else(gorilla2>0 & (gorilla2>.5*max(captive_clade_table$gorilla2)),1,0))) 
captive_clade_table <- captive_clade_table %>%  
  filter(unassigned<3,cladecount>1) 
captive_convergent_genes <- captive_clade_table$Gene
captive_gene_table <- gene_table %>% 
  filter(Gene %in% captive_convergent_genes) %>%
  column_to_rownames(var='Gene') %>%
  t()
HOST_table <- Bxy_metadata %>% 
  dplyr::select(isolate,host_site) %>%
  column_to_rownames(var='isolate') 

p <- gheatmap(ggtree(Bxy_tree) + ylim(-10,NA),
              HOST_table,width=.1,colnames_angle=90,hjust=1)  +
              scale_fill_manual(values=get_color_palette(Bxy_tree$tip.label)) 
p <- p + new_scale_fill()
(p2 <- gheatmap(p,captive_gene_table,offset=.05,colnames_angle=90,hjust=1) +
                scale_fill_viridis_c(option="C"))
ggsave(p2,file =file.path(dir,'output/convergenetly_gained_captive.pdf'))
```

```{r warning = FALSE, message=FALSE}
source('scripts/genomic_island_function.R')
#system(paste0('rm -r ',file.path(dir,'captive_convergent_genes')))
list_of_genes = captive_convergent_genes
pres_abs <- read_csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),
                     col_types = cols())
metadata = read_tsv(file.path(dir,'metadata_edit.txt'),col_types = cols())
window_size=10
get_captive_convergent_genes = function(isolate){

    isolate_old = metadata$isolate_old[metadata$isolate==isolate]
    gff_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.gff')
    genomic_islands_outfile = file.path(dir,'captive_convergent_genes',paste0('gff_window',window_size),
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
    gff_islands = get_genomic_island(pres_abs=pres_abs,
                               list_of_genes=captive_convergent_genes,
                               isolate=isolate,
                               gff_file=gff_file,
                               window_size = window_size,
                               outfile=genomic_islands_outfile)
}

get_captive_convergent_genes('P17.C2')
captive_isolates = Bxy_metadata$isolate[Bxy_metadata$human_ape=='ape']
for (isolate in captive_isolates) {
  print(isolate)
  get_captive_convergent_genes(isolate)
}

#generate heatmap only with clusters longer than 5 genes
captive_PUL_table = data.frame()
captive_PUL_table = data.frame(Gene=captive_convergent_genes)
for  (isolate in captive_isolates) {
  genomic_islands_outfile = file.path(dir,'captive_convergent_genes',paste0('gff_window',window_size),
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
  iso = read_tsv(file=genomic_islands_outfile)
  iso = iso %>% filter(size>5) %>%
  select(Gene,gene_span) %>% 
  mutate(Gene=make.names(Gene, unique=TRUE)) 
  colnames(iso) = c('Gene',isolate)
  captive_PUL_table = captive_PUL_table %>% full_join(iso,by='Gene')}

captive_PUL_tableM = captive_PUL_table %>% 
  column_to_rownames(var='Gene') %>%
  arrange_all() %>% t()
not_captive = Bxy_tree$tip.label[!Bxy_tree$tip.label %in% captive_isolates]
Bxy_tree_captive = drop.tip(Bxy_tree,not_captive)
(cluster_heatmap_plot <- gheatmap(ggtree(Bxy_tree_captive) +geom_tiplab() + ylim(-10,NA),captive_PUL_tableM,colnames_angle=90,hjust=1,offset = .01) +
    scale_fill_manual(values=colorRampPalette(brewer.pal(15, "Paired"))(15)))
ggsave(cluster_heatmap_plot,file=file.path(dir,'captive_convergent_genes',paste0('cluster_window',window_size,'_heatmap.pdf')))
       
captive_PUL_table_df = captive_PUL_table %>% 
  pivot_longer(cols=all_of(captive_isolates),names_to='isolate',values_to='gene_span') %>%
  select(gene_span,isolate) %>% 
  distinct() %>%
  filter(!is.na(gene_span))

updown_size=100
gene_span='group_16996_group_20520'
isolate = 'P17.C2'

get_island_fna = function(isolate,gene_span){

    isolate_old = metadata$isolate_old[metadata$isolate==isolate]
    gff_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.gff')
    fna_file = paste0('results/pangenome/prokka/',isolate_old,'/',isolate_old,'.fna')
    genomic_islands_outfile = file.path(dir,'captive_convergent_genes',paste0('gff_window',window_size),
                    paste0(isolate,'_window',window_size,'_island_gff.txt'))
    island_fasta_outfile = file.path(dir,'captive_convergent_genes',paste0('fna_window',window_size,'updown',updown_size),gene_span,
                                 paste0(isolate,'_',gene_span,'_updown_',updown_size,'.fna'))
    print(island_fasta_outfile)
    output_island_fasta(
          isolate=isolate,
          gene_span = gene_span,
          genomic_islands_outfile = genomic_islands_outfile,
          island_fasta_outfile  = island_fasta_outfile,
          updown_size = updown_size ,
          fna_file = fna_file,
          gff_file = gff_file
          )
}

mcmapply(get_island_fna,captive_PUL_table_df$isolate,captive_PUL_table_df$gene_span,mc.cores=8)
```

```{r}
species = 'Bacteroides_xylanisolvens'
dir = paste0('results/pangenome/',species)
Bxy_sum = read_tsv(file.path(dir,'pairwise_gene_gain_50strains_summary.txt'),col_types = cols())
Bxy_gi = read_tsv(file.path(dir,'pairwise_gene_gain_50strains_gff_islands.txt'),col_types = cols())

df = Bxy_sum
dist_cutoff = .01
df_filt = filter(df,tree_dist>dist_cutoff)

#Pangenome dist
summary(lm_robust(dist_bray_curtis ~ tree_dist,data = df))
summary(lm_robust(dist_bray_curtis ~ I(tree_dist^2) + tree_dist,data = df))
cor.test(df$dist_bray_curtis,df$tree_dist,method = 'spearman')

bray = ggplot(Bxy_sum,aes(x=tree_dist,y=dist_bray_curtis)) + 
  geom_point(color='blue4',alpha = .05,size=3) + 
  theme_bw() +
  stat_smooth(method = "lm", color = 'blue4',formula = y ~ x + I(x^2), size = 1) +
  ylab('Pangenome distance (Bray-Curtis)')

#Events
summary(lm_robust(n_unique_genes ~ tree_dist,data = df))
summary(lm_robust(n_unique_genes ~ I(tree_dist^2) + tree_dist,data = df))
cor.test(df$n_unique_genes,df$tree_dist,method = 'spearman')
summary(lm_robust(bray_CN100 ~ tree_dist ,data = df))
summary(lm_robust(bray_CN100 ~ I(tree_dist^2) + tree_dist, data = df))
cor.test(df$bray_CN100,df$tree_dist,method = 'spearman')
(events = ggplot(df,aes(x=tree_dist)) + 
  geom_point(aes(y=n_unique_genes),color='darkorange',alpha = .05,size=3) + 
  stat_smooth(aes(y=n_unique_genes), 
              method = "lm", color = 'darkorange',formula = y ~ I(x^2) + x, size = 1) +
  geom_point(aes(y=bray_CN100),color='purple',alpha = .1,size=3)   +
  stat_smooth(aes(y=bray_CN100), 
              method = "lm", color = 'purple',formula = y ~ I(x^2) + x, size = 1) +
    theme_bw() +
    scale_y_continuous(
    name = "Number of unique genes",
    sec.axis = sec_axis( trans=~., name="Number of events")
    ))

##Fractions of events/genes by size 
Bxy_gi_filt = Bxy_gi %>% 
  filter(tree_dist>0) %>% 
  mutate(cat='gene') %>%
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,210)))
mean(Bxy_gi_filt$size)
median(Bxy_gi_filt$size)

Bxy_gi_clust = Bxy_gi_filt %>% 
  distinct(comp,tree_dist,name,size) %>%
  mutate(cat='event') %>%
  mutate(size_category=cut(size, breaks=c(0,5,10,15,20,25,50,100,210)))
mean(Bxy_gi_clust$size)
median(Bxy_gi_clust$size)

Bxy_gi_combined = Bxy_gi_filt %>% add_row(Bxy_gi_clust) %>% 
  group_by(cat,size_category) %>%
  summarise(count = n()) %>%
  mutate(freq = count / sum(count)) %>% 
  as.data.frame()

(frac_events_genes = Bxy_gi_combined %>% 
  ggplot(aes(x=size_category,y=freq,fill=cat)) +
  geom_bar(stat="identity", position=position_dodge()) + 
  theme_bw() +
  xlab('Genomic island size')+
  scale_y_continuous(
    name = "Fraction of events",
    sec.axis = sec_axis( trans=~., name="Fraction of genes")
    )+ 
  theme(legend.position = c(.95, .95), 
       legend.justification = c(.95, .95)
       ))

#island size
island_size = Bxy_gi_clust %>% 
  group_by(comp,tree_dist) %>% 
  summarize(mean_size = mean(size))

summary(lm_robust(mean_size~tree_dist,island_size))
summary(lm_robust(mean_size~I(tree_dist^2) + tree_dist,island_size))
cor.test(island_size$mean_size,island_size$tree_dist,method = 'spearman')
summary(lm_robust(mean_size~tree_dist,filter(island_size,tree_dist>.01)))
summary(lm_robust(mean_size~I(tree_dist^2) + tree_dist,filter(island_size,tree_dist>.01)))
cor.test(filter(island_size,tree_dist>.01)$mean_size,filter(island_size,tree_dist>.01)$tree_dist,method = 'spearman')

island_size_plot = island_size %>% 
  ggplot(aes(x=tree_dist,y=mean_size)) + 
  geom_point(color='red2',alpha = .05,size=3)   +
  stat_smooth(data = filter(island_size,tree_dist>.01),aes(y=mean_size), 
              method = "lm", color = 'red2',formula = y ~ I(x^2) + x, size = 1) +
  theme_bw()  +
  ylab('Average genomic island size')

(figure4 = plot_grid(bray,events,island_size_plot,frac_events_genes,nrow=2,labels=c('A','B','C','D')))
ggsave(figure4,file =file.path(dir,'output/figure4.pdf'),width=8,height=8)

```


```{r}
species = 'Bacteroides_xylanisolvens'
dir = paste0('results/pangenome/',species)
Bxy = read_tsv('results/pangenome/Bacteroides_xylanisolvens/pairwise_gene_gain_50strains.txt',col_types = cols())
Bov = read_tsv('results/pangenome/Bacteroides_ovatus/pairwise_gene_gain_50strains.txt',col_types = cols())
Bfr = read_tsv('results/pangenome/Bacteroides_fragilis/pairwise_gene_gain_50strains.txt',col_types = cols())
Bth = read_tsv('results/pangenome/Bacteroides_thetaiotaomicron/pairwise_gene_gain_50strains.txt',col_types = cols())
Bac = Bxy %>% add_row(Bov) %>% add_row(Bfr) %>% add_row(Bth)


Bac <- Bac %>% group_by(taxonomy_Species.iso1)%>% mutate(
  max_tree_dist = max(tree_dist),
  tree_dist_norm = tree_dist/mean(tree_dist)) %>% as.data.frame()

Bac_tree_dist_sum = Bac %>% 
  group_by(taxonomy_Species.iso1) %>% 
  summarize(min_tdn = min(tree_dist_norm),max_tdn = max(tree_dist_norm))
Bac <- Bac %>%  filter(tree_dist_norm>max(Bac_tree_dist_sum$min_tdn),
                       tree_dist_norm<min(Bac_tree_dist_sum$max_tdn))

summary(lm_robust(dist_bray_curtis ~ I(tree_dist_norm^2) + tree_dist_norm + taxonomy_Species.iso1, data = Bac))
summary(lm_robust(dist_bray_curtis ~  tree_dist_norm + taxonomy_Species.iso1, data = Bac))

figure5 = Bac %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=dist_bray_curtis) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=1,size=2) +
  theme_bw()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~ I(x^2) + x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~  I(x^2) + x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  I(x^2) + x, size = 1, se=TRUE)+
  ylab('Pangenome distance (Bray-Curtis)')
  
ggsave(figure5,file = file.path(dir,'output/figure5_bcdist.pdf'),width=8,height=6)

summary(lm_robust(bray_CN100 ~ I(tree_dist_norm^2) + tree_dist_norm + taxonomy_Species.iso1, data = Bac))
summary(lm_robust(bray_CN100 ~  tree_dist_norm + taxonomy_Species.iso1, data = Bac))
(figure5B = Bac %>% 
  ggplot() + 
  aes(x=tree_dist_norm,y=bray_CN100) +
  geom_point(aes(color=taxonomy_Species.iso1),alpha=1,size=2) +
  theme_bw()  + 
  scale_colour_manual(values=c('red2','blue','green2','orange')) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_fragilis'),
             method = "lm",color='red2', formula = y ~ x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_ovatus'),
             method = "lm",color='blue', formula = y ~  x, size = 1, se=TRUE)+
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_thetaiotaomicron'),
             method = "lm",color='green2', formula = y ~ x, size = 1, se=TRUE) +
  stat_smooth(data = filter(Bac,taxonomy_Species.iso1 == 'Bacteroides_xylanisolvens'),
              method = "lm",color='orange', formula = y ~  x, size = 1, se=TRUE)+
    ylab('Number of events'))

ggsave(figure5B,file = file.path(dir,'output/figure5_CN100.pdf'),width=8,height=6)



```
