---
title: "pwcomps"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir ='/stor/work/Ochman/alex/captive_ape_strain_genomes')
library(ape)
library(treeio)
library(ggtree)
library(estimatr)
library(tidyverse)
library(seqinr)
library(stringr)
library(cowplot)
library(harrietr)
library(vegan)
library(harrietr)
library(parallel)
library(data.table)

```

```{r}
source('scripts/genomic_island_function.R')
outdir = ('results/geneLossGain/')
dir.create(outdir)
```


```{r}
get_phylo_independent_comps = function(species,cutoff,outfile) {
  #identify phylogenetically independent pairs of strains
  #given the min cutoff tree dist
  
  #read in input files
  print(species)
  dir = paste0('results/pangenome/',species)
  metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
  tree = read.tree(file.path(dir,paste0(species,'.tre')))
  outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
  tree = drop.tip(tree,outgroup)
  tree_dist <- cophenetic.phylo(tree)
  pw_df <- melt_dist(tree_dist) %>% 
    dplyr::rename(tree_dist=dist) %>% 
    filter(tree_dist > as.numeric(cutoff))
  
  #add metadata to pw comparisons
  metadata_subset <- metadata %>% dplyr::select(isolate,host,sample,taxonomy_Species)
  pw_df_meta  <- pw_df  %>%    
    left_join(metadata_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and 
    left_join(metadata_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) %>%
    filter(host.iso1 != 'missing',host.iso2 != 'missing') %>%
    mutate(diff_host = if_else(host.iso1 != host.iso2, 1, 0)) %>%
    mutate(ape_host = if_else(host.iso1 %in% c('chimp','bonobo','orangutan','gorilla')|
                                host.iso2 %in% c('chimp','bonobo','orangutan','gorilla'),1,0)) %>%
    arrange(desc(diff_host),tree_dist) %>% 
    mutate(comp = paste(iso1,iso2,sep='_')) 
  
  identify_taxa <- function(iso1,iso2) {
    #determine all isolates have the same MRCA as two input isolates
    node = getMRCA(tree,c(iso1,iso2))
    subtree = extract.clade(tree, node)
    return(subtree$tip.label)
  }
  
  taxa = c()
  rows = c()
  for (row in 1:nrow(pw_df_meta)) { #loop through tree distm
    iso1=pw_df_meta[row,1]
    iso2=pw_df_meta[row,2]
    subtree_taxa = identify_taxa(iso1,iso2)
    #add row to list if there's no phylogenetic overlap with isolate comparison already present in the   list
    if (length(intersect(subtree_taxa,taxa))==0){ 
      taxa = c(taxa,subtree_taxa)  #keep track of isolates included in comparisons            
      rows=c(rows,row)
      print(row)
    }
  }
  PI_comps = pw_df_meta[rows,]
  PI_comps$compNum = paste0('P',as.list(1:nrow(PI_comps)))
  write_tsv(PI_comps,file = outfile)
  return(PI_comps)

}

Bxy_PI_pairs = get_phylo_independent_comps('Bacteroides_xylanisolvens',.008,
                                           file.path(outdir,'Bxy_PI_pairs.txt'))
Bov_PI_pairs = get_phylo_independent_comps('Bacteroides_ovatus',.005,
                                           file.path(outdir,'Bov_PI_pairs.txt'))
Bfr_PI_pairs = get_phylo_independent_comps('Bacteroides_fragilis',.0001,
                                           file.path(outdir,'Bfr_PI_pairs.txt'))
Bth_PI_pairs = get_phylo_independent_comps('Bacteroides_thetaiotaomicron',.002,
                                           file.path(outdir,'Bth_PI_pairs.txt'))
makeLong = function(PI_comps){
  print(PI_comps)
  iso1 = PI_comps %>% select_if(stringr::str_detect(names(.), "iso1")|
                                    names(.) %in%
                                c('tree_dist','diff_host','ape_host','comp','compNum'))
  iso2 = PI_comps %>% select_if(stringr::str_detect(names(.), "iso2")|
                                    names(.) %in%
                                c('tree_dist','diff_host','ape_host','comp','compNum')) 
  
  colnames(iso1) = c('isolate',"tree_dist","host","sample","taxonomy_Species",
                   "diff_host","ape_host",'comp','compNum')
  colnames(iso2) = c('isolate',"tree_dist","host","sample","taxonomy_Species",
                   "diff_host","ape_host",'comp','compNum')
  PI_pairs = rbind(iso1,iso2)
  return(PI_pairs)
} 

Bxy_PI_pairs_long = makeLong(Bxy_PI_pairs)
Bov_PI_pairs_long = makeLong(Bov_PI_pairs)
Bfr_PI_pairs_long = makeLong(Bfr_PI_pairs)
Bth_PI_pairs_long = makeLong(Bth_PI_pairs)

```
```{r}
species = 'Bacteroides_ovatus'
PI_pairs_df = Bov_PI_pairs
offset_val = .01  

visualize_PI_pairs <- function(species,PI_pairs_df, offset_val) {
  
dir = paste0('results/pangenome/',species)
metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
tree = read.tree(file.path(dir,paste0(species,'.tre')))
outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
tree = drop.tip(tree,outgroup)  
tree_PI = drop.tip(tree,setdiff(tree$tip.label,PI_pairs_df$isolate))  
 
HOST_table <- metadata %>% 
  select(isolate,host) %>%
  column_to_rownames(var='isolate') 
HOST_table_PI = HOST_table
HOST_table_PI$host[!rownames(HOST_table_PI) %in% PI_pairs_df$isolate] <- NA

predicted.genes = metadata %>% select(isolate,predicted.genes) %>%
  mutate(predicted.genes=as.numeric(predicted.genes)) %>%
  column_to_rownames(var='isolate')
predicted.genes_PI = predicted.genes
predicted.genes_PI$predicted.genes[!rownames(predicted.genes_PI) %in% 
                           PI_pairs_df$isolate] <- NA

get_color_palette <- function(tips) {
  metadata <- metadata %>% filter(isolate %in% tips)
  vec <- sort(unique(metadata$host))
  return(recode(vec,
                          'human'='cadetblue4',
                          'rumen' = 'brown4',
                          'bonobo'='red2',
                          'chimpanzee'='orange2',
                          'orangutan'='purple4',
                          'gorilla'='green3',
                          'chicken'='tan',
                          'mouse' = 'yellow2',
                          'pig' = 'pink',
                          'missing'='white'))}


tree_full = ggtree(tree) %<+% 
  PI_pairs_df +
  geom_tippoint(aes(subset = label %in% c(PI_pairs_df$isolate),
                    color = factor(compNum))) +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
  geom_treescale() 
tree_full <- gheatmap(tree_full + ylim(-10,NA),colnames = FALSE,
    HOST_table,width=.1,colnames_angle=90,hjust=1)  + 
    scale_fill_manual(values=get_color_palette(tree$tip.label))
tree_full <- tree_full + new_scale_fill()
tree_full <- gheatmap(tree_full,predicted.genes,colnames_angle=90,hjust=1,width=.1,offset = offset_val) + 
    scale_fill_viridis_c(direction = -1, option="D")
print(tree_full)

tree_PI_plot = ggtree(tree_PI) %<+% 
  PI_pairs_df +
  geom_tippoint(aes(subset = label %in% c(PI_pairs_df$isolate),
                    color = factor(compNum))) +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
  geom_treescale() 
tree_PI_plot <- gheatmap(tree_PI_plot + ylim(-10,NA),colnames = FALSE,
    HOST_table,width=.1,colnames_angle=90,hjust=1)  + 
    scale_fill_manual(values=get_color_palette(tree_PI$tip.label))
tree_PI_plot <- tree_PI_plot + new_scale_fill()
tree_PI_plot <- gheatmap(tree_PI_plot,predicted.genes,
                         colnames_angle=90,hjust=1,width=.1,offset = offset_val) + 
    scale_fill_viridis_c(direction = -1, option="D")
print(tree_PI_plot)
}

visualize_PI_pairs('Bacteroides_xylanisolvens',Bxy_PI_pairs_long,offset_val = .008)
visualize_PI_pairs('Bacteroides_ovatus',Bov_PI_pairs_long,offset_val = .012)
visualize_PI_pairs('Bacteroides_fragilis',Bfr_PI_pairs_long,offset_val = .001)
visualize_PI_pairs('Bacteroides_thetaiotaomicron',Bth_PI_pairs_long,offset_val = .004)
```

```{r}
pairwise_gene_gain_runner = function(iso1,iso2,species) {
  #set input filepaths to run pairwise_gene_gain
  dir = paste0('results/pangenome/',species)
  metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
  pres_abs_file <- file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv')
  pw_outdir = file.path(dir,'pairwise_gene_gain')
  iso1_old = metadata$isolate_old[metadata$isolate==iso1]
  iso2_old = metadata$isolate_old[metadata$isolate==iso2]
  iso1_gff_file = paste0('results/pangenome/prokka/',iso1_old,'/',iso1_old,'.gff')
  iso2_gff_file = paste0('results/pangenome/prokka/',iso2_old,'/',iso2_old,'.gff')
  countTREE_file=file.path(dir,paste0("count/count_",species,".tre"))
  countOutput_file=file.path(dir,paste0("count/countOutput_",species,"_gainpenalty2.FAMILY"))
  window_size = 2
  pairwise_gene_gain(
    #loaded from functions script
    pres_abs_file  = pres_abs_file,
    iso1=iso1,
    iso2=iso2,
    iso1_gff_file = iso1_gff_file,
    iso2_gff_file = iso2_gff_file,
    window_size = window_size,
    countTREE_file=countTREE_file,
    countOutput_file=countOutput_file,
    pw_outdir = pw_outdir)
}

#pairwise_gene_gain_runner('P21.11A','GCA.003458755.1.ASM345875v1','Bacteroides_xylanisolvens')
#mapply(pairwise_gene_gain_runner,Bxy_PI_pairs$iso1,Bxy_PI_pairs$iso2,'Bacteroides_xylanisolvens')
#mapply(pairwise_gene_gain_runner,Bov_PI_pairs$iso1,Bov_PI_pairs$iso2,'Bacteroides_ovatus')
#mapply(pairwise_gene_gain_runner,Bfr_PI_pairs$iso1,Bfr_PI_pairs$iso2,'Bacteroides_fragilis')
#mapply(pairwise_gene_gain_runner,Bth_PI_pairs$iso1,Bth_PI_pairs$iso2,'Bacteroides_thetaiotaomicron')
```


```{r}

get_summary_output <- function(species,window_size,suffix,outdir){
  outfile = file.path(outdir,paste0(species,'_window',window_size,'_',suffix,'.txt'))
  dir = paste0('results/pangenome/',species)
  pw_outdir = file.path(dir,'pairwise_gene_gain')
  all_files = list.files(file.path('results/pangenome',species,'pairwise_gene_gain'))
  summary_files = all_files[str_detect(all_files,suffix)]
  summary_files = summary_files[str_detect(summary_files,paste0('window',1))]
  summary_files = unlist(paste0(file.path(pw_outdir),'/',summary_files))
  summary_df  = rbindlist(lapply(summary_files,function(x){read.csv(x,sep='\t')}))
  write_tsv(summary_df,file =outfile)
  return(summary_df)
}

Bxy_summary = get_summary_output('Bacteroides_xylanisolvens',2,'summary',outdir)
Bov_summary = get_summary_output('Bacteroides_ovatus',2,'summary',outdir)
Bfr_summary = get_summary_output('Bacteroides_fragilis',2,'summary',outdir)
Bth_summary = get_summary_output('Bacteroides_thetaiotaomicron',2,'summary',outdir)

Bxy_island = get_summary_output('Bacteroides_xylanisolvens',2,'island',outdir)
Bov_island = get_summary_output('Bacteroides_ovatus',2,'island',outdir)
Bfr_island = get_summary_output('Bacteroides_fragilis',2,'island',outdir)
Bth_island = get_summary_output('Bacteroides_thetaiotaomicron',2,'island',outdir)

Bxy_gff = get_summary_output('Bacteroides_xylanisolvens',2,'gff',outdir)
Bov_island = get_summary_output('Bacteroides_ovatus',2,'island',outdir)
Bfr_island = get_summary_output('Bacteroides_fragilis',2,'island',outdir)
Bth_island = get_summary_output('Bacteroides_thetaiotaomicron',2,'island',outdir)
```



```{r}
#summarize gene gain numbers and events
summary_df = Bxy_summary
iso1 = summary_df %>% select(iso1,iso1_gain_events,iso1_gain)
colnames(iso1) = c('isolate','gain_events','gain_numGenes')
iso2 = summary_df %>% select(iso2,iso2_gain_events,iso2_gain)
colnames(iso2) = c('isolate','gain_events','gain_numGenes')
gain = rbind(iso1,iso2) %>% 
  as.data.frame() %>% 
  mutate(diff_gain_numGenes = gain_numGenes-gain_events) %>%
  select(isolate,gain_events,diff_gain_numGenes) %>%
  pivot_longer(cols=c(gain_events,diff_gain_numGenes),names_to='cat',values_to='count') %>%
  mutate(count=as.numeric(count),cat=as.factor(cat)) %>%
  as.data.frame()

#summarize gene loss numbers and events
iso1 = summary_df %>% select(iso1,iso1_loss_events,iso1_loss)
colnames(iso1) = c('isolate','loss_events','loss_numGenes')
iso2 = summary_df %>% select(iso2,iso2_loss_events,iso2_loss)
colnames(iso2) = c('isolate','loss_events','loss_numGenes')
loss = rbind(iso1,iso2) %>% 
  as.data.frame() %>% 
  mutate(diff_loss_numGenes = loss_numGenes-loss_events) %>%
  select(isolate,loss_events,diff_loss_numGenes) %>%
  pivot_longer(cols=c(loss_events,diff_loss_numGenes),names_to='cat',values_to='count') %>%
  mutate(count=as.numeric(count),cat=as.factor(cat)) %>%
  as.data.frame()

tree_PI_plot = tree_PI_plot +   new_scale_fill()
p2 <- facet_plot(tree_PI_plot, panel = 'Gene Loss/Gain', data = loss, 
                geom = geom_barh, 
                mapping = aes(x = -count, fill = as.factor(cat)), 
                stat='identity')
p3 <- facet_plot(p2, panel = 'Gene Loss/Gain', data = gain, 
                geom = geom_barh, 
                mapping = aes(x = count, fill = as.factor(cat)), 
                stat='identity')
top = p3  + 
  theme_bw() + 
  scale_fill_manual(values=c('pink','lightblue','firebrick','blue4')) +
    theme(legend.position="top")
top
```




```{r}
identify_pw_50strain = function(species) {
  #reads in relevant data files
  #identify 50strains based on hclust of tree distances 
  #returns flattened dist matrix of all 1225 pw comparison and metadata
  
  #read in input files
  print(species)
  dir = paste0('results/pangenome/',species)
  metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
  pres_abs <- read_csv(file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv'),col_types = cols())
  tree = read.tree(file.path(dir,paste0("phylogeny/RAxML_bipartitions.",species)))
  #outdir
  pw_outdir = file.path(dir,'pairwise_gene_gain')
  dir.create(pw_outdir)
  #create distance matrix from tree
  outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
  tree = drop.tip(tree,outgroup)
  tree_dist <- cophenetic.phylo(tree)
  
  #dereplication closely related strains
  set.seed(123)
  hc = hclust(as.dist(tree_dist))
  cl = cutree(hc,k=50)
  df = data.frame(cl) %>% 
    rownames_to_column(var='isolate') %>%
    group_by(cl) %>% sample_n(1)
  pw_df <- melt_dist(tree_dist) %>% 
    dplyr::rename(tree_dist=dist) %>%
    filter(iso1 %in% df$isolate, iso2 %in% df$isolate)
  pw_df = pw_df %>% 
    mutate(comp = paste(iso1,iso2,sep='_')) 
  #add metadata to pw comparisons
  metadata_subset <- metadata %>% dplyr::select(isolate,host,sample,taxonomy_Species,predicted.genes)
  pw_df  <- pw_df  %>%    
    left_join(metadata_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and 
    left_join(metadata_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) 
  return(pw_df)
  }
Bxy_pw = identify_pw_50strain('Bacteroides_xylanisolvens')
Bov_pw = identify_pw_50strain('Bacteroides_ovatus')
Bfr_pw = identify_pw_50strain('Bacteroides_fragilis')
Bth_pw = identify_pw_50strain('Bacteroides_thetaiotaomicron')

species = 'Bacteroides_xylanisolvens'
print(species)
dir = paste0('results/pangenome/',species)
Bxy_metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
Bxy_metadata_50pw = Bxy_metadata %>% filter(isolate %in% c(Bxy_pw$iso1,Bxy_pw$iso2))
summary(Bxy_metadata_50pw$predicted.genes)


species = 'Bacteroides_ovatus'
print(species)
dir = paste0('results/pangenome/',species)
Bov_metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
Bov_metadata_50pw = Bov_metadata %>% filter(isolate %in% c(Bov_pw$iso1,Bov_pw$iso2))
summary(Bov_metadata_50pw$predicted.genes)

species = 'Bacteroides_fragilis'
print(species)
dir = paste0('results/pangenome/',species)
Bfr_metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
Bfr_metadata_50pw = Bfr_metadata %>% filter(isolate %in% c(Bov_pw$iso1,Bov_pw$iso2))
summary(Bfr_metadata_50pw$predicted.genes)
```




```{r}

```





```{r}

get_summary_output <- function(species){
  dir = paste0('results/pangenome/',species)
  pw_outdir = file.path(dir,'pairwise_gene_gain')
  all_files = list.files(file.path('results/pangenome',species,'pairwise_gene_gain'))
  summary_files = all_files[str_detect(all_files,'summary')]
  summary_files = unlist(paste0(file.path(pw_outdir),'/',summary_files))
  summary_df  = rbindlist(lapply(summary_files,function(x){read.csv(x,sep='\t')}))
  return(summary_df)
}
get_island_output <- function(species){
  dir = paste0('results/pangenome/',species)
  pw_outdir = file.path(dir,'pairwise_gene_gain')
  all_files = list.files(file.path('results/pangenome',species,'pairwise_gene_gain'))
  island_files = all_files[str_detect(all_files,'island')]
  island_files = unlist(paste0(file.path(pw_outdir),'/',island_files))
  island_df  = rbindlist(lapply(island_files,function(x){read.csv(x,sep='\t')}))
  return(island_df) }


Bxy_summary_df = get_summary_output('Bacteroides_xylanisolvens')
Bxy_island_df = get_island_output('Bacteroides_xylanisolvens')
```




```{r}
island_df = Bxy_island_df
df = island_df  %>% 
    separate(col=geneGainLoss,into=c('iso','gainloss'),remove=F,sep='_') %>%
    mutate(iso_category = 
          dplyr::if_else(iso == 'iso1',paste0(iso1,gainloss),paste0(iso2,gainloss)), 
    size_category=cut(cluster_size, breaks=c(0,5,10,15,20,25,50,250)), 
    size_category = recode(size_category,
                                '(0,5]'='01-5',
                                '(5,10]'='06-10',
                                '(10,15]'='11-15',
                                '(15,20]'='16-20',
                                '(20,25]'='21-25',
                                '(25,50]'='26-50',
                                '(50,250]'='50-250'))
    
  
prop_df = df %>% group_by(iso_category,gainloss,size_category) %>%
    summarise(number_events = n(),
              number_genes = sum(cluster_size)) 
prop_df = prop_df %>% group_by(gainloss,iso_category) %>%
         mutate(total_events = sum(number_events),
                prop_events = number_events / total_events,
                total_genes = sum(number_genes),
                prop_genes= number_genes / total_genes) 
prop_df = prop_df %>% 
  select(iso_category,size_category,prop_events,prop_genes) %>%
  pivot_longer(cols = c(prop_events,prop_genes), names_to = 'category', values_to = 'proportion') %>%
  pivot_wider(names_from = size_category,values_from = proportion,values_fill = 0 ) %>%
  pivot_longer(cols = unique(df$size_category), names_to = 'size_category', values_to = 'proportion')
prop_df$gainloss <- factor(prop_df$gainloss, levels = c('loss','gain'))

#visualize gain events vs prop
gain_loss_events_prop = prop_df  %>% 
      filter(category=='prop_events') %>% 
      ggplot(aes(x=size_category,y=proportion,fill=gainloss)) +
      geom_boxplot(position=position_dodge(width = 1)) + 
      theme_bw() +
      xlab('Event size')+
      scale_y_continuous(
        name = "Proportion of events",
        ) + 
      theme(
        legend.position = c(.95, .95), 
        legend.justification = c(.95, .95)
        ) +
      scale_fill_manual(values = c('blue4','firebrick'))
gain_loss_events_prop 

gain_loss_gene_prop = prop_df %>% 
      filter(category=='prop_genes') %>% 
      ggplot(aes(x=size_category,y=proportion,fill=gainloss)) +
      geom_boxplot(position=position_dodge(width = .9)) + 
      theme_bw() +
      xlab('Event size')+
      scale_y_continuous(
        name = "Proportion of genes",
        ) + 
      theme(
        legend.position = c(.95, .95), 
        legend.justification = c(.95, .95)
        ) +
      scale_fill_manual(values = c('lightblue','pink'))
gain_loss_gene_prop 

bottom = plot_grid(gain_loss_events_prop  ,gain_loss_gene_prop)

plot_grid(top,bottom,ncol=1)
```


```{r}

df = island_df  %>% 
    separate(col=geneGainLoss,into=c('iso','gainloss'),remove=F,sep='_') %>%
    mutate(iso_category = 
          dplyr::if_else(iso == 'iso1',paste0(iso1,gainloss),paste0(iso2,gainloss)), 
    size_category=cut(cluster_size, breaks=c(0,5,10,15,20,25,50,250)), 
    size_category = recode(size_category,
                                '(0,5]'='01-5',
                                '(5,10]'='06-10',
                                '(10,15]'='11-15',
                                '(15,20]'='16-20',
                                '(20,25]'='21-25',
                                '(25,50]'='26-50',
                                '(50,250]'='50-250'))

P50 = function(isolate) {
  res = 0
  half = sum(vec)*.5
  print(half)
  for (x in sort(vec,decreasing = TRUE)) {
    res = sum(res,x)
    if (res>half){
      return(x)
    }
  }}

```


```{r}
dup = island_df  %>% 
    separate(col=geneGainLoss,into=c('iso','gainloss'),remove=F,sep='_') %>%
    mutate(iso_category = 
             dplyr::if_else(iso == 'iso1',paste0(iso1,gainloss),paste0(iso2,gainloss))) %>%
    mutate(size_category=cut(cluster_size, breaks=c(0,5,10,15,20,25,50,250))) %>%
    group_by(iso_category,gainloss,size_category,is_dup) %>%
    summarise(number_events = n(),
              number_genes = sum(cluster_size)) 

dup  %>% 
      filter(size_category=='(0,5]') %>% 
      ggplot(aes(x=is_dup,y=proportion,fill=gainloss)) +
      geom_boxplot(position=position_dodge(width = .9)) + 
      theme_bw() +
      xlab('Event size')+
      scale_y_continuous(
        name = "Proportion of genes",
        ) + 
      theme(
        legend.position = c(.95, .95), 
        legend.justification = c(.95, .95)
        ) +
      scale_fill_manual(values = c('lightblue','pink'))
```

