---
title: "Untitled"
output: html_document
---

Generates all figures and tables
```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)
knitr::opts_knit$set(root.dir ='/stor/work/Ochman/alex/captive_ape_strain_genomes')
library(ape)
library(treeio)
library(ggtree)
library(estimatr)
library(tidyverse)
library(seqinr)
library(stringr)
library(cowplot)
library(harrietr)
library(vegan)
library(ggnewscale)
library(RColorBrewer)
library(Biostrings)
```

#### Input files
```{r}
#inputs annotation, gene table, Bxy metdata,all_prot
Bxydir = 'results/pangenome/Bacteroides_xylanisolvens'
indir = file.path('results/analyses_input_files')
outdir = file.path('results/output_figures_tables')
dir.create(outdir)
annotation = read_tsv(file= file.path(indir,'Bxy_roary_nosplitparalogs_annotation.txt'),
                      col_types = cols())
gene_table = read_tsv(file= file.path(indir,'Bxy_roary_nosplitparalogs_gene_table.txt'),
                      col_types = cols())
Bxy_metadata = read_tsv(file=file.path(indir,'Bxy_metadata.txt'),
                      col_types = cols())
Bxy_tree = read.tree(file = file.path(indir,'Bacteroides_xylanisolvens.tre'))
pres_abs =  read_tsv(file= file.path(indir,'roary_nosplitparalogs_gene_presence_absence.txt'),
                      col_types = cols())

#files 
sulfatase_prot_file = file.path(indir,'sulfatase.faa')
OG_AAI = read_tsv(file=file.path(indir,'sulfatase_OG_AAI.txt'))
sulfa_blastp = read_tsv(file.path(indir,'sulfatase_blastp.txt'),
                      comment='#',col_names = FALSE, col_types = cols())
```


### Figure 1
```{r}
species = 'Bacteroides_xylanisolvens'
offset_val1 = .025 
offset_val2 = .03 
dir = paste0('results/pangenome/',species)
tree_file = file.path(dir,paste0(species,'.tre'))
metadata_file = file.path(dir,'metadata.txt')
PI_pairs_file = file.path("results/geneLossGain/Bxy_PI_pairs.txt")

####

metadata = read_tsv(metadata_file,col_types = cols())
PI_pairs = read_tsv(PI_pairs_file,col_types = cols())
Bxy_metadata = metadata

makeLong = function(PI_comps){
  iso1 = PI_comps %>% select_if(stringr::str_detect(names(.), "iso1")|
                                    names(.) %in%
                                c('tree_dist','diff_host','ape_host','comp','compNum'))
  iso2 = PI_comps %>% select_if(stringr::str_detect(names(.), "iso2")|
                                    names(.) %in%
                                c('tree_dist','diff_host','ape_host','comp','compNum')) 
  
  colnames(iso1) = c('isolate',"tree_dist","host","sample","taxonomy_Species",
                   "diff_host","ape_host",'comp','compNum')
  colnames(iso2) = c('isolate',"tree_dist","host","sample","taxonomy_Species",
                   "diff_host","ape_host",'comp','compNum')
  PI_pairs = rbind(iso1,iso2)
  return(PI_pairs)
} 

PI_pairs_long = makeLong(PI_pairs)

Bxy_tree = read.tree(tree_file)
outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
Bxy_tree = drop.tip(Bxy_tree,outgroup) 

tree_v1 = ggtree(Bxy_tree) %<+% 
  Bxy_metadata +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 90),size=.75) +
  geom_treescale()

cladeA_MCRA <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$clade=='cladeA'])
cladeB_MCRA <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$clade=='cladeB'])
cladeC_MCRA <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$clade=='cladeC'])
mixedhostNode <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$captive_clade=='mixedhost'])
gorilla1Node <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$captive_clade=='gorilla1'])
gorilla2Node <- ape::getMRCA(Bxy_tree,Bxy_metadata$isolate[Bxy_metadata$captive_clade=='gorilla2'])
tree_v2 = ggtree(Bxy_tree) + #add clade labels
   geom_cladelabel(node=cladeA_MCRA,label="cladeA",offset = .005)+
   geom_cladelabel(node=cladeB_MCRA,label="cladeB",offset = .005)+
   geom_cladelabel(node=cladeC_MCRA,label="cladeC",offset = .005)+
   geom_cladelabel(node=mixedhostNode,label="mixedhost",offset = .01) +
   geom_cladelabel(node=gorilla1Node,label="gorilla1",offset = .01)+ 
   geom_cladelabel(node=gorilla2Node,label="gorilla2",offset = .01) +
   geom_tippoint(aes(subset=(label=='GCA.000210075.1.ASM21007v1')),color='red') 

get_color_palette <- function(tips,metadata) {
  metadata <- metadata %>% filter(isolate %in% tips)
  vec <- sort(unique(metadata$host))
  return(recode(vec,
                          'human'='cadetblue4',
                          'rumen' = 'brown4',
                          'bonobo'='red2',
                          'chimpanzee'='orange2',
                          'orangutan'='purple4',
                          'gorilla'='green3',
                          'chicken'='tan',
                          'mouse' = 'yellow2',
                          'pig' = 'pink',
                          'missing'='white'))}


make_full_tree_heatmap <- function(metadata,gg_tree,PI_pairs_long,offset_val1,offset_val2) {
  
HOST_table <- metadata %>% 
  select(isolate,host) %>%
  column_to_rownames(var='isolate') 

predicted.genes = metadata %>% select(isolate,predicted.genes) %>%
  mutate(predicted.genes=as.numeric(predicted.genes)) %>%
  column_to_rownames(var='isolate')

tree_full = gg_tree %<+% 
  PI_pairs_long +
  geom_tippoint(aes(subset = label %in% c(PI_pairs_long$isolate),
                    color = factor(compNum))) +
  geom_nodepoint(aes(subset = suppressWarnings(as.numeric(label)) > 75),size=.75) +
  geom_treescale() 
tree_full_heatmap <- gheatmap(tree_full + ylim(-10,NA),colnames = FALSE,offset=offset_val1,
    HOST_table,width=.1,colnames_angle=90,hjust=1)  + 
    scale_fill_manual(values=get_color_palette(gg_tree$data$label,metadata))
tree_full_heatmap  <- tree_full_heatmap + new_scale_fill()
tree_full_heatmap  <- gheatmap(tree_full_heatmap,predicted.genes,
                               colnames_angle=90,hjust=1,width=.1,offset = offset_val2) + 
    scale_fill_viridis_c(direction = -1, option="D")
return(tree_full_heatmap)
}

(Bxy_tree_full_heatmap = make_full_tree_heatmap(Bxy_metadata,tree_v2,
                                                PI_pairs_long,offset_val1,offset_val2)) 


```

#Figure S1
```{r}
species = 'Bacteroides_fragilis'
offset_val1 = .000
offset_val2 = .001  
dir = paste0('results/pangenome/',species)
tree_file = file.path(dir,paste0(species,'.tre'))
metadata_file = file.path(dir,'metadata.txt')
PI_pairs_file = file.path("results/geneLossGain/Bfr_PI_pairs.txt")


metadata = read_tsv(metadata_file,col_types = cols())
Bfr_metadata = metadata
tree = read.tree(tree_file)
outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
tree = drop.tip(tree,outgroup)  
PI_pairs = read_tsv(PI_pairs_file,col_types = cols())
PI_pairs_long = makeLong(PI_pairs)
(Bfr_tree_full_heatmap = make_full_tree_heatmap(Bfr_metadata,ggtree(tree),
                                                PI_pairs_long,offset_val1,offset_val2)) 

######
######
######

species = 'Bacteroides_ovatus'
offset_val1 = 0
offset_val2 = .012   
dir = paste0('results/pangenome/',species)
tree_file = file.path(dir,paste0(species,'.tre'))
metadata_file = file.path(dir,'metadata.txt')
PI_pairs_file = file.path("results/geneLossGain/Bov_PI_pairs.txt")
Bov_metadata = read_tsv(metadata_file,col_types = cols())
######
tree = read.tree(tree_file)
outgroup = Bov_metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
tree = drop.tip(tree,outgroup)  
PI_pairs = read_tsv(PI_pairs_file,col_types = cols())
PI_pairs_long = makeLong(PI_pairs)
(Bov_tree_full_heatmap = make_full_tree_heatmap(Bov_metadata,ggtree(tree),
                                                PI_pairs_long,offset_val1,offset_val2))

######
######
######

species = 'Bacteroides_thetaiotaomicron'
offset_val1 = 0
offset_val2 = .004   
dir = paste0('results/pangenome/',species)
tree_file = file.path(dir,paste0(species,'.tre'))
metadata_file = file.path(dir,'metadata.txt')
PI_pairs_file = file.path("results/geneLossGain/Bth_PI_pairs.txt")
Bth_metadata = read_tsv(metadata_file,col_types = cols()) ###
######
tree = read.tree(tree_file)
outgroup = Bth_metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
tree = drop.tip(tree,outgroup)  
PI_pairs = read_tsv(PI_pairs_file,col_types = cols())
PI_pairs_long = makeLong(PI_pairs)
(Bth_tree_full_heatmap = make_full_tree_heatmap(Bth_metadata,ggtree(tree),
                                                PI_pairs_long,offset_val1,offset_val2))

```

### Figure 2
```{r}
species = 'Bacteroides_xylanisolvens'
identify_pw_50strain = function(species) {
  #reads in relevant data files
  #identify 50strains based on hclust of tree distances 
  #returns flattened dist matrix of all 1225 pw comparison and metadata
  
  #read in input files
  print(species)
  dir = paste0('results/pangenome/',species)
  metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
  tree_file = file.path(dir,paste0(species,'.tre'))
  tree = read.tree(tree_file)
  
  #create distance matrix from tree
  outgroup = metadata %>% filter(taxonomy_Species!=species) %>% pull(isolate)
  tree = drop.tip(tree,outgroup)
  tree_dist <- cophenetic.phylo(tree)
  
  #dereplication closely related strains
  set.seed(123)
  hc = hclust(as.dist(tree_dist))
  cl = cutree(hc,k=50)
  df = data.frame(cl) %>% 
    rownames_to_column(var='isolate') %>%
    group_by(cl) %>% sample_n(1)
  return(df$isolate)

}

Bxy_rep50 = identify_pw_50strain('Bacteroides_xylanisolvens')
Bov_rep50 = identify_pw_50strain('Bacteroides_ovatus')
Bfr_rep50 = identify_pw_50strain('Bacteroides_fragilis')
Bth_rep50 = identify_pw_50strain('Bacteroides_thetaiotaomicron')
```

```{r}
bray_tree_dist = function(species) {
  
  rep50 = identify_pw_50strain(species)
  dir = paste0('results/pangenome/',species)
  metadata = read_tsv(file.path(dir,'metadata.txt'),col_types = cols())
  
  tree_file = file.path(dir,paste0(species,'.tre'))
  tree = read.tree(tree_file)
  tree_dist <- cophenetic.phylo(tree)[rep50,rep50]
  tree_dist.1 = melt_dist(tree_dist) %>%
    dplyr::rename(tree_dist=dist) 
  
   pres_abs_file <- file.path(dir,'roary_nosplitparalogs/gene_presence_absence.csv')
   pres_abs <- read_csv(pres_abs_file,col_types = cols())
    isblank <- function(x) {as.numeric(str_count(x, pattern = "_"))}
    gene_table <- pres_abs %>% 
      select(Gene,rep50)  %>%
      mutate_at(vars(-Gene),isblank) %>%
      column_to_rownames(var = 'Gene') %>% 
      as.matrix()
  gene_table[is.na(gene_table)] <- 0
  dist_bray <- as.matrix(vegdist(t(gene_table), method="bray",full=T))
  dist_bray = dist_bray[rep50,rep50]
  dist_bray.1 = melt_dist(dist_bray) %>%
      dplyr::rename(bray_curtis=dist) 
  pw_df_bray = tree_dist.1 %>% left_join(dist_bray.1,by=c('iso2','iso1'))
  
  metadata_subset <- metadata %>% dplyr::select(isolate,host,sample,taxonomy_Species)
  pw_df_bray  <- pw_df_bray  %>%    
    left_join(metadata_subset,by=c('iso1'='isolate')) %>%  #add metadata for individual1 and 
    left_join(metadata_subset,by=c('iso2'='isolate'),suffix = c(".iso1", ".iso2")) 
  return(pw_df_bray)
}


Bxy_tree_dist = bray_tree_dist("Bacteroides_xylanisolvens")
Bov_tree_dist = bray_tree_dist("Bacteroides_ovatus")
Bfr_tree_dist = bray_tree_dist("Bacteroides_fragilis")
Bth_tree_dist = bray_tree_dist("Bacteroides_thetaiotaomicron")


```

```{r}
Bt_metadata = Bxy_metadata %>% 
  add_rows(Bov_metadata) %>% 
  add_rows(Bfr_metadata) %>% 
  add_rows(Bth_metadata)
Bt_rep = c(Bxy_rep50,Bov_rep50,Bfr_rep50,Bth_rep50)
Bt_metadata %>% 
  filter(isolate %in% Bt_rep) %>% 
  ggplot(aes(x =taxonomy_Species,y=predicted.genes,fill=host)) +
  geom_dotplot(binaxis = "y", stackdir = "centerwhole") +
  theme_bw()

Bt_pd
```

